// z-schema/dist/schema-cache.mjs
import { getQueryPath, getRemotePath, isAbsoluteUri } from "./utils/uri.mjs";
import { findId, getId } from "./json-schema.mjs";
import { deepClone } from "./utils/clone.mjs";
import { decodeJSONPointer } from "./utils/json.mjs";
import { Report } from "./report.mjs";
import { normalizeOptions } from "./z-schema-options.mjs";
//#region src/schema-cache.ts
function getSafeRemotePath(uri) {
	const remotePath = getRemotePath(uri);
	if (!remotePath) return;
	if (remotePath === "__proto__" || remotePath === "constructor" || remotePath === "prototype") return;
	return remotePath;
}
const getEffectiveId = (schema) => {
	let id = getId(schema);
	if ((!id || !isAbsoluteUri(id)) && typeof schema.id === "string" && isAbsoluteUri(schema.id)) id = schema.id;
	return id;
};
/**
* Shared logic for registering a remote reference schema.
* Used by both the static `ZSchema.setRemoteReference()` (global cache) and
* the instance `validator.setRemoteReference()` (instance cache).
*/
function prepareRemoteSchema(schema, uri, validationOptions, maxCloneDepth) {
	let _schema;
	if (typeof schema === "string") _schema = JSON.parse(schema);
	else _schema = deepClone(schema, maxCloneDepth);
	if (!_schema.id) _schema.id = uri;
	if (validationOptions) _schema.__$validationOptions = normalizeOptions(validationOptions);
	return _schema;
}
var SchemaCache = class SchemaCache {
	static global_cache = Object.create(null);
	cache = Object.create(null);
	constructor(validator) {
		this.validator = validator;
	}
	static cacheSchemaByUri(uri, schema) {
		const remotePath = getSafeRemotePath(uri);
		if (remotePath) this.global_cache[remotePath] = schema;
	}
	cacheSchemaByUri(uri, schema) {
		const remotePath = getSafeRemotePath(uri);
		if (remotePath) this.cache[remotePath] = schema;
	}
	removeFromCacheByUri(uri) {
		const remotePath = getSafeRemotePath(uri);
		if (remotePath) delete this.cache[remotePath];
	}
	checkCacheForUri(uri) {
		const remotePath = getSafeRemotePath(uri);
		return remotePath ? this.cache[remotePath] != null : false;
	}
	getSchema(report, refOrSchema) {
		if (Array.isArray(refOrSchema)) return refOrSchema.map((i) => this.getSchema(report, i));
		if (typeof refOrSchema === "string") return this.getSchemaByUri(report, refOrSchema);
		return deepClone(refOrSchema, this.validator.options.maxRecursionDepth);
	}
	fromCache(path) {
		if (path === "__proto__" || path === "constructor" || path === "prototype") return;
		let found = this.cache[path];
		if (found) return found;
		found = SchemaCache.global_cache[path];
		if (found) {
			const clone = deepClone(found, this.validator.options.maxRecursionDepth);
			if (!clone.id || !isAbsoluteUri(clone.id) && isAbsoluteUri(path)) clone.id = path;
			this.cache[path] = clone;
			return clone;
		}
	}
	getSchemaByUri(report, uri, root) {
		if (root && !isAbsoluteUri(uri)) {
			const rootId = getEffectiveId(root);
			if (rootId && isAbsoluteUri(rootId)) {
				const hashIndex = rootId.indexOf("#");
				const rootBase = hashIndex === -1 ? rootId : rootId.slice(0, hashIndex);
				try {
					uri = new URL(uri, rootBase).toString();
				} catch {}
			}
		}
		const remotePath = getSafeRemotePath(uri);
		const queryPath = getQueryPath(uri);
		let result;
		let resolvedFromAncestor = false;
		if (remotePath) {
			const ancestorReport = report.getAncestor(remotePath);
			if (ancestorReport?.rootSchema) {
				result = ancestorReport.rootSchema;
				resolvedFromAncestor = true;
			}
		}
		if (!result && root && remotePath) {
			const rootId = getEffectiveId(root);
			const rootRemotePath = rootId ? getRemotePath(rootId) : void 0;
			if (rootRemotePath && rootRemotePath === remotePath) result = root;
		}
		if (!result) result = remotePath ? this.fromCache(remotePath) : root;
		if (result && remotePath && isAbsoluteUri(remotePath) && (!result.id || !isAbsoluteUri(result.id))) result.id = remotePath;
		if (result && remotePath) {
			if (result !== root && !resolvedFromAncestor) {
				report.path.push(remotePath);
				let remoteReport;
				let usesAncestorReport = false;
				const anscestorReport = result.id ? report.getAncestor(result.id) : void 0;
				if (anscestorReport) {
					remoteReport = anscestorReport;
					usesAncestorReport = true;
				} else {
					remoteReport = new Report(report);
					const noCache = result.id && isAbsoluteUri(result.id) ? false : true;
					if (this.validator.sc.compileSchema(remoteReport, result, { noCache })) {
						const savedOptions = this.validator.options;
						try {
							this.validator.options = result.__$validationOptions || this.validator.options;
							const parentSchemaUri = typeof result.$schema === "string" ? getRemotePath(result.$schema) : void 0;
							const currentSchemaUri = report.getSchemaId();
							if (!(!!parentSchemaUri && parentSchemaUri.length > 0 && (currentSchemaUri === parentSchemaUri || !!report.getAncestor(parentSchemaUri)))) this.validator.sv.validateSchema(remoteReport, result);
						} finally {
							this.validator.options = savedOptions;
						}
					}
				}
				const remoteReportIsValid = usesAncestorReport ? true : remoteReport.isValid();
				if (!remoteReportIsValid) report.addError("REMOTE_NOT_VALID", [uri], remoteReport);
				report.path.pop();
				if (!remoteReportIsValid) return;
			}
		}
		const resourceRoot = result;
		if (result && queryPath) {
			const parts = queryPath.split("/");
			for (let idx = 0, lim = parts.length; result && idx < lim; idx++) {
				const key = decodeJSONPointer(parts[idx]);
				if (idx === 0) result = findId(result, key, remotePath, remotePath, this.validator.options.maxRecursionDepth);
				else result = result[key];
			}
		}
		if (result && typeof result === "object" && resourceRoot && typeof resourceRoot === "object") result.__$resourceRoot = resourceRoot;
		return result;
	}
};
//#endregion
export { SchemaCache, prepareRemoteSchema };
