// z-schema/dist/validation/array.mjs
import { isUniqueArray } from "../utils/array.mjs";
import { cacheValidationResult, deferOrRunSync, shouldSkipValidate } from "./shared.mjs";
//#region src/validation/array.ts
function additionalItemsValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["ARRAY_ADDITIONAL_ITEMS"])) return;
	if (!Array.isArray(json)) return;
	if (schema.additionalItems === false && Array.isArray(schema.items)) {
		if (json.length > schema.items.length) report.addError("ARRAY_ADDITIONAL_ITEMS", void 0, void 0, schema, "additionalItems");
	}
}
function itemsValidator() {}
function prefixItemsValidator() {}
function maxItemsValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["ARRAY_LENGTH_LONG"])) return;
	if (!Array.isArray(json)) return;
	if (json.length > schema.maxItems) report.addError("ARRAY_LENGTH_LONG", [json.length, schema.maxItems], void 0, schema, "maxItems");
}
function minItemsValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["ARRAY_LENGTH_SHORT"])) return;
	if (!Array.isArray(json)) return;
	if (json.length < schema.minItems) report.addError("ARRAY_LENGTH_SHORT", [json.length, schema.minItems], void 0, schema, "minItems");
}
function uniqueItemsValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["ARRAY_UNIQUE"])) return;
	if (!Array.isArray(json)) return;
	if (schema.uniqueItems === true) {
		const matches = [];
		if (isUniqueArray(json, matches, this.options.maxRecursionDepth) === false) report.addError("ARRAY_UNIQUE", matches, void 0, schema, "uniqueItems");
	}
}
function containsValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["CONTAINS"])) return;
	if (!Array.isArray(json)) return;
	const containsSchema = schema.contains;
	if (containsSchema === void 0) return;
	const Report_ = report.constructor;
	const subReports = [];
	for (let idx = 0; idx < json.length; idx++) {
		const subReport = new Report_(report);
		subReports.push(subReport);
		this._jsonValidate(subReport, containsSchema, json[idx]);
		cacheValidationResult(report, containsSchema, json[idx], subReport.errors.length === 0);
	}
	const addContainsErrorIfNeeded = () => {
		let matchingItems = 0;
		for (const subReport of subReports) if (subReport.errors.length === 0) matchingItems += 1;
		const supportsContainsBounds = this.options.version === "draft2019-09" || this.options.version === "draft2020-12";
		const minContains = supportsContainsBounds && typeof schema.minContains === "number" ? schema.minContains ?? 1 : 1;
		const maxContains = supportsContainsBounds && typeof schema.maxContains === "number" ? schema.maxContains : void 0;
		if (!(matchingItems >= minContains) || !(maxContains === void 0 || matchingItems <= maxContains)) report.addError("CONTAINS", void 0, subReports, schema, void 0);
	};
	deferOrRunSync(report, subReports, addContainsErrorIfNeeded);
}
function maxContainsValidator() {}
function minContainsValidator() {}
//#endregion
export { additionalItemsValidator, containsValidator, itemsValidator, maxContainsValidator, maxItemsValidator, minContainsValidator, minItemsValidator, prefixItemsValidator, uniqueItemsValidator };
