// z-schema/dist/validation/combinators.mjs
import { Report } from "../report.mjs";
import { cacheValidationResult, deferOrRunSync } from "./shared.mjs";
//#region src/validation/combinators.ts
function allOfValidator(report, schema, json) {
	for (let i = 0; i < schema.allOf.length; i++) {
		const validateResult = this._jsonValidate(report, schema.allOf[i], json);
		if (this.options.breakOnFirstError && validateResult === false) break;
	}
}
function anyOfValidator(report, schema, json) {
	const subReports = [];
	for (let i = 0; i < schema.anyOf.length; i++) {
		const subReport = new Report(report);
		subReports.push(subReport);
		this._jsonValidate(subReport, schema.anyOf[i], json);
		cacheValidationResult(report, schema.anyOf[i], json, subReport.errors.length === 0);
	}
	deferOrRunSync(report, subReports, () => {
		let passed = false;
		for (const subReport of subReports) if (subReport.errors.length === 0) {
			passed = true;
			break;
		}
		if (passed === false) report.addError("ANY_OF_MISSING", void 0, subReports, schema, "anyOf");
	});
}
function oneOfValidator(report, schema, json) {
	const subReports = [];
	for (let i = 0; i < schema.oneOf.length; i++) {
		const subReport = new Report(report);
		subReports.push(subReport);
		this._jsonValidate(subReport, schema.oneOf[i], json);
		cacheValidationResult(report, schema.oneOf[i], json, subReport.errors.length === 0);
	}
	deferOrRunSync(report, subReports, () => {
		let passes = 0;
		for (const subReport of subReports) if (subReport.errors.length === 0) passes++;
		if (passes === 0) report.addError("ONE_OF_MISSING", void 0, subReports, schema, "oneOf");
		else if (passes > 1) report.addError("ONE_OF_MULTIPLE", void 0, void 0, schema, "oneOf");
	});
}
function notValidator(report, schema, json) {
	const subReport = new Report(report);
	if (this._jsonValidate(subReport, schema.not, json) === true) report.addError("NOT_PASSED", void 0, void 0, schema, "not");
}
function ifValidator(report, schema, json) {
	if (this.options.version === "draft-04" || this.options.version === "draft-06") return;
	const conditionSchema = schema.if;
	const thenSchema = schema.then;
	const elseSchema = schema.else;
	if (conditionSchema === void 0 || thenSchema === void 0 && elseSchema === void 0) return;
	const conditionReport = new Report(report);
	this._jsonValidate(conditionReport, conditionSchema, json);
	cacheValidationResult(report, conditionSchema, json, conditionReport.errors.length === 0);
	const branchSchema = conditionReport.errors.length === 0 ? thenSchema : elseSchema;
	if (branchSchema === void 0) return;
	this._jsonValidate(report, branchSchema, json);
}
function thenValidator() {}
function elseValidator() {}
//#endregion
export { allOfValidator, anyOfValidator, elseValidator, ifValidator, notValidator, oneOfValidator, thenValidator };
