// z-schema/dist/validation/object.mjs
import { isObject } from "../utils/what-is.mjs";
import { compileSchemaRegex } from "../utils/schema-regex.mjs";
import { difference } from "../utils/array.mjs";
import { deferOrRunSync, shouldSkipValidate, supportsDependentKeywords } from "./shared.mjs";
//#region src/validation/object.ts
function maxPropertiesValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["OBJECT_PROPERTIES_MAXIMUM"])) return;
	if (!isObject(json)) return;
	const keysCount = Object.keys(json).length;
	if (keysCount > schema.maxProperties) report.addError("OBJECT_PROPERTIES_MAXIMUM", [keysCount, schema.maxProperties], void 0, schema, "maxProperties");
}
function minPropertiesValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["OBJECT_PROPERTIES_MINIMUM"])) return;
	if (!isObject(json)) return;
	const keysCount = Object.keys(json).length;
	if (keysCount < schema.minProperties) report.addError("OBJECT_PROPERTIES_MINIMUM", [keysCount, schema.minProperties], void 0, schema, "minProperties");
}
function requiredValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["OBJECT_MISSING_REQUIRED_PROPERTY"])) return;
	if (!isObject(json)) return;
	const idx = schema.required.length;
	for (let i = 0; i < idx; i++) {
		const requiredPropertyName = schema.required[i];
		if (!Object.hasOwn(json, requiredPropertyName)) report.addError("OBJECT_MISSING_REQUIRED_PROPERTY", [requiredPropertyName], void 0, schema, "required");
	}
}
function additionalPropertiesValidator(report, schema, json) {
	if (schema.properties === void 0 && schema.patternProperties === void 0) return propertiesValidator.call(this, report, schema, json);
}
function patternPropertiesValidator(report, schema, json) {
	if (schema.properties === void 0) return propertiesValidator.call(this, report, schema, json);
}
function propertiesValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["OBJECT_ADDITIONAL_PROPERTIES"])) return;
	if (!isObject(json)) return;
	const properties = schema.properties !== void 0 ? schema.properties : {};
	const patternProperties = schema.patternProperties !== void 0 ? schema.patternProperties : {};
	if (schema.additionalProperties === false) {
		let s = Object.keys(json);
		const p = Object.keys(properties);
		const pp = Object.keys(patternProperties);
		s = difference(s, p);
		for (const ppKey of pp) {
			const result = compileSchemaRegex(ppKey);
			if (!result.ok) continue;
			const regExp = result.value;
			for (let idx2 = s.length - 1; idx2 >= 0; idx2--) if (regExp.test(s[idx2]) === true) s.splice(idx2, 1);
		}
		if (s.length > 0) {
			if (Array.isArray(this.options.assumeAdditional)) for (const allowed of this.options.assumeAdditional) {
				const io = s.indexOf(allowed);
				if (io !== -1) s.splice(io, 1);
			}
			if (s.length > 0) for (const extra of s) report.addError("OBJECT_ADDITIONAL_PROPERTIES", [extra], void 0, schema, "properties");
		}
	}
}
function dependenciesValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["OBJECT_DEPENDENCY_KEY"])) return;
	if (!isObject(json)) return;
	const keys = Object.keys(schema.dependencies);
	for (const dependencyName of keys) if (Object.hasOwn(json, dependencyName)) {
		const dependencyDefinition = schema.dependencies[dependencyName];
		if (Array.isArray(dependencyDefinition)) {
			for (const requiredPropertyName of dependencyDefinition) if (!Object.hasOwn(json, requiredPropertyName)) report.addError("OBJECT_DEPENDENCY_KEY", [requiredPropertyName, dependencyName], void 0, schema, "dependencies");
		} else this._jsonValidate(report, dependencyDefinition, json);
	}
}
function dependentSchemasValidator(report, schema, json) {
	if (!supportsDependentKeywords(schema, this.options.version)) return;
	if (!isObject(json) || !isObject(schema.dependentSchemas)) return;
	const keys = Object.keys(schema.dependentSchemas);
	for (const dependencyName of keys) if (Object.hasOwn(json, dependencyName)) {
		const dependencySchema = schema.dependentSchemas[dependencyName];
		this._jsonValidate(report, dependencySchema, json);
	}
}
function dependentRequiredValidator(report, schema, json) {
	if (!supportsDependentKeywords(schema, this.options.version)) return;
	if (shouldSkipValidate(this.validateOptions, ["OBJECT_DEPENDENCY_KEY"])) return;
	if (!isObject(json) || !isObject(schema.dependentRequired)) return;
	const keys = Object.keys(schema.dependentRequired);
	for (const dependencyName of keys) {
		if (!Object.hasOwn(json, dependencyName)) continue;
		const requiredProperties = schema.dependentRequired[dependencyName];
		if (!Array.isArray(requiredProperties)) continue;
		for (const requiredPropertyName of requiredProperties) if (!Object.hasOwn(json, requiredPropertyName)) report.addError("OBJECT_DEPENDENCY_KEY", [requiredPropertyName, dependencyName], void 0, schema, "dependentRequired");
	}
}
function propertyNamesValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["PROPERTY_NAMES"])) return;
	if (!isObject(json)) return;
	const propertyNamesSchema = schema.propertyNames;
	if (propertyNamesSchema === void 0) return;
	const Report_ = report.constructor;
	const keys = Object.keys(json);
	const subReports = [];
	for (const key of keys) {
		const subReport = new Report_(report);
		subReports.push(subReport);
		this._jsonValidate(subReport, propertyNamesSchema, key);
	}
	const addPropertyNameErrors = () => {
		for (let idx = 0; idx < keys.length; idx++) if (subReports[idx].errors.length > 0) report.addError("PROPERTY_NAMES", [keys[idx]], subReports[idx], schema, void 0);
	};
	deferOrRunSync(report, subReports, addPropertyNameErrors);
}
//#endregion
export { additionalPropertiesValidator, dependenciesValidator, dependentRequiredValidator, dependentSchemasValidator, maxPropertiesValidator, minPropertiesValidator, patternPropertiesValidator, propertiesValidator, propertyNamesValidator, requiredValidator };
