// z-schema/dist/validation/shared.mjs
import { isObject } from "../utils/what-is.mjs";
//#region src/validation/shared.ts
const shouldSkipValidate = function(options, errors) {
	return options && Array.isArray(options.includeErrors) && options.includeErrors.length > 0 && !errors.some(function(err) {
		return options.includeErrors.includes(err);
	});
};
const supportsDependentKeywords = (schema, version) => {
	if (typeof schema.$schema === "string") return !/draft-04|draft-06|draft-07/.test(schema.$schema);
	return !(version === "draft-04" || version === "draft-06" || version === "draft-07");
};
const VOCAB_VALIDATION_2019_09 = "https://json-schema.org/draft/2019-09/vocab/validation";
const VOCAB_VALIDATION_2020_12 = "https://json-schema.org/draft/2020-12/vocab/validation";
const VOCAB_FORMAT_2019_09 = "https://json-schema.org/draft/2019-09/vocab/format";
const VOCAB_FORMAT_ASSERTION_2020_12 = "https://json-schema.org/draft/2020-12/vocab/format-assertion";
const VALIDATION_VOCAB_KEYWORDS = new Set([
	"type",
	"multipleOf",
	"maximum",
	"exclusiveMaximum",
	"minimum",
	"exclusiveMinimum",
	"maxLength",
	"minLength",
	"pattern",
	"maxItems",
	"minItems",
	"uniqueItems",
	"maxContains",
	"minContains",
	"maxProperties",
	"minProperties",
	"required",
	"dependentRequired",
	"enum",
	"const",
	"contentEncoding",
	"contentMediaType"
]);
const isValidationVocabularyEnabled = (schema, report, version) => {
	if (version !== "draft2019-09" && version !== "draft2020-12") return true;
	const currentSchemaMeta = schema.__$schemaResolved;
	const rootSchemaMeta = report.rootSchema && typeof report.rootSchema !== "boolean" ? report.rootSchema.__$schemaResolved : void 0;
	const metaSchema = currentSchemaMeta || rootSchemaMeta;
	if (!metaSchema || typeof metaSchema !== "object" || !isObject(metaSchema.$vocabulary)) return true;
	const vocabulary = metaSchema.$vocabulary;
	if (Object.hasOwn(vocabulary, VOCAB_VALIDATION_2019_09) || Object.hasOwn(vocabulary, VOCAB_VALIDATION_2020_12)) return vocabulary[VOCAB_VALIDATION_2019_09] === true || vocabulary[VOCAB_VALIDATION_2020_12] === true;
	return false;
};
/**
* Checks whether the format-assertion vocabulary is enabled in the meta-schema.
* For draft 2019-09: checks if the format vocabulary is set to true.
* For draft 2020-12: checks if the format-assertion vocabulary is present and true.
* Returns true for older drafts (format was always an assertion).
*/
const isFormatAssertionVocabEnabled = (schema, report, version) => {
	if (version !== "draft2019-09" && version !== "draft2020-12") return true;
	const currentSchemaMeta = schema.__$schemaResolved;
	const rootSchemaMeta = report.rootSchema && typeof report.rootSchema !== "boolean" ? report.rootSchema.__$schemaResolved : void 0;
	const metaSchema = currentSchemaMeta || rootSchemaMeta;
	if (!metaSchema || typeof metaSchema !== "object" || !isObject(metaSchema.$vocabulary)) return false;
	const vocabulary = metaSchema.$vocabulary;
	if (Object.hasOwn(vocabulary, VOCAB_FORMAT_ASSERTION_2020_12)) return vocabulary[VOCAB_FORMAT_ASSERTION_2020_12] === true;
	if (Object.hasOwn(vocabulary, VOCAB_FORMAT_2019_09)) return vocabulary[VOCAB_FORMAT_2019_09] === true;
	return false;
};
function cacheValidationResult(report, schema, json, passed) {
	let schemaMap = report.__validationResultCache.get(schema);
	if (!schemaMap) {
		schemaMap = /* @__PURE__ */ new Map();
		report.__validationResultCache.set(schema, schemaMap);
	}
	schemaMap.set(json, passed);
}
function getCachedValidationResult(report, schema, json) {
	return report.__validationResultCache.get(schema)?.get(json);
}
/**
* Shared async-task aggregation pattern.
*
* 1. Collects async tasks from `subReports` into `report`.
* 2. If new async tasks were added, defers `decisionFn` via `report.addAsyncTask`
*    with proper path save/restore so error paths are correct.
* 3. Otherwise, runs `decisionFn` synchronously.
*/
function deferOrRunSync(report, subReports, decisionFn) {
	const asyncTasksBefore = report.asyncTasks.length;
	for (const subReport of subReports) report.asyncTasks.push(...subReport.asyncTasks);
	if (report.asyncTasks.length > asyncTasksBefore) report.addAsyncTaskWithPath((callback) => {
		setTimeout(() => callback(null), 0);
	}, [], () => {
		decisionFn();
	});
	else decisionFn();
}
//#endregion
export { VALIDATION_VOCAB_KEYWORDS, cacheValidationResult, deferOrRunSync, getCachedValidationResult, isFormatAssertionVocabEnabled, isValidationVocabularyEnabled, shouldSkipValidate, supportsDependentKeywords };
