// z-schema/dist/validation/string.mjs
import { whatIs } from "../utils/what-is.mjs";
import { getFormatValidators } from "../format-validators.mjs";
import { compileSchemaRegex } from "../utils/schema-regex.mjs";
import { isFormatAssertionVocabEnabled, shouldSkipValidate } from "./shared.mjs";
import { decodeBase64, isValidBase64 } from "../utils/base64.mjs";
import { unicodeLength } from "../utils/unicode.mjs";
//#region src/validation/string.ts
function minLengthValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["MIN_LENGTH"])) return;
	if (typeof json !== "string") return;
	if (unicodeLength(json) < schema.minLength) report.addError("MIN_LENGTH", [json.length, schema.minLength], void 0, schema, "minLength");
}
function maxLengthValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["MAX_LENGTH"])) return;
	if (typeof json !== "string") return;
	if (unicodeLength(json) > schema.maxLength) report.addError("MAX_LENGTH", [json.length, schema.maxLength], void 0, schema, "maxLength");
}
function patternValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["PATTERN"])) return;
	if (typeof json !== "string") return;
	const result = compileSchemaRegex(schema.pattern);
	if (!result.ok) {
		report.addError("PATTERN", [
			schema.pattern,
			json,
			result.error.message
		], void 0, schema, "pattern");
		return;
	}
	if (!result.value.test(json)) report.addError("PATTERN", [schema.pattern, json], void 0, schema, "pattern");
}
function formatValidator(report, schema, json) {
	if (this.options.formatAssertions === false) return;
	if (this.options.formatAssertions === true) {
		if (!isFormatAssertionVocabEnabled(schema, report, this.options.version)) return;
	}
	const isModernDraft = this.options.version === "draft2019-09" || this.options.version === "draft2020-12";
	const formatValidatorFn = getFormatValidators(this.options)[schema.format];
	if (typeof formatValidatorFn === "function") {
		if (shouldSkipValidate(this.validateOptions, ["INVALID_FORMAT"])) return;
		if (report.hasError("INVALID_TYPE", [schema.type, whatIs(json)])) return;
		if (formatValidatorFn.length === 2) report.addAsyncTaskWithPath(formatValidatorFn, [json], function(result) {
			if (result !== true) report.addError("INVALID_FORMAT", [schema.format, JSON.stringify(json)], void 0, schema, "format");
		});
		else {
			const result = formatValidatorFn.call(this, json);
			if (result instanceof Promise) {
				const promiseResult = result;
				report.addAsyncTaskWithPath(async (callback) => {
					try {
						callback(await promiseResult);
					} catch (_error) {
						callback(false);
					}
				}, [], function(resolvedResult) {
					if (resolvedResult !== true) report.addError("INVALID_FORMAT", [schema.format, JSON.stringify(json)], void 0, schema, "format");
				});
			} else if (result !== true) report.addError("INVALID_FORMAT", [schema.format, JSON.stringify(json)], void 0, schema, "format");
		}
	} else if (this.options.ignoreUnknownFormats !== true && !isModernDraft) report.addError("UNKNOWN_FORMAT", [schema.format], void 0, schema, "format");
}
function contentEncodingValidator(report, schema, json) {
	if (this.options.version !== "draft-07") return;
	if (typeof json !== "string") return;
	if (schema.contentEncoding !== "base64") return;
	if (!isValidBase64(json)) report.addError("INVALID_FORMAT", ["contentEncoding:base64", JSON.stringify(json)], void 0, schema, "contentEncoding");
}
function contentMediaTypeValidator(report, schema, json) {
	if (this.options.version !== "draft-07") return;
	if (typeof json !== "string") return;
	if (schema.contentMediaType !== "application/json") return;
	let payload = json;
	if (schema.contentEncoding === "base64") {
		const decoded = decodeBase64(json);
		if (decoded === void 0) {
			report.addError("INVALID_FORMAT", ["contentEncoding:base64", JSON.stringify(json)], void 0, schema, "contentEncoding");
			return;
		}
		payload = decoded;
	}
	try {
		JSON.parse(payload);
	} catch {
		report.addError("INVALID_FORMAT", ["contentMediaType:application/json", JSON.stringify(json)], void 0, schema, "contentMediaType");
	}
}
//#endregion
export { contentEncodingValidator, contentMediaTypeValidator, formatValidator, maxLengthValidator, minLengthValidator, patternValidator };
