// z-schema/cjs/index.js
Object.defineProperties(exports, {
	__esModule: { value: true },
	[Symbol.toStringTag]: { value: "Module" }
});
//#region \0rolldown/runtime.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));
//#endregion
let validator_lib_isEmail_js = require("validator/lib/isEmail.js");
validator_lib_isEmail_js = __toESM(validator_lib_isEmail_js);
let validator_lib_isIP_js = require("validator/lib/isIP.js");
validator_lib_isIP_js = __toESM(validator_lib_isIP_js);
let validator_lib_isURL_js = require("validator/lib/isURL.js");
validator_lib_isURL_js = __toESM(validator_lib_isURL_js);
let punycode_punycode_js = require("punycode/punycode.js");
punycode_punycode_js = __toESM(punycode_punycode_js);
let safe_regex2 = require("safe-regex2");
safe_regex2 = __toESM(safe_regex2);
//#region src/utils/uri.ts
const getQueryPath = (uri) => {
	const io = uri.indexOf("#");
	return io === -1 ? void 0 : uri.slice(io + 1);
};
const getRemotePath = (uri) => {
	const io = uri.indexOf("#");
	return io === -1 ? uri : uri.slice(0, io);
};
const isAbsoluteUri = (uri) => /^[a-zA-Z][a-zA-Z\d+.-]*:/.test(uri);
//#endregion
//#region src/utils/constants.ts
/**
* Maximum allowed value for {@link ZSchemaOptions.asyncTimeout} in milliseconds.
* Values exceeding this limit are clamped during option normalization to
* prevent resource exhaustion (CWE-400).
*/
const MAX_ASYNC_TIMEOUT = 6e4;
/**
* Maximum allowed length for a JSON Schema `pattern` regular expression string.
* Patterns exceeding this limit are rejected by {@link compileSchemaRegex} to
* mitigate Regular Expression Denial-of-Service (CWE-1333) and regex injection
* (CWE-95).
*/
const MAX_SCHEMA_REGEX_LENGTH = 1e4;
//#endregion
//#region src/utils/what-is.ts
const whatIs = (what) => {
	if (typeof what === "object") {
		if (what === null) return "null";
		if (Array.isArray(what)) return "array";
		return "object";
	}
	if (typeof what === "number") {
		if (Number.isFinite(what)) if (what % 1 === 0) return "integer";
		else return "number";
		if (Number.isNaN(what)) return "not-a-number";
		return "unknown-number";
	}
	return typeof what;
};
function isObject(value) {
	return typeof value === "object" && value !== null && !Array.isArray(value);
}
function isInteger(value) {
	return typeof value === "number" && Number.isFinite(value) && value % 1 === 0;
}
//#endregion
//#region src/json-schema.ts
/**
* Keywords whose values are not JSON Schema sub-schemas and must not be
* traversed during schema walking (id collection, reference collection, etc.).
*/
const NON_SCHEMA_KEYWORDS = [
	"enum",
	"const",
	"default",
	"examples"
];
/** Returns true if the key is an internal z-schema property (prefixed with `__$`). */
const isInternalKey = (key) => key.startsWith("__$");
const getId = (schema) => {
	if (typeof schema.$schema === "string" && schema.$schema.includes("draft-04")) return schema.id;
	return schema.$id ?? schema.id;
};
const findId = (schema, id, targetBaseUri, currentBaseUri, maxDepth = 100, _depth = 0) => {
	if (typeof schema !== "object" || schema === null) return;
	if (!id) return schema;
	if (_depth >= maxDepth) throw new Error(`Maximum recursion depth (${maxDepth}) exceeded in findId. If your schema is deeply nested and valid, increase the maxRecursionDepth option.`);
	const baseUri = currentBaseUri ?? targetBaseUri;
	const schemaId = getId(schema);
	let nextBaseUri = baseUri;
	if (schemaId) {
		if (isAbsoluteUri(schemaId)) nextBaseUri = getRemotePath(schemaId);
		else if (baseUri && isAbsoluteUri(baseUri)) try {
			nextBaseUri = getRemotePath(new URL(schemaId, baseUri).toString());
		} catch {}
	}
	if (!targetBaseUri || nextBaseUri === targetBaseUri) {
		if (schemaId && (schemaId === id || schemaId[0] === "#" && schemaId.substring(1) === id)) return schema;
		if (schema.$anchor === id || schema.$dynamicAnchor === id) return schema;
	}
	let result;
	if (Array.isArray(schema)) for (let i = 0; i < schema.length; i++) {
		result = findId(schema[i], id, targetBaseUri, nextBaseUri, maxDepth, _depth + 1);
		if (result) return result;
	}
	if (isObject(schema)) {
		const keys = Object.keys(schema);
		for (let i = keys.length - 1; i >= 0; i--) {
			const k = keys[i];
			if (isInternalKey(k) || NON_SCHEMA_KEYWORDS.includes(k)) continue;
			result = findId(schema[k], id, targetBaseUri, nextBaseUri, maxDepth, _depth + 1);
			if (result) return result;
		}
	}
};
//#endregion
//#region src/errors.ts
const Errors = {
	INVALID_TYPE: "Expected type {0} but found type {1}",
	INVALID_FORMAT: "Object didn't pass validation for format {0}: {1}",
	ENUM_MISMATCH: "No enum match for: {0}",
	ENUM_CASE_MISMATCH: "Enum does not match case for: {0}",
	ANY_OF_MISSING: "Data does not match any schemas from 'anyOf'",
	ONE_OF_MISSING: "Data does not match any schemas from 'oneOf'",
	ONE_OF_MULTIPLE: "Data is valid against more than one schema from 'oneOf'",
	NOT_PASSED: "Data matches schema from 'not'",
	ARRAY_LENGTH_SHORT: "Array is too short ({0}), minimum {1}",
	ARRAY_LENGTH_LONG: "Array is too long ({0}), maximum {1}",
	ARRAY_UNIQUE: "Array items are not unique (indexes {0} and {1})",
	ARRAY_ADDITIONAL_ITEMS: "Additional items not allowed",
	ARRAY_UNEVALUATED_ITEMS: "Unevaluated items are not allowed",
	MULTIPLE_OF: "Value {0} is not a multiple of {1}",
	MINIMUM: "Value {0} is less than minimum {1}",
	MINIMUM_EXCLUSIVE: "Value {0} is equal or less than exclusive minimum {1}",
	MAXIMUM: "Value {0} is greater than maximum {1}",
	MAXIMUM_EXCLUSIVE: "Value {0} is equal or greater than exclusive maximum {1}",
	OBJECT_PROPERTIES_MINIMUM: "Too few properties defined ({0}), minimum {1}",
	OBJECT_PROPERTIES_MAXIMUM: "Too many properties defined ({0}), maximum {1}",
	OBJECT_MISSING_REQUIRED_PROPERTY: "Missing required property: {0}",
	OBJECT_ADDITIONAL_PROPERTIES: "Additional properties not allowed: {0}",
	OBJECT_UNEVALUATED_PROPERTIES: "Unevaluated properties are not allowed: {0}",
	OBJECT_DEPENDENCY_KEY: "Dependency failed - key must exist: {0} (due to key: {1})",
	MIN_LENGTH: "String is too short ({0} chars), minimum {1}",
	MAX_LENGTH: "String is too long ({0} chars), maximum {1}",
	PATTERN: "String does not match pattern {0}: {1}",
	KEYWORD_TYPE_EXPECTED: "Keyword '{0}' is expected to be of type '{1}'",
	KEYWORD_UNDEFINED_STRICT: "Keyword '{0}' must be defined in strict mode",
	KEYWORD_UNEXPECTED: "Keyword '{0}' is not expected to appear in the schema",
	KEYWORD_MUST_BE: "Keyword '{0}' must be {1}",
	KEYWORD_DEPENDENCY: "Keyword '{0}' requires keyword '{1}'",
	KEYWORD_PATTERN: "Keyword '{0}' is not a valid RegExp pattern: {1}",
	KEYWORD_VALUE_TYPE: "Each element of keyword '{0}' array must be a '{1}'",
	UNKNOWN_FORMAT: "There is no validation function for format '{0}'",
	CUSTOM_MODE_FORCE_PROPERTIES: "{0} must define at least one property if present",
	REF_UNRESOLVED: "Reference has not been resolved during compilation: {0}",
	UNRESOLVABLE_REFERENCE: "Reference could not be resolved: {0}",
	SCHEMA_NOT_REACHABLE: "Validator was not able to read schema with uri: {0}",
	SCHEMA_TYPE_EXPECTED: "Schema is expected to be of type 'object'",
	SCHEMA_NOT_AN_OBJECT: "Schema is not an object: {0}",
	ASYNC_TIMEOUT: "{0} asynchronous task(s) have timed out after {1} ms",
	PARENT_SCHEMA_VALIDATION_FAILED: "Schema failed to validate against its parent schema, see inner errors for details.",
	REMOTE_NOT_VALID: "Remote reference didn't compile successfully: {0}",
	SCHEMA_IS_FALSE: "Boolean schema \"false\" is always invalid.",
	CONST: "Value does not match const: {0}",
	CONTAINS: "Array does not contain an item matching the schema",
	PROPERTY_NAMES: "Property name {0} does not match the propertyNames schema",
	COLLECT_EVALUATED_DEPTH_EXCEEDED: "Schema nesting depth exceeded maximum ({0}) during unevaluated items/properties collection",
	MAX_RECURSION_DEPTH_EXCEEDED: "Maximum recursion depth ({0}) exceeded. If your schema or data is deeply nested and valid, increase the maxRecursionDepth option."
};
var ValidateError = class extends Error {
	name;
	details;
	constructor(message, details) {
		super(message);
		this.name = "z-schema validation error";
		this.details = details;
	}
};
function getValidateError({ message, details }) {
	return new ValidateError(message || "", details);
}
//#endregion
//#region src/utils/properties.ts
function copyProp(from, to, key, fn) {
	if (Object.hasOwn(from, key)) Object.defineProperty(to, key, {
		value: fn ? fn(from[key]) : from[key],
		enumerable: true,
		writable: true,
		configurable: true
	});
}
//#endregion
//#region src/utils/clone.ts
const shallowClone = (src) => {
	if (src == null || typeof src !== "object") return src;
	let res;
	if (Array.isArray(src)) {
		res = [];
		for (let i = 0; i < src.length; i++) res[i] = src[i];
	} else {
		res = {};
		const keys = Object.keys(src).sort();
		for (const key of keys) copyProp(src, res, key);
	}
	return res;
};
const deepClone = (src, maxDepth = 100) => {
	let vidx = 0;
	const visited = /* @__PURE__ */ new Map();
	const cloned = [];
	const cloneDeepInner = (src, _depth) => {
		if (typeof src !== "object" || src === null) return src;
		if (_depth >= maxDepth) throw new Error(`Maximum recursion depth (${maxDepth}) exceeded in deepClone. If your schema or data is deeply nested and valid, increase the maxRecursionDepth option.`);
		let res;
		const cidx = visited.get(src);
		if (cidx !== void 0) return cloned[cidx];
		visited.set(src, vidx++);
		if (Array.isArray(src)) {
			res = [];
			cloned.push(res);
			for (let i = 0; i < src.length; i++) res[i] = cloneDeepInner(src[i], _depth + 1);
		} else {
			res = {};
			cloned.push(res);
			const keys = Object.keys(src).sort();
			for (const key of keys) copyProp(src, res, key, (v) => cloneDeepInner(v, _depth + 1));
		}
		return res;
	};
	return cloneDeepInner(src, 0);
};
//#endregion
//#region src/utils/json.ts
const areEqual = (json1, json2, options, _depth = 0) => {
	const caseInsensitiveComparison = options?.caseInsensitiveComparison || false;
	const maxDepth = options?.maxDepth ?? 100;
	if (json1 === json2) return true;
	if (caseInsensitiveComparison === true && typeof json1 === "string" && typeof json2 === "string" && json1.toUpperCase() === json2.toUpperCase()) return true;
	if (_depth >= maxDepth) throw new Error(`Maximum recursion depth (${maxDepth}) exceeded in areEqual. If your data is deeply nested and valid, increase the maxRecursionDepth option.`);
	let i, len;
	if (Array.isArray(json1) && Array.isArray(json2)) {
		if (json1.length !== json2.length) return false;
		len = json1.length;
		for (i = 0; i < len; i++) if (!areEqual(json1[i], json2[i], options, _depth + 1)) return false;
		return true;
	}
	if (isObject(json1) && isObject(json2)) {
		const keys1 = sortedKeys(json1);
		if (!areEqual(keys1, sortedKeys(json2), options, _depth + 1)) return false;
		len = keys1.length;
		for (i = 0; i < len; i++) if (!areEqual(json1[keys1[i]], json2[keys1[i]], options, _depth + 1)) return false;
		return true;
	}
	return false;
};
const decodeJSONPointer = (str) => {
	return decodeURIComponent(str).replace(/~[0-1]/g, (x) => x === "~1" ? "/" : "~");
};
const sortedKeys = (obj) => Object.keys(obj).sort();
const get = (obj, path) => {
	if (typeof path === "string") path = path.split(".");
	return path.reduce((acc, key) => acc && acc[key] !== void 0 ? acc[key] : void 0, obj);
};
//#endregion
//#region src/utils/symbols.ts
const jsonSymbol = Symbol.for("z-schema/json");
const schemaSymbol = Symbol.for("z-schema/schema");
//#endregion
//#region src/report.ts
const ASYNC_TIMEOUT_POLL_MS = 10;
var Report = class Report {
	asyncTasks = [];
	commonErrorMessage;
	__$recursiveAnchorStack = [];
	__$dynamicScopeStack = [];
	__validationResultCache = /* @__PURE__ */ new Map();
	errors = [];
	json;
	path = [];
	schemaPath = [];
	rootSchema;
	parentReport;
	options;
	reportOptions;
	validateOptions = {};
	constructor(parentOrOptions, reportOptionsOrValidate, validateOptions) {
		this.parentReport = parentOrOptions instanceof Report ? parentOrOptions : void 0;
		this.options = parentOrOptions instanceof Report ? parentOrOptions.options : parentOrOptions || {};
		if (parentOrOptions instanceof Report) {
			this.reportOptions = reportOptionsOrValidate || {};
			this.validateOptions = validateOptions || parentOrOptions.validateOptions;
			this.__$recursiveAnchorStack = [...parentOrOptions.__$recursiveAnchorStack];
			this.__$dynamicScopeStack = [...parentOrOptions.__$dynamicScopeStack];
			this.__validationResultCache = parentOrOptions.__validationResultCache;
		} else {
			this.reportOptions = {};
			this.validateOptions = reportOptionsOrValidate || {};
			this.__$recursiveAnchorStack = [];
			this.__$dynamicScopeStack = [];
			this.__validationResultCache = /* @__PURE__ */ new Map();
		}
	}
	isValid() {
		if (this.asyncTasks.length > 0) throw new Error("Async tasks pending, can't answer isValid");
		return this.errors.length === 0;
	}
	addAsyncTask(fn, args, asyncTaskResultProcessFn) {
		this.asyncTasks.push([
			fn,
			args,
			asyncTaskResultProcessFn
		]);
	}
	/**
	* Like {@link addAsyncTask}, but automatically saves the current `path` and
	* restores it around `processFn`.  This eliminates the manual
	* path-save/restore boilerplate that every async-aware validator would
	* otherwise need.
	*/
	addAsyncTaskWithPath(fn, args, processFn) {
		const pathBefore = shallowClone(this.path);
		this.asyncTasks.push([
			fn,
			args,
			(result) => {
				const backup = this.path;
				this.path = pathBefore;
				processFn(result);
				this.path = backup;
			}
		]);
	}
	getAncestor(id) {
		if (!this.parentReport) return;
		if (this.parentReport.getSchemaId() === id) return this.parentReport;
		return this.parentReport.getAncestor(id);
	}
	processAsyncTasks(timeout, callback) {
		const validationTimeout = Math.min(Math.max(timeout || 2e3, 0), MAX_ASYNC_TIMEOUT);
		const timeoutAt = Date.now() + validationTimeout;
		let tasksCount = this.asyncTasks.length;
		let timedOut = false;
		const finish = () => {
			setTimeout(() => {
				const valid = this.errors.length === 0;
				callback(valid ? void 0 : getValidateError({ details: this.errors }), valid);
			}, 0);
		};
		const respond = (asyncTaskResultProcessFn) => (asyncTaskResult) => {
			if (timedOut) return;
			asyncTaskResultProcessFn(asyncTaskResult);
			if (--tasksCount === 0) finish();
		};
		if (tasksCount === 0 || this.errors.length > 0 && this.options.breakOnFirstError) {
			finish();
			return;
		}
		for (let i = 0; i < this.asyncTasks.length; i++) {
			const [fn, fnArgs, processFn] = this.asyncTasks[i];
			const respondCallback = respond(processFn);
			fn(...fnArgs, respondCallback);
		}
		const checkTimeout = () => {
			if (timedOut || tasksCount <= 0) return;
			if (Date.now() >= timeoutAt) {
				timedOut = true;
				this.addError("ASYNC_TIMEOUT", [tasksCount, validationTimeout]);
				callback(getValidateError({ details: this.errors }), false);
				return;
			}
			setTimeout(checkTimeout, ASYNC_TIMEOUT_POLL_MS);
		};
		setTimeout(checkTimeout, ASYNC_TIMEOUT_POLL_MS);
	}
	getPath(returnPathAsString) {
		let path = [];
		if (this.parentReport) path = path.concat(this.parentReport.path);
		path = path.concat(this.path);
		if (returnPathAsString !== true) return "#/" + path.map(function(segment) {
			segment = segment.toString();
			if (isAbsoluteUri(segment)) return "uri(" + segment + ")";
			return segment.replace(/~/g, "~0").replace(/\//g, "~1");
		}).join("/");
		return path;
	}
	getSchemaPath() {
		let schemaPath = [];
		if (this.parentReport) schemaPath = schemaPath.concat(this.parentReport.schemaPath);
		schemaPath = schemaPath.concat(this.schemaPath);
		return schemaPath;
	}
	getSchemaId() {
		if (!this.rootSchema) return;
		let path = [];
		if (this.parentReport) path = path.concat(this.parentReport.path);
		path = path.concat(this.path);
		while (path.length > 0) {
			const obj = get(this.rootSchema, path);
			if (obj && obj.id) return obj.id;
			path.pop();
		}
		return this.rootSchema.id;
	}
	hasError(errCode, errParams) {
		for (let idx = 0; idx < this.errors.length; idx++) if (this.errors[idx].code === errCode) {
			let match = true;
			for (let idx2 = 0; idx2 < this.errors[idx].params.length; idx2++) if (this.errors[idx].params[idx2] !== errParams[idx2]) match = false;
			if (match) return match;
		}
		return false;
	}
	addError(errCode, errParams, subReports, schema, keyword) {
		if (!errCode) throw new Error("No errorCode passed into addError()");
		this.addCustomError(errCode, Errors[errCode], errParams, subReports, schema, keyword);
	}
	getJson() {
		if (this.json) return this.json;
		if (this.parentReport) return this.parentReport.getJson();
	}
	addCustomError(errorCode, errorMessage, params, subReports, schema, keyword) {
		if (typeof this.reportOptions.maxErrors === "number" && this.errors.length >= this.reportOptions.maxErrors) return;
		if (!errorMessage) throw new Error("No errorMessage known for code " + errorCode);
		params = params || [];
		for (let idx = 0; idx < params.length; idx++) {
			const param = params[idx] === null || isObject(params[idx]) ? JSON.stringify(params[idx]) : params[idx];
			errorMessage = errorMessage.replace("{" + idx + "}", param.toString());
		}
		const err = {
			code: errorCode,
			params,
			message: errorMessage,
			path: this.getPath(this.options.reportPathAsArray),
			schemaPath: this.getSchemaPath(),
			schemaId: this.getSchemaId(),
			keyword
		};
		err[schemaSymbol] = schema;
		err[jsonSymbol] = this.getJson();
		if (schema && typeof schema === "string") err.description = schema;
		else if (schema && typeof schema === "object") {
			if (schema.title) err.title = schema.title;
			if (schema.description) err.description = schema.description;
		}
		if (subReports != null) {
			if (!Array.isArray(subReports)) subReports = [subReports];
			err.inner = [];
			for (const subReport of subReports) for (const error of subReport.errors) err.inner.push(error);
			if (err.inner.length === 0) err.inner = void 0;
		}
		if (Array.isArray(this.validateOptions.excludeErrors) && this.validateOptions.excludeErrors.includes(errorCode)) return;
		this.errors.push(err);
	}
};
//#endregion
//#region src/json-schema-versions.ts
const CURRENT_DEFAULT_SCHEMA_VERSION = "draft2020-12";
const VERSION_SCHEMA_URL_MAPPING = {
	"draft-04": "http://json-schema.org/draft-04/schema#",
	"draft-06": "http://json-schema.org/draft-06/schema#",
	"draft-07": "http://json-schema.org/draft-07/schema#",
	"draft2019-09": "https://json-schema.org/draft/2019-09/schema",
	"draft2020-12": "https://json-schema.org/draft/2020-12/schema"
};
//#endregion
//#region src/z-schema-options.ts
const defaultOptions = {
	version: CURRENT_DEFAULT_SCHEMA_VERSION,
	asyncTimeout: 2e3,
	forceAdditional: false,
	assumeAdditional: false,
	enumCaseInsensitiveComparison: false,
	forceItems: false,
	forceMinItems: false,
	forceMaxItems: false,
	forceMinLength: false,
	forceMaxLength: false,
	forceProperties: false,
	ignoreUnresolvableReferences: false,
	noExtraKeywords: false,
	noTypeless: false,
	noEmptyStrings: false,
	noEmptyArrays: false,
	strictUris: false,
	strictMode: false,
	reportPathAsArray: false,
	breakOnFirstError: false,
	pedanticCheck: false,
	ignoreUnknownFormats: false,
	formatAssertions: null,
	customValidator: null,
	maxRecursionDepth: 100
};
const normalizeOptions = (options) => {
	let normalized;
	if (typeof options === "object") {
		let keys = Object.keys(options);
		for (const key of keys) if (defaultOptions[key] === void 0) throw new Error("Unexpected option passed to constructor: " + key);
		keys = Object.keys(defaultOptions);
		for (const key of keys) if (options[key] === void 0) options[key] = shallowClone(defaultOptions[key]);
		normalized = options;
	} else normalized = shallowClone(defaultOptions);
	if (normalized.asyncTimeout != null) normalized.asyncTimeout = Math.min(Math.max(normalized.asyncTimeout, 0), MAX_ASYNC_TIMEOUT);
	if (normalized.strictMode === true) {
		normalized.forceAdditional = true;
		normalized.forceItems = true;
		normalized.forceMaxLength = true;
		normalized.forceProperties = true;
		normalized.noExtraKeywords = true;
		normalized.noTypeless = true;
		normalized.noEmptyStrings = true;
		normalized.noEmptyArrays = true;
	}
	return normalized;
};
//#endregion
//#region src/schema-cache.ts
function getSafeRemotePath(uri) {
	const remotePath = getRemotePath(uri);
	if (!remotePath) return;
	if (remotePath === "__proto__" || remotePath === "constructor" || remotePath === "prototype") return;
	return remotePath;
}
const getEffectiveId = (schema) => {
	let id = getId(schema);
	if ((!id || !isAbsoluteUri(id)) && typeof schema.id === "string" && isAbsoluteUri(schema.id)) id = schema.id;
	return id;
};
/**
* Shared logic for registering a remote reference schema.
* Used by both the static `ZSchema.setRemoteReference()` (global cache) and
* the instance `validator.setRemoteReference()` (instance cache).
*/
function prepareRemoteSchema(schema, uri, validationOptions, maxCloneDepth) {
	let _schema;
	if (typeof schema === "string") _schema = JSON.parse(schema);
	else _schema = deepClone(schema, maxCloneDepth);
	if (!_schema.id) _schema.id = uri;
	if (validationOptions) _schema.__$validationOptions = normalizeOptions(validationOptions);
	return _schema;
}
var SchemaCache = class SchemaCache {
	static global_cache = Object.create(null);
	cache = Object.create(null);
	constructor(validator) {
		this.validator = validator;
	}
	static cacheSchemaByUri(uri, schema) {
		const remotePath = getSafeRemotePath(uri);
		if (remotePath) this.global_cache[remotePath] = schema;
	}
	cacheSchemaByUri(uri, schema) {
		const remotePath = getSafeRemotePath(uri);
		if (remotePath) this.cache[remotePath] = schema;
	}
	removeFromCacheByUri(uri) {
		const remotePath = getSafeRemotePath(uri);
		if (remotePath) delete this.cache[remotePath];
	}
	checkCacheForUri(uri) {
		const remotePath = getSafeRemotePath(uri);
		return remotePath ? this.cache[remotePath] != null : false;
	}
	getSchema(report, refOrSchema) {
		if (Array.isArray(refOrSchema)) return refOrSchema.map((i) => this.getSchema(report, i));
		if (typeof refOrSchema === "string") return this.getSchemaByUri(report, refOrSchema);
		return deepClone(refOrSchema, this.validator.options.maxRecursionDepth);
	}
	fromCache(path) {
		if (path === "__proto__" || path === "constructor" || path === "prototype") return;
		let found = this.cache[path];
		if (found) return found;
		found = SchemaCache.global_cache[path];
		if (found) {
			const clone = deepClone(found, this.validator.options.maxRecursionDepth);
			if (!clone.id || !isAbsoluteUri(clone.id) && isAbsoluteUri(path)) clone.id = path;
			this.cache[path] = clone;
			return clone;
		}
	}
	getSchemaByUri(report, uri, root) {
		if (root && !isAbsoluteUri(uri)) {
			const rootId = getEffectiveId(root);
			if (rootId && isAbsoluteUri(rootId)) {
				const hashIndex = rootId.indexOf("#");
				const rootBase = hashIndex === -1 ? rootId : rootId.slice(0, hashIndex);
				try {
					uri = new URL(uri, rootBase).toString();
				} catch {}
			}
		}
		const remotePath = getSafeRemotePath(uri);
		const queryPath = getQueryPath(uri);
		let result;
		let resolvedFromAncestor = false;
		if (remotePath) {
			const ancestorReport = report.getAncestor(remotePath);
			if (ancestorReport?.rootSchema) {
				result = ancestorReport.rootSchema;
				resolvedFromAncestor = true;
			}
		}
		if (!result && root && remotePath) {
			const rootId = getEffectiveId(root);
			const rootRemotePath = rootId ? getRemotePath(rootId) : void 0;
			if (rootRemotePath && rootRemotePath === remotePath) result = root;
		}
		if (!result) result = remotePath ? this.fromCache(remotePath) : root;
		if (result && remotePath && isAbsoluteUri(remotePath) && (!result.id || !isAbsoluteUri(result.id))) result.id = remotePath;
		if (result && remotePath) {
			if (result !== root && !resolvedFromAncestor) {
				report.path.push(remotePath);
				let remoteReport;
				let usesAncestorReport = false;
				const anscestorReport = result.id ? report.getAncestor(result.id) : void 0;
				if (anscestorReport) {
					remoteReport = anscestorReport;
					usesAncestorReport = true;
				} else {
					remoteReport = new Report(report);
					const noCache = result.id && isAbsoluteUri(result.id) ? false : true;
					if (this.validator.sc.compileSchema(remoteReport, result, { noCache })) {
						const savedOptions = this.validator.options;
						try {
							this.validator.options = result.__$validationOptions || this.validator.options;
							const parentSchemaUri = typeof result.$schema === "string" ? getRemotePath(result.$schema) : void 0;
							const currentSchemaUri = report.getSchemaId();
							if (!(!!parentSchemaUri && parentSchemaUri.length > 0 && (currentSchemaUri === parentSchemaUri || !!report.getAncestor(parentSchemaUri)))) this.validator.sv.validateSchema(remoteReport, result);
						} finally {
							this.validator.options = savedOptions;
						}
					}
				}
				const remoteReportIsValid = usesAncestorReport ? true : remoteReport.isValid();
				if (!remoteReportIsValid) report.addError("REMOTE_NOT_VALID", [uri], remoteReport);
				report.path.pop();
				if (!remoteReportIsValid) return;
			}
		}
		const resourceRoot = result;
		if (result && queryPath) {
			const parts = queryPath.split("/");
			for (let idx = 0, lim = parts.length; result && idx < lim; idx++) {
				const key = decodeJSONPointer(parts[idx]);
				if (idx === 0) result = findId(result, key, remotePath, remotePath, this.validator.options.maxRecursionDepth);
				else result = result[key];
			}
		}
		if (result && typeof result === "object" && resourceRoot && typeof resourceRoot === "object") result.__$resourceRoot = resourceRoot;
		return result;
	}
};
//#endregion
//#region src/schemas/draft-04-schema.json
var draft_04_schema_default = {
	id: "http://json-schema.org/draft-04/schema#",
	$schema: "http://json-schema.org/draft-04/schema#",
	description: "Core schema meta-schema",
	definitions: {
		"schemaArray": {
			"type": "array",
			"minItems": 1,
			"items": { "$ref": "#" }
		},
		"positiveInteger": {
			"type": "integer",
			"minimum": 0
		},
		"positiveIntegerDefault0": { "allOf": [{ "$ref": "#/definitions/positiveInteger" }, { "default": 0 }] },
		"simpleTypes": { "enum": [
			"array",
			"boolean",
			"integer",
			"null",
			"number",
			"object",
			"string"
		] },
		"stringArray": {
			"type": "array",
			"items": { "type": "string" },
			"minItems": 1,
			"uniqueItems": true
		}
	},
	type: "object",
	properties: {
		"id": { "type": "string" },
		"$schema": { "type": "string" },
		"title": { "type": "string" },
		"description": { "type": "string" },
		"default": {},
		"multipleOf": {
			"type": "number",
			"minimum": 0,
			"exclusiveMinimum": true
		},
		"maximum": { "type": "number" },
		"exclusiveMaximum": {
			"type": "boolean",
			"default": false
		},
		"minimum": { "type": "number" },
		"exclusiveMinimum": {
			"type": "boolean",
			"default": false
		},
		"maxLength": { "$ref": "#/definitions/positiveInteger" },
		"minLength": { "$ref": "#/definitions/positiveIntegerDefault0" },
		"pattern": {
			"type": "string",
			"format": "regex"
		},
		"additionalItems": {
			"anyOf": [{ "type": "boolean" }, { "$ref": "#" }],
			"default": {}
		},
		"items": {
			"anyOf": [{ "$ref": "#" }, { "$ref": "#/definitions/schemaArray" }],
			"default": {}
		},
		"maxItems": { "$ref": "#/definitions/positiveInteger" },
		"minItems": { "$ref": "#/definitions/positiveIntegerDefault0" },
		"uniqueItems": {
			"type": "boolean",
			"default": false
		},
		"maxProperties": { "$ref": "#/definitions/positiveInteger" },
		"minProperties": { "$ref": "#/definitions/positiveIntegerDefault0" },
		"required": { "$ref": "#/definitions/stringArray" },
		"additionalProperties": {
			"anyOf": [{ "type": "boolean" }, { "$ref": "#" }],
			"default": {}
		},
		"definitions": {
			"type": "object",
			"additionalProperties": { "$ref": "#" },
			"default": {}
		},
		"properties": {
			"type": "object",
			"additionalProperties": { "$ref": "#" },
			"default": {}
		},
		"patternProperties": {
			"type": "object",
			"additionalProperties": { "$ref": "#" },
			"default": {}
		},
		"dependencies": {
			"type": "object",
			"additionalProperties": { "anyOf": [{ "$ref": "#" }, { "$ref": "#/definitions/stringArray" }] }
		},
		"enum": {
			"type": "array",
			"minItems": 1,
			"uniqueItems": true
		},
		"type": { "anyOf": [{ "$ref": "#/definitions/simpleTypes" }, {
			"type": "array",
			"items": { "$ref": "#/definitions/simpleTypes" },
			"minItems": 1,
			"uniqueItems": true
		}] },
		"format": { "type": "string" },
		"allOf": { "$ref": "#/definitions/schemaArray" },
		"anyOf": { "$ref": "#/definitions/schemaArray" },
		"oneOf": { "$ref": "#/definitions/schemaArray" },
		"not": { "$ref": "#" }
	},
	dependencies: {
		"exclusiveMaximum": ["maximum"],
		"exclusiveMinimum": ["minimum"]
	},
	"default": {}
};
//#endregion
//#region src/schemas/draft-06-schema.json
var draft_06_schema_default = {
	$schema: "http://json-schema.org/draft-06/schema#",
	$id: "http://json-schema.org/draft-06/schema#",
	title: "Core schema meta-schema",
	definitions: {
		"schemaArray": {
			"type": "array",
			"minItems": 1,
			"items": { "$ref": "#" }
		},
		"nonNegativeInteger": {
			"type": "integer",
			"minimum": 0
		},
		"nonNegativeIntegerDefault0": { "allOf": [{ "$ref": "#/definitions/nonNegativeInteger" }, { "default": 0 }] },
		"simpleTypes": { "enum": [
			"array",
			"boolean",
			"integer",
			"null",
			"number",
			"object",
			"string"
		] },
		"stringArray": {
			"type": "array",
			"items": { "type": "string" },
			"uniqueItems": true,
			"default": []
		}
	},
	type: ["object", "boolean"],
	properties: {
		"$id": {
			"type": "string",
			"format": "uri-reference"
		},
		"$schema": {
			"type": "string",
			"format": "uri"
		},
		"$ref": {
			"type": "string",
			"format": "uri-reference"
		},
		"title": { "type": "string" },
		"description": { "type": "string" },
		"default": {},
		"examples": {
			"type": "array",
			"items": {}
		},
		"multipleOf": {
			"type": "number",
			"exclusiveMinimum": 0
		},
		"maximum": { "type": "number" },
		"exclusiveMaximum": { "type": "number" },
		"minimum": { "type": "number" },
		"exclusiveMinimum": { "type": "number" },
		"maxLength": { "$ref": "#/definitions/nonNegativeInteger" },
		"minLength": { "$ref": "#/definitions/nonNegativeIntegerDefault0" },
		"pattern": {
			"type": "string",
			"format": "regex"
		},
		"additionalItems": { "$ref": "#" },
		"items": {
			"anyOf": [{ "$ref": "#" }, { "$ref": "#/definitions/schemaArray" }],
			"default": {}
		},
		"maxItems": { "$ref": "#/definitions/nonNegativeInteger" },
		"minItems": { "$ref": "#/definitions/nonNegativeIntegerDefault0" },
		"uniqueItems": {
			"type": "boolean",
			"default": false
		},
		"contains": { "$ref": "#" },
		"maxProperties": { "$ref": "#/definitions/nonNegativeInteger" },
		"minProperties": { "$ref": "#/definitions/nonNegativeIntegerDefault0" },
		"required": { "$ref": "#/definitions/stringArray" },
		"additionalProperties": { "$ref": "#" },
		"definitions": {
			"type": "object",
			"additionalProperties": { "$ref": "#" },
			"default": {}
		},
		"properties": {
			"type": "object",
			"additionalProperties": { "$ref": "#" },
			"default": {}
		},
		"patternProperties": {
			"type": "object",
			"additionalProperties": { "$ref": "#" },
			"propertyNames": { "format": "regex" },
			"default": {}
		},
		"dependencies": {
			"type": "object",
			"additionalProperties": { "anyOf": [{ "$ref": "#" }, { "$ref": "#/definitions/stringArray" }] }
		},
		"propertyNames": { "$ref": "#" },
		"const": {},
		"enum": {
			"type": "array",
			"minItems": 1,
			"uniqueItems": true
		},
		"type": { "anyOf": [{ "$ref": "#/definitions/simpleTypes" }, {
			"type": "array",
			"items": { "$ref": "#/definitions/simpleTypes" },
			"minItems": 1,
			"uniqueItems": true
		}] },
		"format": { "type": "string" },
		"allOf": { "$ref": "#/definitions/schemaArray" },
		"anyOf": { "$ref": "#/definitions/schemaArray" },
		"oneOf": { "$ref": "#/definitions/schemaArray" },
		"not": { "$ref": "#" }
	},
	"default": {}
};
//#endregion
//#region src/schemas/draft-07-schema.json
var draft_07_schema_default = {
	$schema: "http://json-schema.org/draft-07/schema#",
	$id: "http://json-schema.org/draft-07/schema#",
	title: "Core schema meta-schema",
	definitions: {
		"schemaArray": {
			"type": "array",
			"minItems": 1,
			"items": { "$ref": "#" }
		},
		"nonNegativeInteger": {
			"type": "integer",
			"minimum": 0
		},
		"nonNegativeIntegerDefault0": { "allOf": [{ "$ref": "#/definitions/nonNegativeInteger" }, { "default": 0 }] },
		"simpleTypes": { "enum": [
			"array",
			"boolean",
			"integer",
			"null",
			"number",
			"object",
			"string"
		] },
		"stringArray": {
			"type": "array",
			"items": { "type": "string" },
			"uniqueItems": true,
			"default": []
		}
	},
	type: ["object", "boolean"],
	properties: {
		"$id": {
			"type": "string",
			"format": "uri-reference"
		},
		"$schema": {
			"type": "string",
			"format": "uri"
		},
		"$ref": {
			"type": "string",
			"format": "uri-reference"
		},
		"$comment": { "type": "string" },
		"title": { "type": "string" },
		"description": { "type": "string" },
		"default": true,
		"readOnly": {
			"type": "boolean",
			"default": false
		},
		"writeOnly": {
			"type": "boolean",
			"default": false
		},
		"examples": {
			"type": "array",
			"items": true
		},
		"multipleOf": {
			"type": "number",
			"exclusiveMinimum": 0
		},
		"maximum": { "type": "number" },
		"exclusiveMaximum": { "type": "number" },
		"minimum": { "type": "number" },
		"exclusiveMinimum": { "type": "number" },
		"maxLength": { "$ref": "#/definitions/nonNegativeInteger" },
		"minLength": { "$ref": "#/definitions/nonNegativeIntegerDefault0" },
		"pattern": {
			"type": "string",
			"format": "regex"
		},
		"additionalItems": { "$ref": "#" },
		"items": {
			"anyOf": [{ "$ref": "#" }, { "$ref": "#/definitions/schemaArray" }],
			"default": true
		},
		"maxItems": { "$ref": "#/definitions/nonNegativeInteger" },
		"minItems": { "$ref": "#/definitions/nonNegativeIntegerDefault0" },
		"uniqueItems": {
			"type": "boolean",
			"default": false
		},
		"contains": { "$ref": "#" },
		"maxProperties": { "$ref": "#/definitions/nonNegativeInteger" },
		"minProperties": { "$ref": "#/definitions/nonNegativeIntegerDefault0" },
		"required": { "$ref": "#/definitions/stringArray" },
		"additionalProperties": { "$ref": "#" },
		"definitions": {
			"type": "object",
			"additionalProperties": { "$ref": "#" },
			"default": {}
		},
		"properties": {
			"type": "object",
			"additionalProperties": { "$ref": "#" },
			"default": {}
		},
		"patternProperties": {
			"type": "object",
			"additionalProperties": { "$ref": "#" },
			"propertyNames": { "format": "regex" },
			"default": {}
		},
		"dependencies": {
			"type": "object",
			"additionalProperties": { "anyOf": [{ "$ref": "#" }, { "$ref": "#/definitions/stringArray" }] }
		},
		"propertyNames": { "$ref": "#" },
		"const": true,
		"enum": {
			"type": "array",
			"items": true,
			"minItems": 1,
			"uniqueItems": true
		},
		"type": { "anyOf": [{ "$ref": "#/definitions/simpleTypes" }, {
			"type": "array",
			"items": { "$ref": "#/definitions/simpleTypes" },
			"minItems": 1,
			"uniqueItems": true
		}] },
		"format": { "type": "string" },
		"contentMediaType": { "type": "string" },
		"contentEncoding": { "type": "string" },
		"if": { "$ref": "#" },
		"then": { "$ref": "#" },
		"else": { "$ref": "#" },
		"allOf": { "$ref": "#/definitions/schemaArray" },
		"anyOf": { "$ref": "#/definitions/schemaArray" },
		"oneOf": { "$ref": "#/definitions/schemaArray" },
		"not": { "$ref": "#" }
	},
	"default": true
};
//#endregion
//#region src/schemas/draft-2019-09-meta-applicator.json
var draft_2019_09_meta_applicator_default = {
	$schema: "https://json-schema.org/draft/2019-09/schema",
	$id: "https://json-schema.org/draft/2019-09/meta/applicator",
	$recursiveAnchor: true,
	title: "Applicator vocabulary meta-schema",
	type: ["object", "boolean"],
	properties: {
		"additionalItems": { "$recursiveRef": "#" },
		"unevaluatedItems": { "$recursiveRef": "#" },
		"items": { "anyOf": [{ "$recursiveRef": "#" }, { "$ref": "#/$defs/schemaArray" }] },
		"contains": { "$recursiveRef": "#" },
		"additionalProperties": { "$recursiveRef": "#" },
		"unevaluatedProperties": { "$recursiveRef": "#" },
		"properties": {
			"type": "object",
			"additionalProperties": { "$recursiveRef": "#" },
			"default": {}
		},
		"patternProperties": {
			"type": "object",
			"additionalProperties": { "$recursiveRef": "#" },
			"propertyNames": { "format": "regex" },
			"default": {}
		},
		"dependentSchemas": {
			"type": "object",
			"additionalProperties": { "$recursiveRef": "#" }
		},
		"propertyNames": { "$recursiveRef": "#" },
		"if": { "$recursiveRef": "#" },
		"then": { "$recursiveRef": "#" },
		"else": { "$recursiveRef": "#" },
		"allOf": { "$ref": "#/$defs/schemaArray" },
		"anyOf": { "$ref": "#/$defs/schemaArray" },
		"oneOf": { "$ref": "#/$defs/schemaArray" },
		"not": { "$recursiveRef": "#" }
	},
	$defs: { "schemaArray": {
		"type": "array",
		"minItems": 1,
		"items": { "$recursiveRef": "#" }
	} }
};
//#endregion
//#region src/schemas/draft-2019-09-meta-content.json
var draft_2019_09_meta_content_default = {
	$schema: "https://json-schema.org/draft/2019-09/schema",
	$id: "https://json-schema.org/draft/2019-09/meta/content",
	$recursiveAnchor: true,
	title: "Content vocabulary meta-schema",
	type: ["object", "boolean"],
	properties: {
		"contentMediaType": { "type": "string" },
		"contentEncoding": { "type": "string" },
		"contentSchema": { "$recursiveRef": "#" }
	}
};
//#endregion
//#region src/schemas/draft-2019-09-meta-core.json
var draft_2019_09_meta_core_default = {
	$schema: "https://json-schema.org/draft/2019-09/schema",
	$id: "https://json-schema.org/draft/2019-09/meta/core",
	$recursiveAnchor: true,
	title: "Core vocabulary meta-schema",
	type: ["object", "boolean"],
	properties: {
		"$id": {
			"type": "string",
			"format": "uri-reference",
			"$comment": "Non-empty fragments not allowed.",
			"pattern": "^[^#]*#?$"
		},
		"$schema": {
			"type": "string",
			"format": "uri"
		},
		"$anchor": {
			"type": "string",
			"pattern": "^[A-Za-z][-A-Za-z0-9.:_]*$"
		},
		"$ref": {
			"type": "string",
			"format": "uri-reference"
		},
		"$recursiveRef": {
			"type": "string",
			"format": "uri-reference"
		},
		"$recursiveAnchor": {
			"type": "boolean",
			"default": false
		},
		"$vocabulary": {
			"type": "object",
			"propertyNames": {
				"type": "string",
				"format": "uri"
			},
			"additionalProperties": { "type": "boolean" }
		},
		"$comment": { "type": "string" },
		"$defs": {
			"type": "object",
			"additionalProperties": { "$recursiveRef": "#" },
			"default": {}
		}
	}
};
//#endregion
//#region src/schemas/draft-2019-09-meta-format.json
var draft_2019_09_meta_format_default = {
	$schema: "https://json-schema.org/draft/2019-09/schema",
	$id: "https://json-schema.org/draft/2019-09/meta/format",
	$recursiveAnchor: true,
	title: "Format vocabulary meta-schema",
	type: ["object", "boolean"],
	properties: { "format": { "type": "string" } }
};
//#endregion
//#region src/schemas/draft-2019-09-meta-meta-data.json
var draft_2019_09_meta_meta_data_default = {
	$schema: "https://json-schema.org/draft/2019-09/schema",
	$id: "https://json-schema.org/draft/2019-09/meta/meta-data",
	$recursiveAnchor: true,
	title: "Meta-data vocabulary meta-schema",
	type: ["object", "boolean"],
	properties: {
		"title": { "type": "string" },
		"description": { "type": "string" },
		"default": true,
		"deprecated": {
			"type": "boolean",
			"default": false
		},
		"readOnly": {
			"type": "boolean",
			"default": false
		},
		"writeOnly": {
			"type": "boolean",
			"default": false
		},
		"examples": {
			"type": "array",
			"items": true
		}
	}
};
//#endregion
//#region src/schemas/draft-2019-09-meta-validation.json
var draft_2019_09_meta_validation_default = {
	$schema: "https://json-schema.org/draft/2019-09/schema",
	$id: "https://json-schema.org/draft/2019-09/meta/validation",
	$recursiveAnchor: true,
	title: "Validation vocabulary meta-schema",
	type: ["object", "boolean"],
	properties: {
		"multipleOf": {
			"type": "number",
			"exclusiveMinimum": 0
		},
		"maximum": { "type": "number" },
		"exclusiveMaximum": { "type": "number" },
		"minimum": { "type": "number" },
		"exclusiveMinimum": { "type": "number" },
		"maxLength": { "$ref": "#/$defs/nonNegativeInteger" },
		"minLength": { "$ref": "#/$defs/nonNegativeIntegerDefault0" },
		"pattern": {
			"type": "string",
			"format": "regex"
		},
		"maxItems": { "$ref": "#/$defs/nonNegativeInteger" },
		"minItems": { "$ref": "#/$defs/nonNegativeIntegerDefault0" },
		"uniqueItems": {
			"type": "boolean",
			"default": false
		},
		"maxContains": { "$ref": "#/$defs/nonNegativeInteger" },
		"minContains": {
			"$ref": "#/$defs/nonNegativeInteger",
			"default": 1
		},
		"maxProperties": { "$ref": "#/$defs/nonNegativeInteger" },
		"minProperties": { "$ref": "#/$defs/nonNegativeIntegerDefault0" },
		"required": { "$ref": "#/$defs/stringArray" },
		"dependentRequired": {
			"type": "object",
			"additionalProperties": { "$ref": "#/$defs/stringArray" }
		},
		"const": true,
		"enum": {
			"type": "array",
			"items": true
		},
		"type": { "anyOf": [{ "$ref": "#/$defs/simpleTypes" }, {
			"type": "array",
			"items": { "$ref": "#/$defs/simpleTypes" },
			"minItems": 1,
			"uniqueItems": true
		}] }
	},
	$defs: {
		"nonNegativeInteger": {
			"type": "integer",
			"minimum": 0
		},
		"nonNegativeIntegerDefault0": {
			"$ref": "#/$defs/nonNegativeInteger",
			"default": 0
		},
		"simpleTypes": { "enum": [
			"array",
			"boolean",
			"integer",
			"null",
			"number",
			"object",
			"string"
		] },
		"stringArray": {
			"type": "array",
			"items": { "type": "string" },
			"uniqueItems": true,
			"default": []
		}
	}
};
//#endregion
//#region src/schemas/draft-2019-09-schema.json
var draft_2019_09_schema_default = {
	$schema: "https://json-schema.org/draft/2019-09/schema",
	$id: "https://json-schema.org/draft/2019-09/schema",
	$vocabulary: {
		"https://json-schema.org/draft/2019-09/vocab/core": true,
		"https://json-schema.org/draft/2019-09/vocab/applicator": true,
		"https://json-schema.org/draft/2019-09/vocab/validation": true,
		"https://json-schema.org/draft/2019-09/vocab/meta-data": true,
		"https://json-schema.org/draft/2019-09/vocab/format": false,
		"https://json-schema.org/draft/2019-09/vocab/content": true
	},
	$recursiveAnchor: true,
	title: "Core and Validation specifications meta-schema",
	allOf: [
		{ "$ref": "meta/core" },
		{ "$ref": "meta/applicator" },
		{ "$ref": "meta/validation" },
		{ "$ref": "meta/meta-data" },
		{ "$ref": "meta/format" },
		{ "$ref": "meta/content" }
	],
	type: ["object", "boolean"],
	properties: {
		"definitions": {
			"$comment": "While no longer an official keyword as it is replaced by $defs, this keyword is retained in the meta-schema to prevent incompatible extensions as it remains in common use.",
			"type": "object",
			"additionalProperties": { "$recursiveRef": "#" },
			"default": {}
		},
		"dependencies": {
			"$comment": "\"dependencies\" is no longer a keyword, but schema authors should avoid redefining it to facilitate a smooth transition to \"dependentSchemas\" and \"dependentRequired\"",
			"type": "object",
			"additionalProperties": { "anyOf": [{ "$recursiveRef": "#" }, { "$ref": "meta/validation#/$defs/stringArray" }] }
		}
	}
};
//#endregion
//#region src/schemas/draft-2020-12-meta-applicator.json
var draft_2020_12_meta_applicator_default = {
	$schema: "https://json-schema.org/draft/2020-12/schema",
	$id: "https://json-schema.org/draft/2020-12/meta/applicator",
	$dynamicAnchor: "meta",
	title: "Applicator vocabulary meta-schema",
	type: ["object", "boolean"],
	properties: {
		"prefixItems": { "$ref": "#/$defs/schemaArray" },
		"items": { "$dynamicRef": "#meta" },
		"contains": { "$dynamicRef": "#meta" },
		"additionalProperties": { "$dynamicRef": "#meta" },
		"properties": {
			"type": "object",
			"additionalProperties": { "$dynamicRef": "#meta" },
			"default": {}
		},
		"patternProperties": {
			"type": "object",
			"additionalProperties": { "$dynamicRef": "#meta" },
			"propertyNames": { "format": "regex" },
			"default": {}
		},
		"dependentSchemas": {
			"type": "object",
			"additionalProperties": { "$dynamicRef": "#meta" },
			"default": {}
		},
		"propertyNames": { "$dynamicRef": "#meta" },
		"if": { "$dynamicRef": "#meta" },
		"then": { "$dynamicRef": "#meta" },
		"else": { "$dynamicRef": "#meta" },
		"allOf": { "$ref": "#/$defs/schemaArray" },
		"anyOf": { "$ref": "#/$defs/schemaArray" },
		"oneOf": { "$ref": "#/$defs/schemaArray" },
		"not": { "$dynamicRef": "#meta" }
	},
	$defs: { "schemaArray": {
		"type": "array",
		"minItems": 1,
		"items": { "$dynamicRef": "#meta" }
	} }
};
//#endregion
//#region src/schemas/draft-2020-12-meta-content.json
var draft_2020_12_meta_content_default = {
	$schema: "https://json-schema.org/draft/2020-12/schema",
	$id: "https://json-schema.org/draft/2020-12/meta/content",
	$dynamicAnchor: "meta",
	title: "Content vocabulary meta-schema",
	type: ["object", "boolean"],
	properties: {
		"contentEncoding": { "type": "string" },
		"contentMediaType": { "type": "string" },
		"contentSchema": { "$dynamicRef": "#meta" }
	}
};
//#endregion
//#region src/schemas/draft-2020-12-meta-core.json
var draft_2020_12_meta_core_default = {
	$schema: "https://json-schema.org/draft/2020-12/schema",
	$id: "https://json-schema.org/draft/2020-12/meta/core",
	$dynamicAnchor: "meta",
	title: "Core vocabulary meta-schema",
	type: ["object", "boolean"],
	properties: {
		"$id": {
			"$ref": "#/$defs/uriReferenceString",
			"$comment": "Non-empty fragments not allowed.",
			"pattern": "^[^#]*#?$"
		},
		"$schema": { "$ref": "#/$defs/uriString" },
		"$ref": { "$ref": "#/$defs/uriReferenceString" },
		"$anchor": { "$ref": "#/$defs/anchorString" },
		"$dynamicRef": { "$ref": "#/$defs/uriReferenceString" },
		"$dynamicAnchor": { "$ref": "#/$defs/anchorString" },
		"$vocabulary": {
			"type": "object",
			"propertyNames": { "$ref": "#/$defs/uriString" },
			"additionalProperties": { "type": "boolean" }
		},
		"$comment": { "type": "string" },
		"$defs": {
			"type": "object",
			"additionalProperties": { "$dynamicRef": "#meta" }
		}
	},
	$defs: {
		"anchorString": {
			"type": "string",
			"pattern": "^[A-Za-z_][-A-Za-z0-9._]*$"
		},
		"uriString": {
			"type": "string",
			"format": "uri"
		},
		"uriReferenceString": {
			"type": "string",
			"format": "uri-reference"
		}
	}
};
//#endregion
//#region src/schemas/draft-2020-12-meta-format-annotation.json
var draft_2020_12_meta_format_annotation_default = {
	$schema: "https://json-schema.org/draft/2020-12/schema",
	$id: "https://json-schema.org/draft/2020-12/meta/format-annotation",
	$dynamicAnchor: "meta",
	title: "Format vocabulary meta-schema for annotation results",
	type: ["object", "boolean"],
	properties: { "format": { "type": "string" } }
};
//#endregion
//#region src/schemas/draft-2020-12-meta-format-assertion.json
var draft_2020_12_meta_format_assertion_default = {
	$schema: "https://json-schema.org/draft/2020-12/schema",
	$id: "https://json-schema.org/draft/2020-12/meta/format-assertion",
	$dynamicAnchor: "meta",
	title: "Format vocabulary meta-schema for assertion results",
	type: ["object", "boolean"],
	properties: { "format": { "type": "string" } }
};
//#endregion
//#region src/schemas/draft-2020-12-meta-meta-data.json
var draft_2020_12_meta_meta_data_default = {
	$schema: "https://json-schema.org/draft/2020-12/schema",
	$id: "https://json-schema.org/draft/2020-12/meta/meta-data",
	$dynamicAnchor: "meta",
	title: "Meta-data vocabulary meta-schema",
	type: ["object", "boolean"],
	properties: {
		"title": { "type": "string" },
		"description": { "type": "string" },
		"default": true,
		"deprecated": {
			"type": "boolean",
			"default": false
		},
		"readOnly": {
			"type": "boolean",
			"default": false
		},
		"writeOnly": {
			"type": "boolean",
			"default": false
		},
		"examples": {
			"type": "array",
			"items": true
		}
	}
};
//#endregion
//#region src/schemas/draft-2020-12-meta-unevaluated.json
var draft_2020_12_meta_unevaluated_default = {
	$schema: "https://json-schema.org/draft/2020-12/schema",
	$id: "https://json-schema.org/draft/2020-12/meta/unevaluated",
	$dynamicAnchor: "meta",
	title: "Unevaluated applicator vocabulary meta-schema",
	type: ["object", "boolean"],
	properties: {
		"unevaluatedItems": { "$dynamicRef": "#meta" },
		"unevaluatedProperties": { "$dynamicRef": "#meta" }
	}
};
//#endregion
//#region src/schemas/draft-2020-12-meta-validation.json
var draft_2020_12_meta_validation_default = {
	$schema: "https://json-schema.org/draft/2020-12/schema",
	$id: "https://json-schema.org/draft/2020-12/meta/validation",
	$dynamicAnchor: "meta",
	title: "Validation vocabulary meta-schema",
	type: ["object", "boolean"],
	properties: {
		"type": { "anyOf": [{ "$ref": "#/$defs/simpleTypes" }, {
			"type": "array",
			"items": { "$ref": "#/$defs/simpleTypes" },
			"minItems": 1,
			"uniqueItems": true
		}] },
		"const": true,
		"enum": {
			"type": "array",
			"items": true
		},
		"multipleOf": {
			"type": "number",
			"exclusiveMinimum": 0
		},
		"maximum": { "type": "number" },
		"exclusiveMaximum": { "type": "number" },
		"minimum": { "type": "number" },
		"exclusiveMinimum": { "type": "number" },
		"maxLength": { "$ref": "#/$defs/nonNegativeInteger" },
		"minLength": { "$ref": "#/$defs/nonNegativeIntegerDefault0" },
		"pattern": {
			"type": "string",
			"format": "regex"
		},
		"maxItems": { "$ref": "#/$defs/nonNegativeInteger" },
		"minItems": { "$ref": "#/$defs/nonNegativeIntegerDefault0" },
		"uniqueItems": {
			"type": "boolean",
			"default": false
		},
		"maxContains": { "$ref": "#/$defs/nonNegativeInteger" },
		"minContains": {
			"$ref": "#/$defs/nonNegativeInteger",
			"default": 1
		},
		"maxProperties": { "$ref": "#/$defs/nonNegativeInteger" },
		"minProperties": { "$ref": "#/$defs/nonNegativeIntegerDefault0" },
		"required": { "$ref": "#/$defs/stringArray" },
		"dependentRequired": {
			"type": "object",
			"additionalProperties": { "$ref": "#/$defs/stringArray" }
		}
	},
	$defs: {
		"nonNegativeInteger": {
			"type": "integer",
			"minimum": 0
		},
		"nonNegativeIntegerDefault0": {
			"$ref": "#/$defs/nonNegativeInteger",
			"default": 0
		},
		"simpleTypes": { "enum": [
			"array",
			"boolean",
			"integer",
			"null",
			"number",
			"object",
			"string"
		] },
		"stringArray": {
			"type": "array",
			"items": { "type": "string" },
			"uniqueItems": true,
			"default": []
		}
	}
};
//#endregion
//#region src/schemas/draft-2020-12-schema.json
var draft_2020_12_schema_default = {
	$schema: "https://json-schema.org/draft/2020-12/schema",
	$id: "https://json-schema.org/draft/2020-12/schema",
	$vocabulary: {
		"https://json-schema.org/draft/2020-12/vocab/core": true,
		"https://json-schema.org/draft/2020-12/vocab/applicator": true,
		"https://json-schema.org/draft/2020-12/vocab/unevaluated": true,
		"https://json-schema.org/draft/2020-12/vocab/validation": true,
		"https://json-schema.org/draft/2020-12/vocab/meta-data": true,
		"https://json-schema.org/draft/2020-12/vocab/format-annotation": true,
		"https://json-schema.org/draft/2020-12/vocab/content": true
	},
	$dynamicAnchor: "meta",
	title: "Core and Validation specifications meta-schema",
	allOf: [
		{ "$ref": "meta/core" },
		{ "$ref": "meta/applicator" },
		{ "$ref": "meta/unevaluated" },
		{ "$ref": "meta/validation" },
		{ "$ref": "meta/meta-data" },
		{ "$ref": "meta/format-annotation" },
		{ "$ref": "meta/content" }
	],
	type: ["object", "boolean"],
	$comment: "This meta-schema also defines keywords that have appeared in previous drafts in order to prevent incompatible extensions as they remain in common use.",
	properties: {
		"definitions": {
			"$comment": "\"definitions\" has been replaced by \"$defs\".",
			"type": "object",
			"additionalProperties": { "$dynamicRef": "#meta" },
			"deprecated": true,
			"default": {}
		},
		"dependencies": {
			"$comment": "\"dependencies\" has been split and replaced by \"dependentSchemas\" and \"dependentRequired\" in order to serve their differing semantics.",
			"type": "object",
			"additionalProperties": { "anyOf": [{ "$dynamicRef": "#meta" }, { "$ref": "meta/validation#/$defs/stringArray" }] },
			"deprecated": true,
			"default": {}
		},
		"$recursiveAnchor": {
			"$comment": "\"$recursiveAnchor\" has been replaced by \"$dynamicAnchor\".",
			"$ref": "meta/core#/$defs/anchorString",
			"deprecated": true
		},
		"$recursiveRef": {
			"$comment": "\"$recursiveRef\" has been replaced by \"$dynamicRef\".",
			"$ref": "meta/core#/$defs/uriReferenceString",
			"deprecated": true
		}
	}
};
//#endregion
//#region src/z-schema-versions.ts
const Draft4Schema = draft_04_schema_default;
const Draft6Schema = draft_06_schema_default;
const Draft7Schema = draft_07_schema_default;
const Draft201909Schema = draft_2019_09_schema_default;
const Draft201909MetaApplicator = draft_2019_09_meta_applicator_default;
const Draft201909MetaContent = draft_2019_09_meta_content_default;
const Draft201909MetaCore = draft_2019_09_meta_core_default;
const Draft201909MetaFormat = draft_2019_09_meta_format_default;
const Draft201909MetaMetaData = draft_2019_09_meta_meta_data_default;
const Draft201909MetaValidation = draft_2019_09_meta_validation_default;
const Draft202012Schema = draft_2020_12_schema_default;
const Draft202012MetaApplicator = draft_2020_12_meta_applicator_default;
const Draft202012MetaContent = draft_2020_12_meta_content_default;
const Draft202012MetaCore = draft_2020_12_meta_core_default;
const Draft202012MetaFormatAnnotation = draft_2020_12_meta_format_annotation_default;
const Draft202012MetaFormatAssertion = draft_2020_12_meta_format_assertion_default;
const Draft202012MetaMetaData = draft_2020_12_meta_meta_data_default;
const Draft202012MetaUnevaluated = draft_2020_12_meta_unevaluated_default;
const Draft202012MetaValidation = draft_2020_12_meta_validation_default;
const registerRemoteReference = (uri, schema, validationOptions) => {
	const preparedSchema = schema;
	if (!preparedSchema.id) preparedSchema.id = uri;
	if (validationOptions) preparedSchema.__$validationOptions = normalizeOptions(validationOptions);
	SchemaCache.cacheSchemaByUri(uri, preparedSchema);
};
registerRemoteReference("http://json-schema.org/draft-04/schema", Draft4Schema, { version: "none" });
registerRemoteReference("http://json-schema.org/draft-06/schema", Draft6Schema, { version: "none" });
registerRemoteReference("http://json-schema.org/draft-07/schema", Draft7Schema, { version: "none" });
registerRemoteReference("https://json-schema.org/draft/2019-09/schema", Draft201909Schema, { version: "none" });
registerRemoteReference("https://json-schema.org/draft/2019-09/meta/applicator", Draft201909MetaApplicator, { version: "none" });
registerRemoteReference("https://json-schema.org/draft/2019-09/meta/content", Draft201909MetaContent, { version: "none" });
registerRemoteReference("https://json-schema.org/draft/2019-09/meta/core", Draft201909MetaCore, { version: "none" });
registerRemoteReference("https://json-schema.org/draft/2019-09/meta/format", Draft201909MetaFormat, { version: "none" });
registerRemoteReference("https://json-schema.org/draft/2019-09/meta/meta-data", Draft201909MetaMetaData, { version: "none" });
registerRemoteReference("https://json-schema.org/draft/2019-09/meta/validation", Draft201909MetaValidation, { version: "none" });
registerRemoteReference("https://json-schema.org/draft/2020-12/schema", Draft202012Schema, { version: "none" });
registerRemoteReference("https://json-schema.org/draft/2020-12/meta/applicator", Draft202012MetaApplicator, { version: "none" });
registerRemoteReference("https://json-schema.org/draft/2020-12/meta/content", Draft202012MetaContent, { version: "none" });
registerRemoteReference("https://json-schema.org/draft/2020-12/meta/core", Draft202012MetaCore, { version: "none" });
registerRemoteReference("https://json-schema.org/draft/2020-12/meta/format-annotation", Draft202012MetaFormatAnnotation, { version: "none" });
registerRemoteReference("https://json-schema.org/draft/2020-12/meta/format-assertion", Draft202012MetaFormatAssertion, { version: "none" });
registerRemoteReference("https://json-schema.org/draft/2020-12/meta/meta-data", Draft202012MetaMetaData, { version: "none" });
registerRemoteReference("https://json-schema.org/draft/2020-12/meta/unevaluated", Draft202012MetaUnevaluated, { version: "none" });
registerRemoteReference("https://json-schema.org/draft/2020-12/meta/validation", Draft202012MetaValidation, { version: "none" });
//#endregion
//#region src/utils/date.ts
const isLeapYear = (year) => year % 4 === 0 && (year % 100 !== 0 || year % 400 === 0);
const getDaysInMonth = (year, month) => {
	switch (month) {
		case 2: return isLeapYear(year) ? 29 : 28;
		case 4:
		case 6:
		case 9:
		case 11: return 30;
		default: return 31;
	}
};
const isValidRfc3339Date = (year, month, day) => {
	if (month < 1 || month > 12) return false;
	const maxDay = getDaysInMonth(year, month);
	return day >= 1 && day <= maxDay;
};
//#endregion
//#region src/utils/hostname.ts
const IDN_SEPARATOR_REGEX = /[\u3002\uff0e\uff61]/g;
const IDN_SEPARATOR_TEST_REGEX = /[\u3002\uff0e\uff61]/;
const splitHostnameLabels = (hostname) => {
	if (hostname.length === 0 || hostname.length > 255) return null;
	if (hostname.startsWith(".") || hostname.endsWith(".")) return null;
	const labels = hostname.split(".");
	if (labels.some((label) => label.length === 0 || label.length > 63)) return null;
	return labels;
};
const isAsciiHostnameLabel = (label) => /^[a-z0-9](?:[a-z0-9-]*[a-z0-9])?$/i.test(label);
const isGreek = (char) => /\p{Script=Greek}/u.test(char);
const isHebrew = (char) => /\p{Script=Hebrew}/u.test(char);
const hasCjkKanaOrHan = (input) => /[\p{Script=Hiragana}\p{Script=Katakana}\p{Script=Han}]/u.test(input);
const toUnicodeLabel = (label) => {
	if (!/^xn--/i.test(label)) return label;
	try {
		return punycode_punycode_js.default.toUnicode(label.toLowerCase());
	} catch (_e) {
		return null;
	}
};
const isValidIdnUnicodeLabel = (label) => {
	if (label.startsWith("-") || label.endsWith("-")) return false;
	if (label.length >= 4 && label[2] === "-" && label[3] === "-" && !/^xn--/i.test(label)) return false;
	if (/^\p{M}/u.test(label)) return false;
	if (/[\u302e\u302f\u0640\u07fa]/u.test(label)) return false;
	for (let idx = 0; idx < label.length; idx++) {
		const char = label[idx];
		if (char === "·") {
			if (idx === 0 || idx === label.length - 1 || label[idx - 1] !== "l" || label[idx + 1] !== "l") return false;
		}
		if (char === "͵") {
			if (idx === label.length - 1 || !isGreek(label[idx + 1])) return false;
		}
		if (char === "׳" || char === "״") {
			if (idx === 0 || !isHebrew(label[idx - 1])) return false;
		}
		if (char === "‍") {
			if (idx === 0 || label[idx - 1] !== "्") return false;
		}
	}
	if (label.includes("・") && !hasCjkKanaOrHan(label.replace(/\u30fb/g, ""))) return false;
	const hasArabicIndic = /[\u0660-\u0669]/.test(label);
	const hasExtendedArabicIndic = /[\u06f0-\u06f9]/.test(label);
	if (hasArabicIndic && hasExtendedArabicIndic) return false;
	return true;
};
const isValidHostname = (hostname) => {
	if (IDN_SEPARATOR_TEST_REGEX.test(hostname) || /[^\x00-\x7F]/.test(hostname)) return false;
	if (validator_lib_isIP_js.default.default(hostname, 4)) return false;
	const labels = splitHostnameLabels(hostname);
	if (labels === null) return false;
	for (const label of labels) {
		if (!isAsciiHostnameLabel(label)) return false;
		if (/^xn--/i.test(label)) {
			const unicodeLabel = toUnicodeLabel(label);
			if (unicodeLabel === null || !isValidIdnUnicodeLabel(unicodeLabel)) return false;
		}
	}
	return true;
};
const isValidIdnHostname = (hostname) => {
	const labels = splitHostnameLabels(hostname.replace(IDN_SEPARATOR_REGEX, "."));
	if (labels === null) return false;
	for (const label of labels) {
		const unicodeLabel = toUnicodeLabel(label);
		if (unicodeLabel === null || !isValidIdnUnicodeLabel(unicodeLabel)) return false;
	}
	return true;
};
//#endregion
//#region src/utils/time.ts
const toUtcTime = (hour, minute, offsetSign, offsetHour, offsetMinute) => {
	const localTotalMinutes = hour * 60 + minute;
	const offsetTotalMinutes = offsetHour * 60 + offsetMinute;
	const normalizedUtcTotalMinutes = ((offsetSign === "+" ? localTotalMinutes - offsetTotalMinutes : localTotalMinutes + offsetTotalMinutes) % 1440 + 1440) % 1440;
	return {
		hour: Math.floor(normalizedUtcTotalMinutes / 60),
		minute: normalizedUtcTotalMinutes % 60
	};
};
const RFC3339_TIME_REGEX = /^([0-9]{2}):([0-9]{2}):([0-9]{2})(\.[0-9]+)?(z|([+-][0-9]{2}:[0-9]{2}))$/i;
const parseRfc3339Time = (time) => {
	const matches = RFC3339_TIME_REGEX.exec(time);
	if (matches === null) return null;
	const hour = parseInt(matches[1], 10);
	const minute = parseInt(matches[2], 10);
	const second = parseInt(matches[3], 10);
	if (hour > 23 || minute > 59 || second > 60) return null;
	let utcHour = hour;
	let utcMinute = minute;
	if (matches[5].toLowerCase() !== "z") {
		const offsetMatches = /^([+-])([0-9]{2}):([0-9]{2})$/.exec(matches[5]);
		if (offsetMatches === null) return null;
		const offsetSign = offsetMatches[1];
		const offsetHour = parseInt(offsetMatches[2], 10);
		const offsetMinute = parseInt(offsetMatches[3], 10);
		if (offsetHour > 23 || offsetMinute > 59) return null;
		const utc = toUtcTime(hour, minute, offsetSign, offsetHour, offsetMinute);
		utcHour = utc.hour;
		utcMinute = utc.minute;
	}
	if (second === 60 && (utcHour !== 23 || utcMinute !== 59)) return null;
	return {
		hour,
		minute,
		second,
		utcHour,
		utcMinute
	};
};
//#endregion
//#region src/format-validators.ts
const dateValidator = (date) => {
	if (typeof date !== "string") return true;
	const matches = /^([0-9]{4})-([0-9]{2})-([0-9]{2})$/.exec(date);
	if (matches === null) return false;
	return isValidRfc3339Date(parseInt(matches[1], 10), parseInt(matches[2], 10), parseInt(matches[3], 10));
};
const dateTimeValidator = (dateTime) => {
	if (typeof dateTime !== "string") return true;
	const s = dateTime.toLowerCase().split("t");
	if (s.length !== 2) return false;
	const datePart = s[0];
	const timePart = s[1];
	const dateMatches = /^([0-9]{4})-([0-9]{2})-([0-9]{2})$/.exec(datePart);
	if (dateMatches === null) return false;
	if (!isValidRfc3339Date(parseInt(dateMatches[1], 10), parseInt(dateMatches[2], 10), parseInt(dateMatches[3], 10))) return false;
	return parseRfc3339Time(timePart) !== null;
};
const emailValidator = (email) => {
	if (typeof email !== "string") return true;
	if (validator_lib_isEmail_js.default.default(email, {
		require_tld: true,
		allow_ip_domain: true
	})) return true;
	const ipv6Literal = /^(.+)@\[IPv6:([^\]]+)\]$/i.exec(email);
	if (!ipv6Literal) return false;
	const localPart = ipv6Literal[1];
	const addressPart = ipv6Literal[2];
	if (!validator_lib_isIP_js.default.default(addressPart, 6)) return false;
	return validator_lib_isEmail_js.default.default(`${localPart}@example.com`, { require_tld: true });
};
const hostnameValidator = (hostname) => {
	if (typeof hostname !== "string") return true;
	return isValidHostname(hostname);
};
const ipv4Validator = (ipv4) => {
	if (typeof ipv4 !== "string") return true;
	return validator_lib_isIP_js.default.default(ipv4, 4);
};
const ipv6Validator = (ipv6) => {
	if (typeof ipv6 !== "string") return true;
	if (ipv6.includes("%")) return false;
	return validator_lib_isIP_js.default.default(ipv6, 6);
};
const regexValidator = (input) => {
	if (typeof input !== "string") return true;
	const invalidEscapes = new Set(["a"]);
	for (let idx = 0; idx < input.length; idx++) {
		if (input[idx] !== "\\") continue;
		idx++;
		if (idx >= input.length) return false;
		const escaped = input[idx];
		if (invalidEscapes.has(escaped)) return false;
	}
	try {
		RegExp(input);
		return true;
	} catch (_e) {
		return false;
	}
};
const durationValidator = (input) => {
	if (typeof input !== "string") return true;
	if (!/^P[\x00-\x7F]*$/.test(input)) return false;
	if (!input.startsWith("P")) return false;
	const body = input.slice(1);
	if (body.length === 0) return false;
	if (body.includes("W")) return /^\d+W$/.test(body);
	const parts = body.split("T");
	if (parts.length > 2) return false;
	const datePart = parts[0];
	const timePart = parts.length === 2 ? parts[1] : void 0;
	if (datePart.length > 0 && !/^(?:\d+Y(?:\d+M(?:\d+D)?)?|\d+M(?:\d+D)?|\d+D)$/.test(datePart)) return false;
	const hasDateComponent = /\d+[YMD]/.test(datePart);
	let hasTimeComponent = false;
	if (timePart !== void 0) {
		if (timePart.length === 0) return false;
		if (!/^(?:\d+H(?:\d+M(?:\d+S)?)?|\d+M(?:\d+S)?|\d+S)$/.test(timePart)) return false;
		hasTimeComponent = /\d+[HMS]/.test(timePart);
		if (!hasTimeComponent) return false;
	}
	return hasDateComponent || hasTimeComponent;
};
const uuidValidator = (input) => {
	if (typeof input !== "string") return true;
	return /^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/.test(input);
};
const strictUriValidator = (uri) => typeof uri !== "string" || validator_lib_isURL_js.default.default(uri);
const hasValidPercentEncoding = (str) => {
	for (let i = 0; i < str.length; i++) if (str[i] === "%") {
		if (i + 2 >= str.length || !/[0-9a-fA-F]/.test(str[i + 1]) || !/[0-9a-fA-F]/.test(str[i + 2])) return false;
	}
	return true;
};
const uriValidator = function(uri) {
	if (typeof uri !== "string") return true;
	if (/[^\x00-\x7F]/.test(uri)) return false;
	if (!hasValidPercentEncoding(uri)) return false;
	const match = uri.match(/^([a-zA-Z][a-zA-Z0-9+.-]*):\/\/([^/?#]*)/);
	if (match) {
		const authority = match[2];
		const atIndex = authority.indexOf("@");
		if (atIndex > 0) {
			const userinfo = authority.substring(0, atIndex);
			if (userinfo.includes("[") || userinfo.includes("]")) return false;
		}
		let hostPort = atIndex >= 0 ? authority.substring(atIndex + 1) : authority;
		if (hostPort.startsWith("[")) {
			const bracketEnd = hostPort.indexOf("]");
			if (bracketEnd >= 0) hostPort = hostPort.substring(bracketEnd + 1);
		}
		const colonIndex = hostPort.lastIndexOf(":");
		if (colonIndex >= 0) {
			const port = hostPort.substring(colonIndex + 1);
			if (port.length > 0 && !/^\d+$/.test(port)) return false;
		}
	}
	return /^[a-zA-Z][a-zA-Z0-9+.-]*:[^"\\<>^{}^`| ]*$/.test(uri);
};
const uriReferenceValidator = (uri) => {
	if (typeof uri !== "string") return true;
	if (/[^\x00-\x7F]/.test(uri)) return false;
	return /^([a-zA-Z][a-zA-Z0-9+.-]*:)?[^"\\<>^{}^`| ]*$/.test(uri);
};
const uriTemplateValidator = (uri) => {
	if (typeof uri !== "string") return true;
	if (!/^([a-zA-Z][a-zA-Z0-9+.-]*:)?[^"\\<>^`| ]*$/.test(uri)) return false;
	let inExpression = false;
	for (let idx = 0; idx < uri.length; idx++) {
		const ch = uri[idx];
		if (ch === "{") {
			if (inExpression) return false;
			inExpression = true;
		} else if (ch === "}") {
			if (!inExpression) return false;
			inExpression = false;
		}
	}
	return !inExpression;
};
const hasValidTildeEscapes = (segment) => {
	for (let i = 0; i < segment.length; i++) if (segment[i] === "~") {
		const next = segment[i + 1];
		if (next !== "0" && next !== "1") return false;
		i++;
	}
	return true;
};
const jsonPointerValidator = (pointer) => {
	if (typeof pointer !== "string") return true;
	if (pointer === "") return true;
	if (!/^(?:\/[^/]*)+$/.test(pointer)) return false;
	const tokens = pointer.split("/").slice(1);
	for (const token of tokens) if (!hasValidTildeEscapes(token)) return false;
	return true;
};
const relativeJsonPointerValidator = (pointer) => {
	if (typeof pointer !== "string") return true;
	const match = pointer.match(/^(0|[1-9]\d*)(.*)$/);
	if (!match) return false;
	const suffix = match[2];
	if (suffix === "" || suffix === "#") return true;
	if (!suffix.startsWith("/")) return false;
	if (!/^(?:\/[^/]*)+$/.test(suffix)) return false;
	const tokens = suffix.split("/").slice(1);
	for (const token of tokens) if (!hasValidTildeEscapes(token)) return false;
	return true;
};
const timeValidator = (time) => {
	if (typeof time !== "string") return true;
	return parseRfc3339Time(time) !== null;
};
const idnEmailValidator = (email) => {
	if (typeof email !== "string") return true;
	return /^[^\s@]+@[^\s@]+$/.test(email);
};
const idnHostnameValidator = (hostname) => {
	if (typeof hostname !== "string") return true;
	return isValidIdnHostname(hostname);
};
const iriValidator = (iri) => {
	if (typeof iri !== "string") return true;
	if (!/^[a-zA-Z][a-zA-Z0-9+.-]*:[^"\\<>^{}^`| ]*$/u.test(iri)) return false;
	try {
		new URL(iri);
		return true;
	} catch (_e) {
		return false;
	}
};
const iriReferenceValidator = (iriReference) => {
	if (typeof iriReference !== "string") return true;
	return /^([a-zA-Z][a-zA-Z0-9+.-]*:)?[^"\\<>^{}^`| ]*$/u.test(iriReference);
};
const inbuiltValidators = {
	date: dateValidator,
	"date-time": dateTimeValidator,
	email: emailValidator,
	hostname: hostnameValidator,
	"host-name": hostnameValidator,
	ipv4: ipv4Validator,
	ipv6: ipv6Validator,
	regex: regexValidator,
	uri: uriValidator,
	"strict-uri": strictUriValidator,
	"uri-reference": uriReferenceValidator,
	"uri-template": uriTemplateValidator,
	"json-pointer": jsonPointerValidator,
	"relative-json-pointer": relativeJsonPointerValidator,
	time: timeValidator,
	"idn-email": idnEmailValidator,
	"idn-hostname": idnHostnameValidator,
	iri: iriValidator,
	"iri-reference": iriReferenceValidator,
	duration: durationValidator,
	uuid: uuidValidator
};
const customValidators = {};
function getFormatValidators(options) {
	return {
		...inbuiltValidators,
		...options?.strictUris ? { uri: strictUriValidator } : {},
		...customValidators,
		...options?.customFormats || {}
	};
}
function registerFormat(name, validatorFunction) {
	customValidators[name] = validatorFunction;
}
function unregisterFormat(name) {
	delete customValidators[name];
}
function getSupportedFormats(customFormats) {
	const merged = {
		...inbuiltValidators,
		...customValidators,
		...customFormats
	};
	return sortedKeys(merged).filter((key) => merged[key] != null);
}
function isFormatSupported(name, customFormats) {
	if (customFormats) {
		const custom = customFormats[name];
		if (custom === null) return false;
		if (custom != null) return true;
	}
	if (name in customValidators) return customValidators[name] != null;
	return name in inbuiltValidators;
}
function getRegisteredFormats() {
	return sortedKeys(customValidators);
}
//#endregion
//#region src/utils/schema-regex.ts
function compileSchemaRegex(pattern) {
	if (pattern.length > 1e4) return {
		ok: false,
		error: {
			pattern,
			message: `Pattern length ${pattern.length} exceeds maximum allowed length of ${MAX_SCHEMA_REGEX_LENGTH}`
		}
	};
	if (!(0, safe_regex2.default)(pattern)) return {
		ok: false,
		error: {
			pattern,
			message: "Pattern rejected as potentially unsafe (ReDoS)"
		}
	};
	if (/\\[pP]{/.test(pattern) || /[\u{10000}-\u{10FFFF}]/u.test(pattern) || /\\uD[89AB][0-9A-Fa-f]{2}\\uD[CDEF][0-9A-Fa-f]{2}/.test(pattern)) try {
		return {
			ok: true,
			value: new RegExp(pattern, "u")
		};
	} catch (e) {
		return {
			ok: false,
			error: {
				pattern,
				message: e && e.message ? e.message : "Invalid regular expression"
			}
		};
	}
	else try {
		return {
			ok: true,
			value: new RegExp(pattern)
		};
	} catch (e) {
		return {
			ok: false,
			error: {
				pattern,
				message: e && e.message ? e.message : "Invalid regular expression"
			}
		};
	}
}
//#endregion
//#region src/utils/array.ts
/**
* Check if all elements in an array are unique.
*
* Uses a Set-based fast path for arrays of pure primitives (O(n)).
* Falls back to pairwise deep comparison (O(n²)) when the array contains
* objects or arrays that need structural equality checks.
*/
const isUniqueArray = (arr, indexes, maxDepth) => {
	const l = arr.length;
	if (l <= 1) return true;
	let allPrimitive = true;
	for (let i = 0; i < l; i++) {
		const v = arr[i];
		if (v !== null && typeof v === "object") {
			allPrimitive = false;
			break;
		}
	}
	if (allPrimitive) {
		const seen = /* @__PURE__ */ new Set();
		for (let i = 0; i < l; i++) {
			const v = arr[i];
			const key = typeof v + ":" + String(v);
			if (seen.has(key)) {
				if (indexes) for (let j = 0; j < i; j++) {
					const prev = arr[j];
					if (typeof prev === typeof v && prev === v) {
						indexes.push(j, i);
						break;
					}
				}
				return false;
			}
			seen.add(key);
		}
		return true;
	}
	for (let i = 0; i < l; i++) for (let j = i + 1; j < l; j++) if (areEqual(arr[i], arr[j], { maxDepth })) {
		if (indexes) indexes.push(i, j);
		return false;
	}
	return true;
};
const difference = (bigSet, subSet) => {
	const exclusions = new Set(subSet);
	const arr = [];
	let idx = bigSet.length;
	while (idx--) if (!exclusions.has(bigSet[idx])) arr.push(bigSet[idx]);
	return arr;
};
//#endregion
//#region src/validation/shared.ts
const shouldSkipValidate = function(options, errors) {
	return options && Array.isArray(options.includeErrors) && options.includeErrors.length > 0 && !errors.some(function(err) {
		return options.includeErrors.includes(err);
	});
};
const supportsDependentKeywords = (schema, version) => {
	if (typeof schema.$schema === "string") return !/draft-04|draft-06|draft-07/.test(schema.$schema);
	return !(version === "draft-04" || version === "draft-06" || version === "draft-07");
};
const VOCAB_VALIDATION_2019_09 = "https://json-schema.org/draft/2019-09/vocab/validation";
const VOCAB_VALIDATION_2020_12 = "https://json-schema.org/draft/2020-12/vocab/validation";
const VOCAB_FORMAT_2019_09 = "https://json-schema.org/draft/2019-09/vocab/format";
const VOCAB_FORMAT_ASSERTION_2020_12 = "https://json-schema.org/draft/2020-12/vocab/format-assertion";
const VALIDATION_VOCAB_KEYWORDS = new Set([
	"type",
	"multipleOf",
	"maximum",
	"exclusiveMaximum",
	"minimum",
	"exclusiveMinimum",
	"maxLength",
	"minLength",
	"pattern",
	"maxItems",
	"minItems",
	"uniqueItems",
	"maxContains",
	"minContains",
	"maxProperties",
	"minProperties",
	"required",
	"dependentRequired",
	"enum",
	"const",
	"contentEncoding",
	"contentMediaType"
]);
const isValidationVocabularyEnabled = (schema, report, version) => {
	if (version !== "draft2019-09" && version !== "draft2020-12") return true;
	const currentSchemaMeta = schema.__$schemaResolved;
	const rootSchemaMeta = report.rootSchema && typeof report.rootSchema !== "boolean" ? report.rootSchema.__$schemaResolved : void 0;
	const metaSchema = currentSchemaMeta || rootSchemaMeta;
	if (!metaSchema || typeof metaSchema !== "object" || !isObject(metaSchema.$vocabulary)) return true;
	const vocabulary = metaSchema.$vocabulary;
	if (Object.hasOwn(vocabulary, VOCAB_VALIDATION_2019_09) || Object.hasOwn(vocabulary, VOCAB_VALIDATION_2020_12)) return vocabulary[VOCAB_VALIDATION_2019_09] === true || vocabulary[VOCAB_VALIDATION_2020_12] === true;
	return false;
};
/**
* Checks whether the format-assertion vocabulary is enabled in the meta-schema.
* For draft 2019-09: checks if the format vocabulary is set to true.
* For draft 2020-12: checks if the format-assertion vocabulary is present and true.
* Returns true for older drafts (format was always an assertion).
*/
const isFormatAssertionVocabEnabled = (schema, report, version) => {
	if (version !== "draft2019-09" && version !== "draft2020-12") return true;
	const currentSchemaMeta = schema.__$schemaResolved;
	const rootSchemaMeta = report.rootSchema && typeof report.rootSchema !== "boolean" ? report.rootSchema.__$schemaResolved : void 0;
	const metaSchema = currentSchemaMeta || rootSchemaMeta;
	if (!metaSchema || typeof metaSchema !== "object" || !isObject(metaSchema.$vocabulary)) return false;
	const vocabulary = metaSchema.$vocabulary;
	if (Object.hasOwn(vocabulary, VOCAB_FORMAT_ASSERTION_2020_12)) return vocabulary[VOCAB_FORMAT_ASSERTION_2020_12] === true;
	if (Object.hasOwn(vocabulary, VOCAB_FORMAT_2019_09)) return vocabulary[VOCAB_FORMAT_2019_09] === true;
	return false;
};
function cacheValidationResult(report, schema, json, passed) {
	let schemaMap = report.__validationResultCache.get(schema);
	if (!schemaMap) {
		schemaMap = /* @__PURE__ */ new Map();
		report.__validationResultCache.set(schema, schemaMap);
	}
	schemaMap.set(json, passed);
}
function getCachedValidationResult(report, schema, json) {
	return report.__validationResultCache.get(schema)?.get(json);
}
/**
* Shared async-task aggregation pattern.
*
* 1. Collects async tasks from `subReports` into `report`.
* 2. If new async tasks were added, defers `decisionFn` via `report.addAsyncTask`
*    with proper path save/restore so error paths are correct.
* 3. Otherwise, runs `decisionFn` synchronously.
*/
function deferOrRunSync(report, subReports, decisionFn) {
	const asyncTasksBefore = report.asyncTasks.length;
	for (const subReport of subReports) report.asyncTasks.push(...subReport.asyncTasks);
	if (report.asyncTasks.length > asyncTasksBefore) report.addAsyncTaskWithPath((callback) => {
		setTimeout(() => callback(null), 0);
	}, [], () => {
		decisionFn();
	});
	else decisionFn();
}
//#endregion
//#region src/validation/array.ts
function additionalItemsValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["ARRAY_ADDITIONAL_ITEMS"])) return;
	if (!Array.isArray(json)) return;
	if (schema.additionalItems === false && Array.isArray(schema.items)) {
		if (json.length > schema.items.length) report.addError("ARRAY_ADDITIONAL_ITEMS", void 0, void 0, schema, "additionalItems");
	}
}
function itemsValidator() {}
function prefixItemsValidator() {}
function maxItemsValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["ARRAY_LENGTH_LONG"])) return;
	if (!Array.isArray(json)) return;
	if (json.length > schema.maxItems) report.addError("ARRAY_LENGTH_LONG", [json.length, schema.maxItems], void 0, schema, "maxItems");
}
function minItemsValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["ARRAY_LENGTH_SHORT"])) return;
	if (!Array.isArray(json)) return;
	if (json.length < schema.minItems) report.addError("ARRAY_LENGTH_SHORT", [json.length, schema.minItems], void 0, schema, "minItems");
}
function uniqueItemsValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["ARRAY_UNIQUE"])) return;
	if (!Array.isArray(json)) return;
	if (schema.uniqueItems === true) {
		const matches = [];
		if (isUniqueArray(json, matches, this.options.maxRecursionDepth) === false) report.addError("ARRAY_UNIQUE", matches, void 0, schema, "uniqueItems");
	}
}
function containsValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["CONTAINS"])) return;
	if (!Array.isArray(json)) return;
	const containsSchema = schema.contains;
	if (containsSchema === void 0) return;
	const Report_ = report.constructor;
	const subReports = [];
	for (let idx = 0; idx < json.length; idx++) {
		const subReport = new Report_(report);
		subReports.push(subReport);
		this._jsonValidate(subReport, containsSchema, json[idx]);
		cacheValidationResult(report, containsSchema, json[idx], subReport.errors.length === 0);
	}
	const addContainsErrorIfNeeded = () => {
		let matchingItems = 0;
		for (const subReport of subReports) if (subReport.errors.length === 0) matchingItems += 1;
		const supportsContainsBounds = this.options.version === "draft2019-09" || this.options.version === "draft2020-12";
		const minContains = supportsContainsBounds && typeof schema.minContains === "number" ? schema.minContains ?? 1 : 1;
		const maxContains = supportsContainsBounds && typeof schema.maxContains === "number" ? schema.maxContains : void 0;
		if (!(matchingItems >= minContains) || !(maxContains === void 0 || matchingItems <= maxContains)) report.addError("CONTAINS", void 0, subReports, schema, void 0);
	};
	deferOrRunSync(report, subReports, addContainsErrorIfNeeded);
}
function maxContainsValidator() {}
function minContainsValidator() {}
//#endregion
//#region src/validation/combinators.ts
function allOfValidator(report, schema, json) {
	for (let i = 0; i < schema.allOf.length; i++) {
		const validateResult = this._jsonValidate(report, schema.allOf[i], json);
		if (this.options.breakOnFirstError && validateResult === false) break;
	}
}
function anyOfValidator(report, schema, json) {
	const subReports = [];
	for (let i = 0; i < schema.anyOf.length; i++) {
		const subReport = new Report(report);
		subReports.push(subReport);
		this._jsonValidate(subReport, schema.anyOf[i], json);
		cacheValidationResult(report, schema.anyOf[i], json, subReport.errors.length === 0);
	}
	deferOrRunSync(report, subReports, () => {
		let passed = false;
		for (const subReport of subReports) if (subReport.errors.length === 0) {
			passed = true;
			break;
		}
		if (passed === false) report.addError("ANY_OF_MISSING", void 0, subReports, schema, "anyOf");
	});
}
function oneOfValidator(report, schema, json) {
	const subReports = [];
	for (let i = 0; i < schema.oneOf.length; i++) {
		const subReport = new Report(report);
		subReports.push(subReport);
		this._jsonValidate(subReport, schema.oneOf[i], json);
		cacheValidationResult(report, schema.oneOf[i], json, subReport.errors.length === 0);
	}
	deferOrRunSync(report, subReports, () => {
		let passes = 0;
		for (const subReport of subReports) if (subReport.errors.length === 0) passes++;
		if (passes === 0) report.addError("ONE_OF_MISSING", void 0, subReports, schema, "oneOf");
		else if (passes > 1) report.addError("ONE_OF_MULTIPLE", void 0, void 0, schema, "oneOf");
	});
}
function notValidator(report, schema, json) {
	const subReport = new Report(report);
	if (this._jsonValidate(subReport, schema.not, json) === true) report.addError("NOT_PASSED", void 0, void 0, schema, "not");
}
function ifValidator(report, schema, json) {
	if (this.options.version === "draft-04" || this.options.version === "draft-06") return;
	const conditionSchema = schema.if;
	const thenSchema = schema.then;
	const elseSchema = schema.else;
	if (conditionSchema === void 0 || thenSchema === void 0 && elseSchema === void 0) return;
	const conditionReport = new Report(report);
	this._jsonValidate(conditionReport, conditionSchema, json);
	cacheValidationResult(report, conditionSchema, json, conditionReport.errors.length === 0);
	const branchSchema = conditionReport.errors.length === 0 ? thenSchema : elseSchema;
	if (branchSchema === void 0) return;
	this._jsonValidate(report, branchSchema, json);
}
function thenValidator() {}
function elseValidator() {}
//#endregion
//#region src/validation/numeric.ts
function multipleOfValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["MULTIPLE_OF"])) return;
	if (typeof json !== "number") return;
	const result = json / schema.multipleOf;
	if (!Number.isFinite(result) || Math.abs(result - Math.round(result)) >= 1e-10) report.addError("MULTIPLE_OF", [json, schema.multipleOf], void 0, schema, "multipleOf");
}
function maximumValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["MAXIMUM", "MAXIMUM_EXCLUSIVE"])) return;
	if (typeof json !== "number") return;
	if (schema.exclusiveMaximum !== true) {
		if (json > schema.maximum) report.addError("MAXIMUM", [json, schema.maximum], void 0, schema, "maximum");
	} else if (json >= schema.maximum) report.addError("MAXIMUM_EXCLUSIVE", [json, schema.maximum], void 0, schema, "maximum");
}
function exclusiveMaximumValidator(report, schema, json) {
	if (typeof schema.exclusiveMaximum === "number") {
		if (shouldSkipValidate(this.validateOptions, ["MAXIMUM_EXCLUSIVE"])) return;
		if (typeof json !== "number") return;
		if (json >= schema.exclusiveMaximum) report.addError("MAXIMUM_EXCLUSIVE", [json, schema.exclusiveMaximum], void 0, schema, "exclusiveMaximum");
	}
}
function minimumValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["MINIMUM", "MINIMUM_EXCLUSIVE"])) return;
	if (typeof json !== "number") return;
	if (schema.exclusiveMinimum !== true) {
		if (json < schema.minimum) report.addError("MINIMUM", [json, schema.minimum], void 0, schema, "minimum");
	} else if (json <= schema.minimum) report.addError("MINIMUM_EXCLUSIVE", [json, schema.minimum], void 0, schema, "minimum");
}
function exclusiveMinimumValidator(report, schema, json) {
	if (typeof schema.exclusiveMinimum === "number") {
		if (shouldSkipValidate(this.validateOptions, ["MINIMUM_EXCLUSIVE"])) return;
		if (typeof json !== "number") return;
		if (json <= schema.exclusiveMinimum) report.addError("MINIMUM_EXCLUSIVE", [json, schema.exclusiveMinimum], void 0, schema, "exclusiveMinimum");
	}
}
//#endregion
//#region src/validation/object.ts
function maxPropertiesValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["OBJECT_PROPERTIES_MAXIMUM"])) return;
	if (!isObject(json)) return;
	const keysCount = Object.keys(json).length;
	if (keysCount > schema.maxProperties) report.addError("OBJECT_PROPERTIES_MAXIMUM", [keysCount, schema.maxProperties], void 0, schema, "maxProperties");
}
function minPropertiesValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["OBJECT_PROPERTIES_MINIMUM"])) return;
	if (!isObject(json)) return;
	const keysCount = Object.keys(json).length;
	if (keysCount < schema.minProperties) report.addError("OBJECT_PROPERTIES_MINIMUM", [keysCount, schema.minProperties], void 0, schema, "minProperties");
}
function requiredValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["OBJECT_MISSING_REQUIRED_PROPERTY"])) return;
	if (!isObject(json)) return;
	const idx = schema.required.length;
	for (let i = 0; i < idx; i++) {
		const requiredPropertyName = schema.required[i];
		if (!Object.hasOwn(json, requiredPropertyName)) report.addError("OBJECT_MISSING_REQUIRED_PROPERTY", [requiredPropertyName], void 0, schema, "required");
	}
}
function additionalPropertiesValidator(report, schema, json) {
	if (schema.properties === void 0 && schema.patternProperties === void 0) return propertiesValidator.call(this, report, schema, json);
}
function patternPropertiesValidator(report, schema, json) {
	if (schema.properties === void 0) return propertiesValidator.call(this, report, schema, json);
}
function propertiesValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["OBJECT_ADDITIONAL_PROPERTIES"])) return;
	if (!isObject(json)) return;
	const properties = schema.properties !== void 0 ? schema.properties : {};
	const patternProperties = schema.patternProperties !== void 0 ? schema.patternProperties : {};
	if (schema.additionalProperties === false) {
		let s = Object.keys(json);
		const p = Object.keys(properties);
		const pp = Object.keys(patternProperties);
		s = difference(s, p);
		for (const ppKey of pp) {
			const result = compileSchemaRegex(ppKey);
			if (!result.ok) continue;
			const regExp = result.value;
			for (let idx2 = s.length - 1; idx2 >= 0; idx2--) if (regExp.test(s[idx2]) === true) s.splice(idx2, 1);
		}
		if (s.length > 0) {
			if (Array.isArray(this.options.assumeAdditional)) for (const allowed of this.options.assumeAdditional) {
				const io = s.indexOf(allowed);
				if (io !== -1) s.splice(io, 1);
			}
			if (s.length > 0) for (const extra of s) report.addError("OBJECT_ADDITIONAL_PROPERTIES", [extra], void 0, schema, "properties");
		}
	}
}
function dependenciesValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["OBJECT_DEPENDENCY_KEY"])) return;
	if (!isObject(json)) return;
	const keys = Object.keys(schema.dependencies);
	for (const dependencyName of keys) if (Object.hasOwn(json, dependencyName)) {
		const dependencyDefinition = schema.dependencies[dependencyName];
		if (Array.isArray(dependencyDefinition)) {
			for (const requiredPropertyName of dependencyDefinition) if (!Object.hasOwn(json, requiredPropertyName)) report.addError("OBJECT_DEPENDENCY_KEY", [requiredPropertyName, dependencyName], void 0, schema, "dependencies");
		} else this._jsonValidate(report, dependencyDefinition, json);
	}
}
function dependentSchemasValidator(report, schema, json) {
	if (!supportsDependentKeywords(schema, this.options.version)) return;
	if (!isObject(json) || !isObject(schema.dependentSchemas)) return;
	const keys = Object.keys(schema.dependentSchemas);
	for (const dependencyName of keys) if (Object.hasOwn(json, dependencyName)) {
		const dependencySchema = schema.dependentSchemas[dependencyName];
		this._jsonValidate(report, dependencySchema, json);
	}
}
function dependentRequiredValidator(report, schema, json) {
	if (!supportsDependentKeywords(schema, this.options.version)) return;
	if (shouldSkipValidate(this.validateOptions, ["OBJECT_DEPENDENCY_KEY"])) return;
	if (!isObject(json) || !isObject(schema.dependentRequired)) return;
	const keys = Object.keys(schema.dependentRequired);
	for (const dependencyName of keys) {
		if (!Object.hasOwn(json, dependencyName)) continue;
		const requiredProperties = schema.dependentRequired[dependencyName];
		if (!Array.isArray(requiredProperties)) continue;
		for (const requiredPropertyName of requiredProperties) if (!Object.hasOwn(json, requiredPropertyName)) report.addError("OBJECT_DEPENDENCY_KEY", [requiredPropertyName, dependencyName], void 0, schema, "dependentRequired");
	}
}
function propertyNamesValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["PROPERTY_NAMES"])) return;
	if (!isObject(json)) return;
	const propertyNamesSchema = schema.propertyNames;
	if (propertyNamesSchema === void 0) return;
	const Report_ = report.constructor;
	const keys = Object.keys(json);
	const subReports = [];
	for (const key of keys) {
		const subReport = new Report_(report);
		subReports.push(subReport);
		this._jsonValidate(subReport, propertyNamesSchema, key);
	}
	const addPropertyNameErrors = () => {
		for (let idx = 0; idx < keys.length; idx++) if (subReports[idx].errors.length > 0) report.addError("PROPERTY_NAMES", [keys[idx]], subReports[idx], schema, void 0);
	};
	deferOrRunSync(report, subReports, addPropertyNameErrors);
}
//#endregion
//#region src/validation/ref.ts
const getDynamicRefAnchorName = (dynamicRef) => {
	const hashIdx = dynamicRef.indexOf("#");
	if (hashIdx === -1) return;
	const fragment = dynamicRef.slice(hashIdx + 1);
	if (!fragment || fragment[0] === "/") return;
	return fragment;
};
const findDynamicAnchorInScope = (scopeSchema, anchorName) => {
	const scopeId = getId(scopeSchema);
	const scopeBaseUri = scopeId ? getRemotePath(scopeId) : void 0;
	const found = findId(scopeSchema, anchorName, scopeBaseUri, scopeBaseUri);
	if (found && found.$dynamicAnchor === anchorName) return found;
};
/**
* Resolves the effective target for a $recursiveRef, walking the recursive anchor stack.
*/
const resolveRecursiveRef = (schema, recursiveAnchorStack) => {
	const resolved = schema.__$recursiveRefResolved;
	if (!resolved) return;
	let target = resolved;
	if (typeof target === "object" && target.$recursiveAnchor === true) {
		const dynamicTarget = recursiveAnchorStack[0];
		if (dynamicTarget) target = dynamicTarget;
	}
	return target;
};
/**
* Resolves the effective target for a $dynamicRef, walking the dynamic scope stack.
*/
const resolveDynamicRef = (schema, dynamicScopeStack) => {
	const resolved = schema.__$dynamicRefResolved;
	if (typeof resolved === "undefined" || !schema.$dynamicRef) return resolved;
	let target = resolved;
	const anchorName = getDynamicRefAnchorName(schema.$dynamicRef);
	if (anchorName && typeof target === "object" && target.$dynamicAnchor === anchorName) for (let scopeIdx = 0; scopeIdx < dynamicScopeStack.length; scopeIdx++) {
		const scopeSchema = dynamicScopeStack[scopeIdx];
		const scopedTarget = findDynamicAnchorInScope(scopeSchema, anchorName);
		if (scopedTarget) {
			target = scopedTarget;
			break;
		}
	}
	return target;
};
//#endregion
//#region src/utils/base64.ts
const base64Pattern = /^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/;
const isValidBase64 = (value) => {
	if (value.length % 4 !== 0) return false;
	return base64Pattern.test(value);
};
const decodeBase64 = (value) => {
	if (!isValidBase64(value)) return;
	if (typeof atob === "function") try {
		return atob(value);
	} catch {
		return;
	}
	if (typeof Buffer !== "undefined") try {
		return Buffer.from(value, "base64").toString("utf8");
	} catch {
		return;
	}
};
//#endregion
//#region src/utils/unicode.ts
/**
* Returns the number of Unicode code points in the string.
* Uses the built-in string iterator which correctly handles surrogate pairs.
*/
function unicodeLength(str) {
	let count = 0;
	for (const _cp of str) count++;
	return count;
}
//#endregion
//#region src/validation/string.ts
function minLengthValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["MIN_LENGTH"])) return;
	if (typeof json !== "string") return;
	if (unicodeLength(json) < schema.minLength) report.addError("MIN_LENGTH", [json.length, schema.minLength], void 0, schema, "minLength");
}
function maxLengthValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["MAX_LENGTH"])) return;
	if (typeof json !== "string") return;
	if (unicodeLength(json) > schema.maxLength) report.addError("MAX_LENGTH", [json.length, schema.maxLength], void 0, schema, "maxLength");
}
function patternValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["PATTERN"])) return;
	if (typeof json !== "string") return;
	const result = compileSchemaRegex(schema.pattern);
	if (!result.ok) {
		report.addError("PATTERN", [
			schema.pattern,
			json,
			result.error.message
		], void 0, schema, "pattern");
		return;
	}
	if (!result.value.test(json)) report.addError("PATTERN", [schema.pattern, json], void 0, schema, "pattern");
}
function formatValidator(report, schema, json) {
	if (this.options.formatAssertions === false) return;
	if (this.options.formatAssertions === true) {
		if (!isFormatAssertionVocabEnabled(schema, report, this.options.version)) return;
	}
	const isModernDraft = this.options.version === "draft2019-09" || this.options.version === "draft2020-12";
	const formatValidatorFn = getFormatValidators(this.options)[schema.format];
	if (typeof formatValidatorFn === "function") {
		if (shouldSkipValidate(this.validateOptions, ["INVALID_FORMAT"])) return;
		if (report.hasError("INVALID_TYPE", [schema.type, whatIs(json)])) return;
		if (formatValidatorFn.length === 2) report.addAsyncTaskWithPath(formatValidatorFn, [json], function(result) {
			if (result !== true) report.addError("INVALID_FORMAT", [schema.format, JSON.stringify(json)], void 0, schema, "format");
		});
		else {
			const result = formatValidatorFn.call(this, json);
			if (result instanceof Promise) {
				const promiseResult = result;
				report.addAsyncTaskWithPath(async (callback) => {
					try {
						callback(await promiseResult);
					} catch (_error) {
						callback(false);
					}
				}, [], function(resolvedResult) {
					if (resolvedResult !== true) report.addError("INVALID_FORMAT", [schema.format, JSON.stringify(json)], void 0, schema, "format");
				});
			} else if (result !== true) report.addError("INVALID_FORMAT", [schema.format, JSON.stringify(json)], void 0, schema, "format");
		}
	} else if (this.options.ignoreUnknownFormats !== true && !isModernDraft) report.addError("UNKNOWN_FORMAT", [schema.format], void 0, schema, "format");
}
function contentEncodingValidator(report, schema, json) {
	if (this.options.version !== "draft-07") return;
	if (typeof json !== "string") return;
	if (schema.contentEncoding !== "base64") return;
	if (!isValidBase64(json)) report.addError("INVALID_FORMAT", ["contentEncoding:base64", JSON.stringify(json)], void 0, schema, "contentEncoding");
}
function contentMediaTypeValidator(report, schema, json) {
	if (this.options.version !== "draft-07") return;
	if (typeof json !== "string") return;
	if (schema.contentMediaType !== "application/json") return;
	let payload = json;
	if (schema.contentEncoding === "base64") {
		const decoded = decodeBase64(json);
		if (decoded === void 0) {
			report.addError("INVALID_FORMAT", ["contentEncoding:base64", JSON.stringify(json)], void 0, schema, "contentEncoding");
			return;
		}
		payload = decoded;
	}
	try {
		JSON.parse(payload);
	} catch {
		report.addError("INVALID_FORMAT", ["contentMediaType:application/json", JSON.stringify(json)], void 0, schema, "contentMediaType");
	}
}
//#endregion
//#region src/validation/type.ts
function typeValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["INVALID_TYPE"])) return;
	const jsonType = whatIs(json);
	if (typeof schema.type === "string") {
		if (jsonType !== schema.type && (jsonType !== "integer" || schema.type !== "number")) report.addError("INVALID_TYPE", [schema.type, jsonType], void 0, schema, "type");
	} else if (!schema.type.includes(jsonType) && (jsonType !== "integer" || !schema.type.includes("number"))) report.addError("INVALID_TYPE", [JSON.stringify(schema.type), jsonType], void 0, schema, "type");
}
function enumValidator(report, schema, json) {
	if (shouldSkipValidate(this.validateOptions, ["ENUM_CASE_MISMATCH", "ENUM_MISMATCH"])) return;
	let match = false, caseInsensitiveMatch = false;
	for (const enumVal of schema.enum) if (areEqual(json, enumVal, { maxDepth: this.options.maxRecursionDepth })) {
		match = true;
		break;
	} else if (areEqual(json, enumVal, {
		caseInsensitiveComparison: true,
		maxDepth: this.options.maxRecursionDepth
	})) caseInsensitiveMatch = true;
	if (match === false) {
		const error = caseInsensitiveMatch && this.options.enumCaseInsensitiveComparison ? "ENUM_CASE_MISMATCH" : "ENUM_MISMATCH";
		report.addError(error, [JSON.stringify(json)], void 0, schema, "enum");
	}
}
function constValidator(report, schema, json) {
	const constValue = schema.const;
	if (areEqual(json, constValue, { maxDepth: this.options.maxRecursionDepth }) === false) report.addError("CONST", [JSON.stringify(constValue)], void 0, schema, void 0);
}
//#endregion
//#region src/json-validation.ts
function collectEvaluated(args) {
	const { report, currentSchema, json, mode, depth } = args;
	if (!currentSchema || typeof currentSchema === "boolean") return /* @__PURE__ */ new Set();
	if (depth > (this.options.maxRecursionDepth ?? 100)) {
		report.addError("COLLECT_EVALUATED_DEPTH_EXCEEDED", [depth]);
		return /* @__PURE__ */ new Set();
	}
	const evaluated = /* @__PURE__ */ new Set();
	const merge = (other) => {
		if (other === "all") return true;
		for (const v of other) evaluated.add(v);
		return false;
	};
	const recurse = (subSchema) => {
		if (mode === "items") return collectEvaluated.call(this, {
			report,
			currentSchema: subSchema,
			json,
			mode: "items",
			jsonArr: args.jsonArr,
			depth: depth + 1
		});
		return collectEvaluated.call(this, {
			report,
			currentSchema: subSchema,
			json,
			mode: "properties",
			jsonData: args.jsonData,
			depth: depth + 1
		});
	};
	if (mode === "items") {
		const jsonArr = args.jsonArr;
		if (Array.isArray(currentSchema.prefixItems)) {
			const len = Math.min(currentSchema.prefixItems.length, jsonArr.length);
			for (let i = 0; i < len; i++) evaluated.add(i);
		}
		if (currentSchema.items !== void 0) {
			if (Array.isArray(currentSchema.items)) {
				const len = Math.min(currentSchema.items.length, jsonArr.length);
				for (let i = 0; i < len; i++) evaluated.add(i);
			} else if (currentSchema.items !== false) return "all";
		}
		if (currentSchema.additionalItems !== void 0 && currentSchema.additionalItems !== false && Array.isArray(currentSchema.items)) return "all";
		if (currentSchema.contains !== void 0) for (let i = 0; i < jsonArr.length; i++) {
			let passed = getCachedValidationResult(report, currentSchema.contains, jsonArr[i]);
			if (passed === void 0) {
				const subReport = new Report(report);
				validate.call(this, subReport, currentSchema.contains, jsonArr[i]);
				passed = subReport.errors.length === 0;
			}
			if (passed) evaluated.add(i);
		}
		if (currentSchema.unevaluatedItems === true) return "all";
	} else {
		const jsonData = args.jsonData;
		if (isObject(currentSchema.properties)) {
			for (const key of Object.keys(currentSchema.properties)) if (Object.hasOwn(jsonData, key)) evaluated.add(key);
		}
		if (isObject(currentSchema.patternProperties)) for (const pattern of Object.keys(currentSchema.patternProperties)) {
			const result = compileSchemaRegex(pattern);
			if (result.ok) {
				for (const key of Object.keys(jsonData)) if (result.value.test(key)) evaluated.add(key);
			}
		}
		if (currentSchema.additionalProperties !== void 0) {
			const propKeys = isObject(currentSchema.properties) ? Object.keys(currentSchema.properties) : [];
			const patternRegexes = [];
			if (isObject(currentSchema.patternProperties)) for (const pattern of Object.keys(currentSchema.patternProperties)) {
				const result = compileSchemaRegex(pattern);
				if (result.ok) patternRegexes.push(result.value);
			}
			for (const key of Object.keys(jsonData)) {
				if (propKeys.includes(key)) continue;
				if (patternRegexes.some((re) => re.test(key))) continue;
				evaluated.add(key);
			}
		}
		if (isObject(currentSchema.dependentSchemas)) {
			for (const [depKey, depSchema] of Object.entries(currentSchema.dependentSchemas)) if (Object.hasOwn(jsonData, depKey)) {
				if (merge(recurse(depSchema))) return "all";
			}
		}
		if (currentSchema.unevaluatedProperties === true) return "all";
	}
	if (Array.isArray(currentSchema.allOf)) {
		for (const subSchema of currentSchema.allOf) if (merge(recurse(subSchema))) return "all";
	}
	if (Array.isArray(currentSchema.anyOf)) for (const subSchema of currentSchema.anyOf) {
		let passed = getCachedValidationResult(report, subSchema, json);
		if (passed === void 0) {
			const subReport = new Report(report);
			validate.call(this, subReport, subSchema, json);
			passed = subReport.errors.length === 0;
		}
		if (passed) {
			if (merge(recurse(subSchema))) return "all";
		}
	}
	if (Array.isArray(currentSchema.oneOf)) for (const subSchema of currentSchema.oneOf) {
		let passed = getCachedValidationResult(report, subSchema, json);
		if (passed === void 0) {
			const subReport = new Report(report);
			validate.call(this, subReport, subSchema, json);
			passed = subReport.errors.length === 0;
		}
		if (passed) {
			if (merge(recurse(subSchema))) return "all";
		}
	}
	if (currentSchema.if !== void 0) {
		let condPassed = getCachedValidationResult(report, currentSchema.if, json);
		if (condPassed === void 0) {
			const condReport = new Report(report);
			validate.call(this, condReport, currentSchema.if, json);
			condPassed = condReport.errors.length === 0;
		}
		if (condPassed) {
			if (merge(recurse(currentSchema.if))) return "all";
			if (currentSchema.then !== void 0) {
				if (merge(recurse(currentSchema.then))) return "all";
			}
		} else if (currentSchema.else !== void 0) {
			if (merge(recurse(currentSchema.else))) return "all";
		}
	}
	if (currentSchema.__$refResolved && currentSchema.__$refResolved !== currentSchema) {
		if (merge(recurse(currentSchema.__$refResolved))) return "all";
	}
	const recursiveTarget = resolveRecursiveRef(currentSchema, report.__$recursiveAnchorStack);
	if (recursiveTarget && recursiveTarget !== currentSchema) {
		if (merge(recurse(recursiveTarget))) return "all";
	}
	const dynamicTarget = resolveDynamicRef(currentSchema, report.__$dynamicScopeStack);
	if (dynamicTarget && dynamicTarget !== currentSchema) {
		if (merge(recurse(dynamicTarget))) return "all";
	}
	return evaluated;
}
function unevaluatedItemsValidator(report, schema, json) {
	if (!Array.isArray(json)) return;
	if (schema.unevaluatedItems === true) return;
	const unevalSchema = schema.unevaluatedItems;
	if (unevalSchema === void 0) return;
	if (json.length === 0) return;
	const evaluatedItems = collectEvaluated.call(this, {
		report,
		currentSchema: schema,
		json,
		mode: "items",
		jsonArr: json,
		depth: 0
	});
	if (evaluatedItems === "all") return;
	const unevaluatedIndices = [];
	for (let i = 0; i < json.length; i++) if (!evaluatedItems.has(i)) unevaluatedIndices.push(i);
	if (unevaluatedIndices.length === 0) return;
	if (unevalSchema === false) report.addError("ARRAY_UNEVALUATED_ITEMS", void 0, void 0, schema, "unevaluatedItems");
	else for (const idx of unevaluatedIndices) {
		const subReport = new Report(report);
		validate.call(this, subReport, unevalSchema, json[idx]);
		if (subReport.errors.length > 0) {
			report.addError("ARRAY_UNEVALUATED_ITEMS", void 0, void 0, schema, "unevaluatedItems");
			break;
		}
	}
}
function unevaluatedPropertiesValidator(report, schema, json) {
	if (!isObject(json)) return;
	if (schema.unevaluatedProperties === true) return;
	const unevalSchema = schema.unevaluatedProperties;
	if (unevalSchema === void 0) return;
	const allKeys = Object.keys(json);
	if (allKeys.length === 0) return;
	const evaluatedProperties = collectEvaluated.call(this, {
		report,
		currentSchema: schema,
		json,
		mode: "properties",
		jsonData: json,
		depth: 0
	});
	if (evaluatedProperties === "all") return;
	const unevaluatedKeys = allKeys.filter((key) => !evaluatedProperties.has(key));
	if (unevaluatedKeys.length === 0) return;
	if (unevalSchema === false) report.addError("OBJECT_UNEVALUATED_PROPERTIES", [unevaluatedKeys.join(", ")], void 0, schema, "unevaluatedProperties");
	else for (const key of unevaluatedKeys) {
		const subReport = new Report(report);
		validate.call(this, subReport, unevalSchema, json[key]);
		if (subReport.errors.length > 0) report.addError("OBJECT_UNEVALUATED_PROPERTIES", [key], void 0, schema, "unevaluatedProperties");
	}
}
function definitionsValidator() {}
const JsonValidators = {
	id: () => {},
	$id: () => {},
	$ref: () => {},
	$schema: () => {},
	$dynamicAnchor: () => {},
	$dynamicRef: () => {},
	$anchor: () => {},
	$defs: () => {},
	$vocabulary: () => {},
	$recursiveAnchor: () => {},
	$recursiveRef: () => {},
	examples: () => {},
	title: () => {},
	description: () => {},
	default: () => {},
	type: typeValidator,
	enum: enumValidator,
	const: constValidator,
	multipleOf: multipleOfValidator,
	maximum: maximumValidator,
	exclusiveMaximum: exclusiveMaximumValidator,
	minimum: minimumValidator,
	exclusiveMinimum: exclusiveMinimumValidator,
	maxLength: maxLengthValidator,
	minLength: minLengthValidator,
	pattern: patternValidator,
	format: formatValidator,
	contentEncoding: contentEncodingValidator,
	contentMediaType: contentMediaTypeValidator,
	additionalItems: additionalItemsValidator,
	items: itemsValidator,
	prefixItems: prefixItemsValidator,
	maxItems: maxItemsValidator,
	minItems: minItemsValidator,
	uniqueItems: uniqueItemsValidator,
	contains: containsValidator,
	maxContains: maxContainsValidator,
	minContains: minContainsValidator,
	unevaluatedItems: unevaluatedItemsValidator,
	maxProperties: maxPropertiesValidator,
	minProperties: minPropertiesValidator,
	required: requiredValidator,
	additionalProperties: additionalPropertiesValidator,
	patternProperties: patternPropertiesValidator,
	properties: propertiesValidator,
	dependencies: dependenciesValidator,
	dependentSchemas: dependentSchemasValidator,
	dependentRequired: dependentRequiredValidator,
	propertyNames: propertyNamesValidator,
	unevaluatedProperties: unevaluatedPropertiesValidator,
	allOf: allOfValidator,
	anyOf: anyOfValidator,
	oneOf: oneOfValidator,
	not: notValidator,
	if: ifValidator,
	then: thenValidator,
	else: elseValidator,
	definitions: definitionsValidator
};
const recurseArray = function(report, schema, json) {
	const schemaUri = typeof schema.$schema === "string" ? schema.$schema : void 0;
	const prefixItems = (schemaUri === "https://json-schema.org/draft/2020-12/schema" || !schemaUri && this.options.version === "draft2020-12") && Array.isArray(schema.prefixItems) ? schema.prefixItems : void 0;
	if (prefixItems) {
		for (let idx = 0; idx < json.length; idx++) if (idx < prefixItems.length) {
			report.path.push(idx);
			validate.call(this, report, prefixItems[idx], json[idx]);
			report.path.pop();
		} else if (schema.items !== void 0 && !Array.isArray(schema.items)) {
			report.path.push(idx);
			report.schemaPath.push("items");
			validate.call(this, report, schema.items, json[idx]);
			report.schemaPath.pop();
			report.path.pop();
		}
		return;
	}
	if (Array.isArray(schema.items)) {
		for (let idx = 0; idx < json.length; idx++) if (idx < schema.items.length) {
			report.path.push(idx);
			validate.call(this, report, schema.items[idx], json[idx]);
			report.path.pop();
		} else if (typeof schema.additionalItems === "object") {
			report.path.push(idx);
			validate.call(this, report, schema.additionalItems, json[idx]);
			report.path.pop();
		}
	} else if (typeof schema.items === "object" || typeof schema.items === "boolean") for (let idx = 0; idx < json.length; idx++) {
		report.path.push(idx);
		report.schemaPath.push("items");
		validate.call(this, report, schema.items, json[idx]);
		report.schemaPath.pop();
		report.path.pop();
	}
};
const recurseObject = function(report, schema, json) {
	let additionalProperties = schema.additionalProperties;
	if (additionalProperties === true || additionalProperties === void 0) additionalProperties = {};
	const p = schema.properties ? Object.keys(schema.properties) : [];
	const pp = schema.patternProperties ? Object.keys(schema.patternProperties) : [];
	const keys = Object.keys(json);
	for (const m of keys) {
		const propertyValue = json[m];
		const s = [];
		if (p.includes(m)) s.push(schema.properties[m]);
		for (const regexString of pp) {
			const result = compileSchemaRegex(regexString);
			if (result.ok && result.value.test(m) === true) s.push(schema.patternProperties[regexString]);
		}
		if (s.length === 0 && additionalProperties !== false) s.push(additionalProperties);
		for (const schema_s of s) {
			report.path.push(m);
			if (p.includes(m)) {
				report.schemaPath.push("properties");
				report.schemaPath.push(m);
			} else report.schemaPath.push("additionalProperties");
			validate.call(this, report, schema_s, propertyValue);
			report.path.pop();
			report.schemaPath.pop();
			if (p.includes(m)) report.schemaPath.pop();
		}
	}
};
function validate(report, schema, json) {
	report.commonErrorMessage = "JSON_OBJECT_VALIDATION_FAILED";
	if (schema === true) return true;
	if (schema === false) {
		report.addError("SCHEMA_IS_FALSE", [], void 0, schema);
		return false;
	}
	if (!isObject(schema)) {
		report.addError("SCHEMA_NOT_AN_OBJECT", [whatIs(schema)], void 0, schema);
		return false;
	}
	let keys = Object.keys(schema);
	if (keys.length === 0) return true;
	let isRoot = false;
	if (!report.rootSchema) {
		report.rootSchema = schema;
		isRoot = true;
	}
	const recursiveAnchorStack = report.__$recursiveAnchorStack;
	const dynamicScopeStack = report.__$dynamicScopeStack;
	let pushedRecursiveAnchor = false;
	let pushedDynamicScope = false;
	const schemaId = getId(schema);
	const dynamicScopeEntry = schema.__$resourceRoot || (isRoot || typeof schemaId === "string" ? schema : void 0);
	if (dynamicScopeEntry && dynamicScopeStack[dynamicScopeStack.length - 1] !== dynamicScopeEntry) {
		dynamicScopeStack.push(dynamicScopeEntry);
		pushedDynamicScope = true;
	}
	if (schema.$recursiveAnchor === true) {
		recursiveAnchorStack.push(schema);
		pushedRecursiveAnchor = true;
	}
	if (schema.$ref !== void 0) if (this.options.version === "draft2019-09" || this.options.version === "draft2020-12") {
		if (!schema.__$refResolved) report.addError("REF_UNRESOLVED", [schema.$ref], void 0, schema);
		else validate.call(this, report, schema.__$refResolved, json);
		keys = keys.filter((key) => key !== "$ref");
	} else {
		let maxRefs = 99;
		while (schema.$ref && maxRefs > 0) {
			if (!schema.__$refResolved) {
				report.addError("REF_UNRESOLVED", [schema.$ref], void 0, schema);
				break;
			} else if (schema.__$refResolved === schema) break;
			else {
				schema = schema.__$refResolved;
				keys = Object.keys(schema);
			}
			maxRefs--;
		}
		if (maxRefs === 0) throw new Error("Circular dependency by $ref references!");
		report.schemaPath = [];
	}
	if (schema.$recursiveRef !== void 0) {
		if (this.options.version === "draft2019-09" || this.options.version === "draft2020-12") {
			const recursiveRefTarget = resolveRecursiveRef(schema, recursiveAnchorStack);
			if (!recursiveRefTarget) report.addError("REF_UNRESOLVED", [schema.$recursiveRef], void 0, schema);
			else validate.call(this, report, recursiveRefTarget, json);
			keys = keys.filter((key) => key !== "$recursiveRef");
		}
	}
	if (schema.$dynamicRef !== void 0) {
		if (this.options.version === "draft2020-12") {
			const dynamicRefTarget = resolveDynamicRef(schema, dynamicScopeStack);
			if (typeof dynamicRefTarget === "undefined") report.addError("REF_UNRESOLVED", [schema.$dynamicRef], void 0, schema);
			else validate.call(this, report, dynamicRefTarget, json);
			keys = keys.filter((key) => key !== "$dynamicRef");
		}
	}
	const validationVocabularyEnabled = isValidationVocabularyEnabled(schema, report, this.options.version);
	if (!validationVocabularyEnabled) keys = keys.filter((key) => !VALIDATION_VOCAB_KEYWORDS.has(key));
	if (validationVocabularyEnabled && schema.type) {
		keys.splice(keys.indexOf("type"), 1);
		report.schemaPath.push("type");
		JsonValidators.type.call(this, report, schema, json);
		report.schemaPath.pop();
		if (report.errors.length && this.options.breakOnFirstError) {
			if (pushedRecursiveAnchor) recursiveAnchorStack.pop();
			if (pushedDynamicScope) dynamicScopeStack.pop();
			return false;
		}
	}
	const deferredUnevaluatedKeys = [];
	for (const key of keys) {
		if (key === "unevaluatedItems" || key === "unevaluatedProperties") {
			deferredUnevaluatedKeys.push(key);
			continue;
		}
		const validator = JsonValidators[key];
		if (validator) {
			validator.call(this, report, schema, json);
			if (report.errors.length && this.options.breakOnFirstError) break;
		}
	}
	if (deferredUnevaluatedKeys.length > 0 && !(report.errors.length > 0 && this.options.breakOnFirstError)) for (const key of deferredUnevaluatedKeys) {
		const validator = JsonValidators[key];
		if (validator) {
			validator.call(this, report, schema, json);
			if (report.errors.length && this.options.breakOnFirstError) break;
		}
	}
	if (report.errors.length === 0 || this.options.breakOnFirstError === false) {
		if (Array.isArray(json)) recurseArray.call(this, report, schema, json);
		else if (isObject(json)) recurseObject.call(this, report, schema, json);
	}
	if (typeof this.options.customValidator === "function") this.options.customValidator.call(this, report, schema, json);
	if (pushedRecursiveAnchor) recursiveAnchorStack.pop();
	if (pushedDynamicScope) dynamicScopeStack.pop();
	if (isRoot) report.rootSchema = void 0;
	return report.errors.length === 0;
}
//#endregion
//#region src/z-schema-reader.ts
let _schemaReader;
function getSchemaReader() {
	return _schemaReader;
}
function setSchemaReader(schemaReader) {
	_schemaReader = schemaReader;
}
//#endregion
//#region src/schema-compiler.ts
const UNSAFE_TARGETS = [
	Object.prototype,
	Function.prototype,
	Array.prototype
];
/** Returns true if `obj` is a built-in prototype that must not be mutated. */
function isUnsafeTarget(obj) {
	return UNSAFE_TARGETS.includes(obj);
}
/** Safely assign a property on `obj`, refusing prototype-polluting keys. */
function safeSetProperty(obj, key, value) {
	if (isUnsafeTarget(obj)) return;
	/** Reject property names that could pollute Object.prototype (CWE-1321). */
	if (key !== "__proto__" && key !== "constructor" && key !== "prototype") obj[key] = value;
}
/** Safely delete a property from `obj`, refusing to mutate built-in prototypes (CWE-1321). */
function safeDeleteProperty(obj, key) {
	if (isUnsafeTarget(obj)) return;
	if (key !== "__proto__" && key !== "constructor" && key !== "prototype") delete obj[key];
}
const collectIds = (obj, maxDepth = 100) => {
	const ids = [];
	function walk(node, scope, _depth = 0) {
		if (typeof node !== "object" || node == null) return;
		if (_depth >= maxDepth) throw new Error(`Maximum recursion depth (${maxDepth}) exceeded in collectIds. If your schema is deeply nested and valid, increase the maxRecursionDepth option.`);
		let addedScope = false;
		const nodeId = getId(node);
		if (typeof nodeId === "string") {
			let type = isAbsoluteUri(nodeId) ? "absolute" : "relative";
			if (scope.length === 0) type = "root";
			const id = {
				id: nodeId,
				type,
				obj: node
			};
			if (type === "absolute" || type === "root" && isAbsoluteUri(nodeId)) id.absoluteUri = nodeId;
			else if (type === "root" && typeof node.id === "string" && isAbsoluteUri(node.id) && node.id !== nodeId) id.absoluteUri = resolveSchemaScopeId(node.id, node, nodeId);
			else if (type === "relative") {
				id.absoluteParent = scope.filter((x) => x.type === "absolute" || x.type === "root" && x.absoluteUri).slice(-1)[0];
				if (id.absoluteParent) id.absoluteUri = resolveSchemaScopeId(id.absoluteParent.absoluteUri || id.absoluteParent.id, node, id.id);
			}
			ids.push(id);
			scope.push(id);
			addedScope = true;
		}
		if (Array.isArray(node)) for (const item of node) walk(item, scope, _depth + 1);
		else for (const key of Object.keys(node)) {
			if (isInternalKey(key) || NON_SCHEMA_KEYWORDS.includes(key)) continue;
			walk(node[key], scope, _depth + 1);
		}
		if (addedScope) scope.pop();
	}
	walk(obj, []);
	return ids;
};
const collectReferences = (obj, results, scope, path, options, maxDepth = 100, _depth = 0) => {
	results = results || [];
	scope = scope || [];
	path = path || [];
	options = options || {};
	if (typeof obj !== "object" || obj === null) return results;
	if (_depth >= maxDepth) throw new Error(`Maximum recursion depth (${maxDepth}) exceeded in collectReferences. If your schema is deeply nested and valid, increase the maxRecursionDepth option.`);
	const hasRef = typeof obj.$ref === "string" && typeof obj.__$refResolved === "undefined";
	let addedScope = false;
	const isRootScope = scope.length === 0;
	let scopeId = getId(obj);
	if (typeof obj.id === "string" && isAbsoluteUri(obj.id) && (!scopeId || !isAbsoluteUri(scopeId))) scopeId = obj.id;
	if (typeof scopeId === "string" && (isRootScope || !hasRef || options.useRefObjectScope === true)) {
		const base = scope.length > 0 ? scope[scope.length - 1] : void 0;
		scope.push(resolveSchemaScopeId(base, obj, scopeId));
		addedScope = true;
	}
	if (hasRef) results.push({
		ref: resolveReference(scope[scope.length - 1], obj.$ref),
		key: "$ref",
		obj,
		path: path.slice(0)
	});
	if (typeof obj.$recursiveRef === "string" && typeof obj.__$recursiveRefResolved === "undefined") results.push({
		ref: resolveReference(scope[scope.length - 1], obj.$recursiveRef),
		key: "$recursiveRef",
		obj,
		path: path.slice(0)
	});
	if (typeof obj.$dynamicRef === "string" && typeof obj.__$dynamicRefResolved === "undefined") results.push({
		ref: resolveReference(scope[scope.length - 1], obj.$dynamicRef),
		key: "$dynamicRef",
		obj,
		path: path.slice(0)
	});
	if (typeof obj.$schema === "string" && typeof obj.__$schemaResolved === "undefined") results.push({
		ref: resolveReference(scope[scope.length - 1], obj.$schema),
		key: "$schema",
		obj,
		path: path.slice(0)
	});
	if (Array.isArray(obj)) for (let i = 0; i < obj.length; i++) {
		path.push(i);
		collectReferences(obj[i], results, scope, path, options, maxDepth, _depth + 1);
		path.pop();
	}
	else {
		const keys = Object.keys(obj);
		for (const key of keys) {
			if (isInternalKey(key) || NON_SCHEMA_KEYWORDS.includes(key)) continue;
			path.push(key);
			collectReferences(obj[key], results, scope, path, options, maxDepth, _depth + 1);
			path.pop();
		}
	}
	if (addedScope) scope.pop();
	return results;
};
const resolveReference = (base, ref) => {
	if (isAbsoluteUri(ref)) return ref;
	const baseStr = base ?? "";
	if (ref[0] === "#") {
		const hashIndex = baseStr.indexOf("#");
		return (hashIndex === -1 ? baseStr : baseStr.slice(0, hashIndex)) + ref;
	}
	if (!baseStr) return ref;
	const hashIndex = baseStr.indexOf("#");
	const baseNoFrag = hashIndex === -1 ? baseStr : baseStr.slice(0, hashIndex);
	if (isAbsoluteUri(baseNoFrag)) try {
		return new URL(ref, baseNoFrag).toString();
	} catch {}
	let baseDir = baseNoFrag;
	if (!baseDir.endsWith("/")) {
		const lastSlash = baseDir.lastIndexOf("/");
		baseDir = lastSlash === -1 ? "" : baseDir.slice(0, lastSlash + 1);
	}
	return baseDir + ref;
};
const isSimpleIdentifier = (id) => id[0] !== "#" && !id.includes("/") && !id.includes(".") && !id.includes("#");
const resolveIdScope = (base, id) => {
	if (isAbsoluteUri(id)) return id;
	const baseStr = base ?? "";
	if (isSimpleIdentifier(id)) {
		const hashIndex = baseStr.indexOf("#");
		return (hashIndex === -1 ? baseStr : baseStr.slice(0, hashIndex)) + "#" + id;
	}
	return resolveReference(base, id);
};
const resolveSchemaScopeId = (base, schema, id) => {
	if (typeof schema.$id === "string") return resolveReference(base, id);
	return resolveIdScope(base, id);
};
var SchemaCompiler = class {
	constructor(validator) {
		this.validator = validator;
	}
	collectAndCacheIds(schema) {
		const ids = collectIds(schema, this.validator.options.maxRecursionDepth);
		for (const item of ids) if (item.absoluteUri) {
			this.validator.scache.cacheSchemaByUri(item.absoluteUri, item.obj);
			if (item.type === "relative" && item.absoluteParent && isSimpleIdentifier(item.id)) {
				const altAbsoluteUri = resolveReference(item.absoluteParent.absoluteUri || item.absoluteParent.id, item.id);
				this.validator.scache.cacheSchemaByUri(altAbsoluteUri, item.obj);
			}
		} else if (item.type === "root") this.validator.scache.cacheSchemaByUri(item.id, item.obj);
	}
	compileSchema(report, schema, options) {
		report.commonErrorMessage = "SCHEMA_COMPILATION_FAILED";
		if (typeof schema === "string") {
			const loadedSchema = this.validator.scache.getSchemaByUri(report, schema);
			if (typeof loadedSchema === "undefined") {
				report.addError("SCHEMA_NOT_REACHABLE", [schema]);
				return false;
			}
			schema = loadedSchema;
		}
		if (Array.isArray(schema)) {
			if (!options?.noCache) schema.forEach((s) => this.collectAndCacheIds(s));
			return this.compileArrayOfSchemas(report, schema);
		} else if (typeof schema === "boolean") return true;
		else if (!options?.noCache) this.collectAndCacheIds(schema);
		const canMutateSchemaObject = schema !== Object.prototype && schema !== Function.prototype && schema !== Array.prototype;
		if (canMutateSchemaObject && schema.__$compiled && schema.id && this.validator.scache.checkCacheForUri(schema.id) === false) schema.__$compiled = void 0;
		if (schema.__$compiled) return true;
		if (canMutateSchemaObject && !schema.$schema && this.validator.options.version !== "none") schema.$schema = this.validator.getDefaultSchemaId();
		if (schema.id && typeof schema.id === "string" && !options?.noCache) this.validator.scache.cacheSchemaByUri(schema.id, schema);
		let isRoot = false;
		if (!report.rootSchema) {
			report.rootSchema = schema;
			isRoot = true;
		}
		const isValidExceptReferences = report.isValid();
		if (canMutateSchemaObject) safeDeleteProperty(schema, "__$missingReferences");
		const useRefObjectScope = this.validator.options.version === "draft2019-09" || this.validator.options.version === "draft2020-12";
		const refs = collectReferences(schema, void 0, void 0, void 0, { useRefObjectScope }, this.validator.options.maxRecursionDepth);
		for (const refObj of refs) {
			let response = this.validator.scache.getSchemaByUri(report, refObj.ref, schema);
			if (typeof response === "undefined") {
				const schemaReader = getSchemaReader();
				if (schemaReader) {
					const remotePath = getRemotePath(refObj.ref);
					const s = schemaReader(remotePath);
					if (s) {
						s.id = remotePath;
						const subreport = new Report(report);
						if (!this.compileSchema(subreport, s)) report.errors = report.errors.concat(subreport.errors);
						else response = this.validator.scache.getSchemaByUri(report, refObj.ref, schema);
					}
				}
			}
			if (typeof response === "undefined") {
				const hasNotValid = report.hasError("REMOTE_NOT_VALID", [refObj.ref]);
				const isAbsolute = isAbsoluteUri(refObj.ref);
				let isDownloaded = false;
				const ignoreUnresolvableRemotes = this.validator.options.ignoreUnresolvableReferences === true;
				if (isAbsolute) isDownloaded = this.validator.scache.checkCacheForUri(refObj.ref);
				if (hasNotValid) {} else if (ignoreUnresolvableRemotes && isAbsolute) {} else if (isDownloaded) {} else {
					report.path.push(...refObj.path);
					report.addError("UNRESOLVABLE_REFERENCE", [refObj.ref]);
					report.path = report.path.slice(0, -refObj.path.length);
					if (isValidExceptReferences && canMutateSchemaObject && !isUnsafeTarget(schema)) {
						schema.__$missingReferences = schema.__$missingReferences || [];
						schema.__$missingReferences.push(refObj);
					}
				}
			}
			safeSetProperty(refObj.obj, `__${refObj.key}Resolved`, response);
		}
		const isValid = report.isValid();
		if (isValid && canMutateSchemaObject) schema.__$compiled = true;
		if (isRoot) report.rootSchema = void 0;
		return isValid;
	}
	compileArrayOfSchemas(report, arr) {
		let compiled = 0, lastLoopCompiled;
		do {
			report.errors = report.errors.filter((e) => e.code !== "UNRESOLVABLE_REFERENCE");
			lastLoopCompiled = compiled;
			compiled = this.compileArrayOfSchemasLoop(report, arr);
			for (const sch of arr) if (sch.__$missingReferences) {
				for (let idx2 = sch.__$missingReferences.length - 1; idx2 >= 0; idx2--) {
					const refObj = sch.__$missingReferences[idx2];
					const response = arr.find((x) => x.id === refObj.ref);
					if (response) {
						safeSetProperty(refObj.obj, `__${refObj.key}Resolved`, response);
						sch.__$missingReferences.splice(idx2, 1);
					}
				}
				if (sch.__$missingReferences.length === 0) safeDeleteProperty(sch, "__$missingReferences");
			}
		} while (compiled !== arr.length && compiled !== lastLoopCompiled);
		return report.isValid();
	}
	compileArrayOfSchemasLoop(mainReport, arr) {
		let compiledCount = 0;
		for (const schema of arr) {
			const report = new Report(mainReport);
			if (this.compileSchema(report, schema)) compiledCount++;
			mainReport.errors = mainReport.errors.concat(report.errors);
		}
		return compiledCount;
	}
};
//#endregion
//#region src/schema-validator.ts
const SchemaValidators = {
	$ref: function(report, schema) {
		if (typeof schema.$ref !== "string") report.addError("KEYWORD_TYPE_EXPECTED", ["$ref", "string"], void 0, schema, "$ref");
	},
	$schema: function(report, schema) {
		if (typeof schema.$schema !== "string") report.addError("KEYWORD_TYPE_EXPECTED", ["$schema", "string"], void 0, schema, "$schema");
	},
	multipleOf: function(report, schema) {
		if (typeof schema.multipleOf !== "number") report.addError("KEYWORD_TYPE_EXPECTED", ["multipleOf", "number"], void 0, schema, "multipleOf");
		else if (schema.multipleOf <= 0) report.addError("KEYWORD_MUST_BE", ["multipleOf", "strictly greater than 0"], void 0, schema, "multipleOf");
	},
	maximum: function(report, schema) {
		if (typeof schema.maximum !== "number") report.addError("KEYWORD_TYPE_EXPECTED", ["maximum", "number"], void 0, schema, "maximum");
	},
	exclusiveMaximum: function(report, schema) {
		if (report.options.version === "draft-04") {
			if (typeof schema.exclusiveMaximum !== "boolean") report.addError("KEYWORD_TYPE_EXPECTED", ["exclusiveMaximum", "boolean"], void 0, schema, "exclusiveMaximum");
			else if (schema.maximum === void 0) report.addError("KEYWORD_DEPENDENCY", ["exclusiveMaximum", "maximum"], void 0, schema, "exclusiveMaximum");
		} else if (typeof schema.exclusiveMaximum !== "boolean" && typeof schema.exclusiveMaximum !== "number") report.addError("KEYWORD_TYPE_EXPECTED", ["exclusiveMaximum", ["boolean", "number"]], void 0, schema, "exclusiveMaximum");
	},
	minimum: function(report, schema) {
		if (typeof schema.minimum !== "number") report.addError("KEYWORD_TYPE_EXPECTED", ["minimum", "number"], void 0, schema, "minimum");
	},
	exclusiveMinimum: function(report, schema) {
		if (report.options.version === "draft-04") {
			if (typeof schema.exclusiveMinimum !== "boolean") report.addError("KEYWORD_TYPE_EXPECTED", ["exclusiveMinimum", "boolean"], void 0, schema, "exclusiveMinimum");
			else if (schema.minimum === void 0) report.addError("KEYWORD_DEPENDENCY", ["exclusiveMinimum", "minimum"], void 0, schema, "exclusiveMinimum");
		} else if (typeof schema.exclusiveMinimum !== "boolean" && typeof schema.exclusiveMinimum !== "number") report.addError("KEYWORD_TYPE_EXPECTED", ["exclusiveMinimum", ["boolean", "number"]], void 0, schema, "exclusiveMinimum");
	},
	maxLength: function(report, schema) {
		if (!isInteger(schema.maxLength)) report.addError("KEYWORD_TYPE_EXPECTED", ["maxLength", "integer"], void 0, schema, "maxLength");
		else if (schema.maxLength < 0) report.addError("KEYWORD_MUST_BE", ["maxLength", "greater than, or equal to 0"], void 0, schema, "maxLength");
	},
	minLength: function(report, schema) {
		if (!isInteger(schema.minLength)) report.addError("KEYWORD_TYPE_EXPECTED", ["minLength", "integer"], void 0, schema, "minLength");
		else if (schema.minLength < 0) report.addError("KEYWORD_MUST_BE", ["minLength", "greater than, or equal to 0"], void 0, schema, "minLength");
	},
	pattern: function(report, schema) {
		if (typeof schema.pattern !== "string") report.addError("KEYWORD_TYPE_EXPECTED", ["pattern", "string"], void 0, schema, "pattern");
		else {
			const result = compileSchemaRegex(schema.pattern);
			if (!result.ok) report.addError("KEYWORD_PATTERN", [
				"pattern",
				schema.pattern,
				result.error.message
			], void 0, schema, "pattern");
		}
	},
	additionalItems: function(report, schema) {
		if (typeof schema.additionalItems !== "boolean" && !isObject(schema.additionalItems)) report.addError("KEYWORD_TYPE_EXPECTED", ["additionalItems", ["boolean", "object"]], void 0, schema, "additionalItems");
		else if (isObject(schema.additionalItems)) {
			report.path.push("additionalItems");
			this.validateSchema(report, schema.additionalItems);
			report.path.pop();
		}
	},
	items: function(report, schema) {
		if (Array.isArray(schema.items)) for (let idx = 0; idx < schema.items.length; idx++) {
			report.path.push("items");
			report.path.push(idx);
			this.validateSchema(report, schema.items[idx]);
			report.path.pop();
			report.path.pop();
		}
		else if (isObject(schema.items) || report.options.version !== "draft-04" && typeof schema.items === "boolean") {
			report.path.push("items");
			this.validateSchema(report, schema.items);
			report.path.pop();
		} else report.addError("KEYWORD_TYPE_EXPECTED", ["items", report.options.version === "draft-04" ? ["array", "object"] : [
			"array",
			"object",
			"boolean"
		]], void 0, schema, "items");
		if (this.options.forceAdditional === true && schema.additionalItems === void 0 && Array.isArray(schema.items)) report.addError("KEYWORD_UNDEFINED_STRICT", ["additionalItems"], void 0, schema, "additionalItems");
		if (this.options.assumeAdditional && schema.additionalItems === void 0 && Array.isArray(schema.items)) schema.additionalItems = false;
	},
	maxItems: function(report, schema) {
		if (typeof schema.maxItems !== "number") report.addError("KEYWORD_TYPE_EXPECTED", ["maxItems", "integer"], void 0, schema, "maxItems");
		else if (schema.maxItems < 0) report.addError("KEYWORD_MUST_BE", ["maxItems", "greater than, or equal to 0"], void 0, schema, "maxItems");
	},
	minItems: function(report, schema) {
		if (!isInteger(schema.minItems)) report.addError("KEYWORD_TYPE_EXPECTED", ["minItems", "integer"], void 0, schema, "minItems");
		else if (schema.minItems < 0) report.addError("KEYWORD_MUST_BE", ["minItems", "greater than, or equal to 0"], void 0, schema, "minItems");
	},
	uniqueItems: function(report, schema) {
		if (typeof schema.uniqueItems !== "boolean") report.addError("KEYWORD_TYPE_EXPECTED", ["uniqueItems", "boolean"], void 0, schema, "uniqueItems");
	},
	maxProperties: function(report, schema) {
		if (!isInteger(schema.maxProperties)) report.addError("KEYWORD_TYPE_EXPECTED", ["maxProperties", "integer"], void 0, schema, "maxProperties");
		else if (schema.maxProperties < 0) report.addError("KEYWORD_MUST_BE", ["maxProperties", "greater than, or equal to 0"], void 0, schema, "maxProperties");
	},
	minProperties: function(report, schema) {
		if (!isInteger(schema.minProperties)) report.addError("KEYWORD_TYPE_EXPECTED", ["minProperties", "integer"], void 0, schema, "minProperties");
		else if (schema.minProperties < 0) report.addError("KEYWORD_MUST_BE", ["minProperties", "greater than, or equal to 0"], void 0, schema, "minProperties");
	},
	required: function(report, schema) {
		if (!Array.isArray(schema.required)) report.addError("KEYWORD_TYPE_EXPECTED", ["required", "array"], void 0, schema, "required");
		else if (report.options.version === "draft-04" && schema.required.length === 0) report.addError("KEYWORD_MUST_BE", ["required", "an array with at least one element"], void 0, schema, "required");
		else {
			for (const item of schema.required) if (typeof item !== "string") report.addError("KEYWORD_VALUE_TYPE", ["required", "string"], void 0, schema, "required");
			if (isUniqueArray(schema.required) === false) report.addError("KEYWORD_MUST_BE", ["required", "an array with unique items"], void 0, schema, "required");
		}
	},
	additionalProperties: function(report, schema) {
		if (typeof schema.additionalProperties !== "boolean" && !isObject(schema.additionalProperties)) report.addError("KEYWORD_TYPE_EXPECTED", ["additionalProperties", ["boolean", "object"]], void 0, schema, "additionalProperties");
		else if (isObject(schema.additionalProperties)) {
			report.path.push("additionalProperties");
			this.validateSchema(report, schema.additionalProperties);
			report.path.pop();
		}
	},
	properties: function(report, schema) {
		if (!isObject(schema.properties)) {
			report.addError("KEYWORD_TYPE_EXPECTED", ["properties", "object"], void 0, schema, "properties");
			return;
		}
		const keys = Object.keys(schema.properties);
		for (const key of keys) {
			const val = schema.properties[key];
			report.path.push("properties");
			report.path.push(key);
			this.validateSchema(report, val);
			report.path.pop();
			report.path.pop();
		}
		if (this.options.forceAdditional === true && schema.additionalProperties === void 0) report.addError("KEYWORD_UNDEFINED_STRICT", ["additionalProperties"], void 0, schema, "additionalProperties");
		if (this.options.assumeAdditional && schema.additionalProperties === void 0) schema.additionalProperties = false;
		if (this.options.forceProperties === true && keys.length === 0) report.addError("CUSTOM_MODE_FORCE_PROPERTIES", ["properties"], void 0, schema, "properties");
	},
	patternProperties: function(report, schema) {
		if (!isObject(schema.patternProperties)) {
			report.addError("KEYWORD_TYPE_EXPECTED", ["patternProperties", "object"], void 0, schema, "patternProperties");
			return;
		}
		const keys = Object.keys(schema.patternProperties);
		for (const key of keys) {
			const val = schema.patternProperties[key];
			const result = compileSchemaRegex(key);
			if (!result.ok) report.addError("KEYWORD_PATTERN", [
				"patternProperties",
				key,
				result.error.message
			], void 0, schema, "patternProperties");
			report.path.push("patternProperties");
			report.path.push(key);
			this.validateSchema(report, val);
			report.path.pop();
			report.path.pop();
		}
		if (this.options.forceProperties === true && keys.length === 0) report.addError("CUSTOM_MODE_FORCE_PROPERTIES", ["patternProperties"], void 0, schema, "patternProperties");
	},
	dependencies: function(report, schema) {
		if (!isObject(schema.dependencies)) report.addError("KEYWORD_TYPE_EXPECTED", ["dependencies", "object"], void 0, schema, "dependencies");
		else {
			const keys = Object.keys(schema.dependencies);
			for (const schemaKey of keys) {
				const schemaDependency = schema.dependencies[schemaKey];
				if (isObject(schemaDependency) || report.options.version !== "draft-04" && typeof schemaDependency === "boolean") {
					report.path.push("dependencies");
					report.path.push(schemaKey);
					this.validateSchema(report, schemaDependency);
					report.path.pop();
					report.path.pop();
				} else if (Array.isArray(schemaDependency)) {
					const depArray = schemaDependency;
					if (report.options.version === "draft-04" && depArray.length === 0) report.addError("KEYWORD_MUST_BE", ["dependencies", "not empty array"], void 0, schema, "dependencies");
					for (const dep of depArray) if (typeof dep !== "string") report.addError("KEYWORD_VALUE_TYPE", ["dependencies", "string"], void 0, schema, "dependencies");
					if (isUniqueArray(depArray) === false) report.addError("KEYWORD_MUST_BE", ["dependencies", "an array with unique items"], void 0, schema, "dependencies");
				} else report.addError("KEYWORD_VALUE_TYPE", ["dependencies", report.options.version === "draft-04" ? "object or array" : "boolean, object or array"], void 0, schema, "dependencies");
			}
		}
	},
	enum: function(report, schema) {
		if (Array.isArray(schema.enum) === false) report.addError("KEYWORD_TYPE_EXPECTED", ["enum", "array"], void 0, schema, "enum");
		else if (schema.enum.length === 0) report.addError("KEYWORD_MUST_BE", ["enum", "an array with at least one element"], void 0, schema, "enum");
		else if (isUniqueArray(schema.enum) === false) report.addError("KEYWORD_MUST_BE", ["enum", "an array with unique elements"], void 0, schema, "enum");
	},
	type: function(report, schema) {
		const primitiveTypes = [
			"array",
			"boolean",
			"integer",
			"number",
			"null",
			"object",
			"string"
		];
		const primitiveTypeStr = primitiveTypes.join(",");
		const isArray = Array.isArray(schema.type);
		if (Array.isArray(schema.type)) {
			for (const typeItem of schema.type) if (!primitiveTypes.includes(typeItem)) report.addError("KEYWORD_TYPE_EXPECTED", ["type", primitiveTypeStr], void 0, schema, "type");
			if (isUniqueArray(schema.type) === false) report.addError("KEYWORD_MUST_BE", ["type", "an object with unique properties"], void 0, schema, "type");
		} else if (typeof schema.type === "string") {
			if (!primitiveTypes.includes(schema.type)) report.addError("KEYWORD_TYPE_EXPECTED", ["type", primitiveTypeStr], void 0, schema, "type");
		} else report.addError("KEYWORD_TYPE_EXPECTED", ["type", ["string", "array"]], void 0, schema, "type");
		if (this.options.noEmptyStrings === true) {
			if (schema.type === "string" || isArray && schema.type.includes("string")) {
				if (schema.minLength === void 0 && schema.enum === void 0 && schema.format === void 0) schema.minLength = 1;
			}
		}
		if (this.options.noEmptyArrays === true) {
			if (schema.type === "array" || isArray && schema.type.includes("array")) {
				if (schema.minItems === void 0) schema.minItems = 1;
			}
		}
		if (this.options.forceProperties === true) {
			if (schema.type === "object" || isArray && schema.type.includes("object")) {
				if (schema.properties === void 0 && schema.patternProperties === void 0) report.addError("KEYWORD_UNDEFINED_STRICT", ["properties"], void 0, schema, "properties");
			}
		}
		if (this.options.forceItems === true) {
			if (schema.type === "array" || isArray && schema.type.includes("array")) {
				if (schema.items === void 0) report.addError("KEYWORD_UNDEFINED_STRICT", ["items"], void 0, schema, "items");
			}
		}
		if (this.options.forceMinItems === true) {
			if (schema.type === "array" || isArray && schema.type.includes("array")) {
				if (schema.minItems === void 0) report.addError("KEYWORD_UNDEFINED_STRICT", ["minItems"], void 0, schema, "minItems");
			}
		}
		if (this.options.forceMaxItems === true) {
			if (schema.type === "array" || isArray && schema.type.includes("array")) {
				if (schema.maxItems === void 0) report.addError("KEYWORD_UNDEFINED_STRICT", ["maxItems"], void 0, schema, "maxItems");
			}
		}
		if (this.options.forceMinLength === true) {
			if (schema.type === "string" || isArray && schema.type.includes("string")) {
				if (schema.minLength === void 0 && schema.format === void 0 && schema.enum === void 0 && schema.pattern === void 0) report.addError("KEYWORD_UNDEFINED_STRICT", ["minLength"], void 0, schema, "minLength");
			}
		}
		if (this.options.forceMaxLength === true) {
			if (schema.type === "string" || isArray && schema.type.includes("string")) {
				if (schema.maxLength === void 0 && schema.format === void 0 && schema.enum === void 0 && schema.pattern === void 0) report.addError("KEYWORD_UNDEFINED_STRICT", ["maxLength"], void 0, schema, "maxLength");
			}
		}
	},
	allOf: function(report, schema) {
		if (Array.isArray(schema.allOf) === false) report.addError("KEYWORD_TYPE_EXPECTED", ["allOf", "array"], void 0, schema, "allOf");
		else if (schema.allOf.length === 0) report.addError("KEYWORD_MUST_BE", ["allOf", "an array with at least one element"], void 0, schema, "allOf");
		else for (let idx = 0; idx < schema.allOf.length; idx++) {
			report.path.push("allOf");
			report.path.push(idx);
			this.validateSchema(report, schema.allOf[idx]);
			report.path.pop();
			report.path.pop();
		}
	},
	anyOf: function(report, schema) {
		if (Array.isArray(schema.anyOf) === false) report.addError("KEYWORD_TYPE_EXPECTED", ["anyOf", "array"], void 0, schema, "anyOf");
		else if (schema.anyOf.length === 0) report.addError("KEYWORD_MUST_BE", ["anyOf", "an array with at least one element"], void 0, schema, "anyOf");
		else for (let idx = 0; idx < schema.anyOf.length; idx++) {
			report.path.push("anyOf");
			report.path.push(idx);
			this.validateSchema(report, schema.anyOf[idx]);
			report.path.pop();
			report.path.pop();
		}
	},
	oneOf: function(report, schema) {
		if (Array.isArray(schema.oneOf) === false) report.addError("KEYWORD_TYPE_EXPECTED", ["oneOf", "array"], void 0, schema, "oneOf");
		else if (schema.oneOf.length === 0) report.addError("KEYWORD_MUST_BE", ["oneOf", "an array with at least one element"], void 0, schema, "oneOf");
		else for (let idx = 0; idx < schema.oneOf.length; idx++) {
			report.path.push("oneOf");
			report.path.push(idx);
			this.validateSchema(report, schema.oneOf[idx]);
			report.path.pop();
			report.path.pop();
		}
	},
	not: function(report, schema) {
		const notSchema = schema.not;
		if (!(report.options.version === "draft-04" ? isObject(notSchema) : typeof notSchema === "boolean" || isObject(notSchema))) report.addError("KEYWORD_TYPE_EXPECTED", ["not", report.options.version === "draft-04" ? "object" : ["boolean", "object"]], void 0, schema, "not");
		else {
			report.path.push("not");
			this.validateSchema(report, notSchema);
			report.path.pop();
		}
	},
	if: function(report, schema) {
		if (report.options.version !== "draft-07") return;
		const ifSchema = schema.if;
		if (!(typeof ifSchema === "boolean" || isObject(ifSchema))) {
			report.addError("KEYWORD_TYPE_EXPECTED", ["if", ["boolean", "object"]], void 0, schema, "if");
			return;
		}
		report.path.push("if");
		this.validateSchema(report, ifSchema);
		report.path.pop();
	},
	then: function(report, schema) {
		if (report.options.version !== "draft-07") return;
		const thenSchema = schema.then;
		if (!(typeof thenSchema === "boolean" || isObject(thenSchema))) {
			report.addError("KEYWORD_TYPE_EXPECTED", ["then", ["boolean", "object"]], void 0, schema, "then");
			return;
		}
		report.path.push("then");
		this.validateSchema(report, thenSchema);
		report.path.pop();
	},
	else: function(report, schema) {
		if (report.options.version !== "draft-07") return;
		const elseSchema = schema.else;
		if (!(typeof elseSchema === "boolean" || isObject(elseSchema))) {
			report.addError("KEYWORD_TYPE_EXPECTED", ["else", ["boolean", "object"]], void 0, schema, "else");
			return;
		}
		report.path.push("else");
		this.validateSchema(report, elseSchema);
		report.path.pop();
	},
	definitions: function(report, schema) {
		if (!isObject(schema.definitions)) report.addError("KEYWORD_TYPE_EXPECTED", ["definitions", "object"], void 0, schema, "definitions");
		else {
			const keys = Object.keys(schema.definitions);
			for (const key of keys) {
				const val = schema.definitions[key];
				report.path.push("definitions");
				report.path.push(key);
				this.validateSchema(report, val);
				report.path.pop();
				report.path.pop();
			}
		}
	},
	$defs: function(report, schema) {
		if (report.options.version !== "draft2019-09" && report.options.version !== "draft2020-12") return;
		if (!isObject(schema.$defs)) {
			report.addError("KEYWORD_TYPE_EXPECTED", ["$defs", "object"], void 0, schema, "$defs");
			return;
		}
		const keys = Object.keys(schema.$defs);
		for (const key of keys) {
			const val = schema.$defs[key];
			report.path.push("$defs");
			report.path.push(key);
			this.validateSchema(report, val);
			report.path.pop();
			report.path.pop();
		}
	},
	format: function(report, schema) {
		if (this.options.formatAssertions === false) return;
		if (typeof schema.format !== "string") report.addError("KEYWORD_TYPE_EXPECTED", ["format", "string"], void 0, schema, "format");
		else {
			const isModernDraft = this.options.version === "draft2019-09" || this.options.version === "draft2020-12";
			if (!isFormatSupported(schema.format, this.options.customFormats) && this.options.ignoreUnknownFormats !== true && !isModernDraft) report.addError("UNKNOWN_FORMAT", [schema.format], void 0, schema, "format");
		}
	},
	contentEncoding: function(report, schema) {
		if (report.options.version !== "draft-07") return;
		if (typeof schema.contentEncoding !== "string") report.addError("KEYWORD_TYPE_EXPECTED", ["contentEncoding", "string"], void 0, schema, "contentEncoding");
	},
	contentMediaType: function(report, schema) {
		if (report.options.version !== "draft-07") return;
		if (typeof schema.contentMediaType !== "string") report.addError("KEYWORD_TYPE_EXPECTED", ["contentMediaType", "string"], void 0, schema, "contentMediaType");
	},
	id: function(report, schema) {
		if (typeof schema.id !== "string") report.addError("KEYWORD_TYPE_EXPECTED", ["id", "string"], void 0, schema, "id");
	},
	title: function(report, schema) {
		if (typeof schema.title !== "string") report.addError("KEYWORD_TYPE_EXPECTED", ["title", "string"], void 0, schema, "title");
	},
	description: function(report, schema) {
		if (typeof schema.description !== "string") report.addError("KEYWORD_TYPE_EXPECTED", ["description", "string"], void 0, schema, "description");
	},
	default: function() {}
};
var SchemaValidator = class {
	constructor(validator) {
		this.validator = validator;
	}
	get options() {
		return this.validator.options;
	}
	validateArrayOfSchemas(report, arr) {
		for (const schema of arr) this.validateSchema(report, schema);
		return report.isValid();
	}
	validateSchema(report, schema) {
		report.commonErrorMessage = "SCHEMA_VALIDATION_FAILED";
		if (Array.isArray(schema)) return this.validateArrayOfSchemas(report, schema);
		if (typeof schema === "boolean") return true;
		if (schema.__$validated) return true;
		const hasParentSchema = schema.$schema && getId(schema) !== schema.$schema;
		if (hasParentSchema) {
			if (schema.__$schemaResolved && schema.__$schemaResolved !== schema) {
				const subReport = new Report(report);
				if (validate.call(this.validator, subReport, schema.__$schemaResolved, schema) === false) report.addError("PARENT_SCHEMA_VALIDATION_FAILED", void 0, subReport, schema, "$schema");
			} else if (this.validator.options.ignoreUnresolvableReferences !== true) report.addError("REF_UNRESOLVED", [schema.$schema], void 0, schema, "$schema");
		}
		if (this.validator.options.noTypeless === true) {
			if (schema.type !== void 0) {
				let schemas = [];
				if (Array.isArray(schema.anyOf)) schemas = schemas.concat(schema.anyOf);
				if (Array.isArray(schema.oneOf)) schemas = schemas.concat(schema.oneOf);
				if (Array.isArray(schema.allOf)) schemas = schemas.concat(schema.allOf);
				schemas.forEach(function(sch) {
					if (!sch.type) sch.type = schema.type;
				});
			}
			if (schema.enum === void 0 && schema.type === void 0 && schema.anyOf === void 0 && schema.oneOf === void 0 && schema.not === void 0 && schema.$ref === void 0) report.addError("KEYWORD_UNDEFINED_STRICT", ["type"], void 0, schema, "type");
		}
		const keys = Object.keys(schema);
		for (const key of keys) {
			if (key.startsWith("__")) continue;
			if (Object.hasOwn(SchemaValidators, key)) SchemaValidators[key].call(this, report, schema);
			else if (!hasParentSchema) {
				if (this.validator.options.noExtraKeywords === true) report.addError("KEYWORD_UNEXPECTED", [key], void 0, schema, void 0);
			}
		}
		if (this.validator.options.pedanticCheck === true) {
			if (schema.enum) {
				const tmpSchema = shallowClone(schema);
				delete tmpSchema.enum;
				delete tmpSchema.default;
				report.path.push("enum");
				for (let idx = 0; idx < schema.enum.length; idx++) {
					report.path.push(idx);
					validate.call(this.validator, report, tmpSchema, schema.enum[idx]);
					report.path.pop();
				}
				report.path.pop();
			}
			if (schema.default) {
				report.path.push("default");
				validate.call(this.validator, report, schema, schema.default);
				report.path.pop();
			}
		}
		const isValid = report.isValid();
		if (isValid) schema.__$validated = true;
		return isValid;
	}
};
//#endregion
//#region src/z-schema-base.ts
/**
* Module-private symbol used by `ZSchema.create()` to authorise construction.
* Not exported — external code cannot instantiate ZSchema variants directly.
*/
const FACTORY_TOKEN = Symbol("ZSchema.factory");
var ZSchemaBase = class {
	scache;
	sc;
	sv;
	validateOptions = {};
	options;
	constructor(options, token) {
		if (token !== FACTORY_TOKEN) throw new Error("do not use new ZSchema(), use ZSchema.create() instead");
		this.scache = new SchemaCache(this);
		this.sc = new SchemaCompiler(this);
		this.sv = new SchemaValidator(this);
		this.options = normalizeOptions(options);
	}
	/**
	* Internal recursive JSON validation — delegates to the `validate` function
	* in `json-validation.ts`. Exposed as a method so that per-keyword validator
	* modules (array, combinators, object) can call back into the core validator
	* via `this` without importing `json-validation.ts` directly (which would
	* create a circular dependency).
	*/
	_jsonValidate(report, schema, json) {
		return validate.call(this, report, schema, json);
	}
	getDefaultSchemaId() {
		return this.options.version && this.options.version !== "none" ? VERSION_SCHEMA_URL_MAPPING[this.options.version] : VERSION_SCHEMA_URL_MAPPING[defaultOptions.version];
	}
	_validate(json, schema, options, callback) {
		if (typeof options === "function") {
			callback = options;
			options = {};
		}
		if (!options) options = {};
		this.validateOptions = options;
		if (typeof schema !== "string" && typeof schema !== "boolean" && !isObject(schema)) {
			const e = /* @__PURE__ */ new Error("Invalid .validate call - schema must be a string or object but " + whatIs(schema) + " was passed!");
			if (callback) {
				setTimeout(function() {
					callback(e, false);
				}, 0);
				return;
			}
			throw e;
		}
		let foundError = false;
		const report = new Report(this.options, options);
		report.json = json;
		let _schema;
		if (typeof schema === "string") {
			const schemaName = schema;
			_schema = this.scache.getSchema(report, schemaName);
			if (!_schema) {
				const e = /* @__PURE__ */ new Error("Schema with id '" + schemaName + "' wasn't found in the validator cache!");
				if (callback) {
					setTimeout(function() {
						callback(e, false);
					}, 0);
					return;
				}
				throw e;
			}
		} else _schema = this.scache.getSchema(report, schema);
		let compiled = false;
		if (!foundError) compiled = this.sc.compileSchema(report, _schema);
		if (!compiled) foundError = true;
		let validated = false;
		if (!foundError) validated = this.sv.validateSchema(report, _schema);
		if (!validated) foundError = true;
		if (options.schemaPath) {
			report.rootSchema = _schema;
			_schema = get(_schema, options.schemaPath);
			if (!_schema) {
				const e = /* @__PURE__ */ new Error("Schema path '" + options.schemaPath + "' wasn't found in the schema!");
				if (callback) {
					setTimeout(function() {
						callback(e, false);
					}, 0);
					return;
				}
				throw e;
			}
		}
		if (!foundError) validate.call(this, report, _schema, json);
		if (callback) {
			report.processAsyncTasks(this.options.asyncTimeout, callback);
			return;
		} else if (report.asyncTasks.length > 0) throw new Error("This validation has async tasks and cannot be done in sync mode, please provide callback argument.");
		if (!report.isValid()) throw getValidateError({
			message: report.commonErrorMessage,
			details: report.errors
		});
		return true;
	}
	_validateSchema(schemaOrArr) {
		if (Array.isArray(schemaOrArr) && schemaOrArr.length === 0) throw new Error(".compileSchema was called with an empty array");
		const report = new Report(this.options);
		if (Array.isArray(schemaOrArr)) {
			const arr = this.scache.getSchema(report, schemaOrArr);
			if (this.sc.compileSchema(report, arr)) this.sv.validateSchema(report, arr);
		} else {
			const schema = this.scache.getSchema(report, schemaOrArr);
			if (this.sc.compileSchema(report, schema)) this.sv.validateSchema(report, schema);
		}
		if (!report.isValid()) throw getValidateError({
			message: report.commonErrorMessage,
			details: report.errors
		});
		return true;
	}
	/**
	* Register a format validator on this instance only (does not affect other instances or the global registry).
	* @param name - The format name.
	* @param validatorFunction - A sync or async function `(value: unknown) => boolean | Promise<boolean>`.
	*/
	registerFormat(name, validatorFunction) {
		if (!this.options.customFormats) this.options.customFormats = {};
		this.options.customFormats[name] = validatorFunction;
	}
	/**
	* Unregister an instance-scoped format validator.
	* @param name - The format name to unregister.
	*/
	unregisterFormat(name) {
		if (!this.options.customFormats) this.options.customFormats = {};
		this.options.customFormats[name] = null;
	}
	/** Returns the names of format validators registered on this instance. */
	getRegisteredFormats() {
		return sortedKeys(this.options.customFormats || {}).filter((key) => this.options.customFormats?.[key] != null);
	}
	/** Returns all supported format names (global + instance-registered). */
	getSupportedFormats() {
		return getSupportedFormats(this.options.customFormats);
	}
	/**
	* Register a remote schema in this instance's cache so `$ref` can resolve to it.
	* @param uri - The URI the schema will be known by.
	* @param schema - The schema object or JSON string.
	* @param validationOptions - Optional options used for schema preparation.
	*/
	setRemoteReference(uri, schema, validationOptions) {
		const _schema = prepareRemoteSchema(schema, uri, validationOptions, this.options.maxRecursionDepth);
		this.scache.cacheSchemaByUri(uri, _schema);
	}
	/**
	* Extract unresolvable `$ref` URIs from a validation error.
	* @param err - A `ValidateError` from a failed validation.
	* @returns An array of unresolvable reference URIs.
	*/
	getMissingReferences(err) {
		if (!err) return [];
		const details = err.details || [];
		const missingRefs = [];
		function collect(details) {
			for (const detail of details) {
				if (detail.code === "UNRESOLVABLE_REFERENCE" || detail.code === "SCHEMA_NOT_REACHABLE") missingRefs.push(detail.params[0]);
				if (detail.inner) collect(detail.inner);
			}
		}
		collect(details);
		return missingRefs;
	}
	/**
	* Extract unresolvable **remote** `$ref` URIs from a validation error (local fragment-only refs are excluded).
	* @param err - A `ValidateError` from a failed validation.
	* @returns An array of remote reference base URIs.
	*/
	getMissingRemoteReferences(err) {
		const missingReferences = this.getMissingReferences(err);
		const missingRemoteReferences = [];
		for (const ref of missingReferences) {
			const remoteReference = getRemotePath(ref);
			if (remoteReference && !missingRemoteReferences.includes(remoteReference)) missingRemoteReferences.push(remoteReference);
		}
		return missingRemoteReferences;
	}
	/**
	* Resolve a previously compiled schema by its `$id` / `id`, cleaning up internal bookkeeping properties
	* and inlining resolved `$ref` targets.
	* @param schemaId - The schema identifier to look up.
	* @returns A clean, resolved copy of the schema, or `undefined` if not found.
	*/
	getResolvedSchema(schemaId) {
		const report = new Report(this.options);
		const schema = this.scache.getSchemaByUri(report, schemaId);
		if (!schema) return void 0;
		const clonedSchema = deepClone(schema, this.options.maxRecursionDepth);
		const visited = /* @__PURE__ */ new WeakSet();
		const cleanup = function(schema) {
			let key;
			const typeOf = whatIs(schema);
			if (typeOf !== "object" && typeOf !== "array") return;
			if (visited.has(schema)) return;
			visited.add(schema);
			if (schema.$ref && schema.__$refResolved) {
				const from = schema.__$refResolved;
				const to = schema;
				delete schema.$ref;
				delete schema.__$refResolved;
				for (key in from) copyProp(from, to, key);
			}
			for (key in schema) if (Object.hasOwn(schema, key)) if (isInternalKey(key)) delete schema[key];
			else cleanup(schema[key]);
		};
		cleanup(clonedSchema);
		return clonedSchema;
	}
};
//#endregion
//#region src/z-schema.ts
var ZSchema = class ZSchema extends ZSchemaBase {
	/** @internal Use ZSchema.create() instead. */
	constructor(options, token) {
		super(options, token);
	}
	/**
	* Register a global format validator available to all instances.
	* @param name - The format name (e.g. `'email'`, `'date'`).
	* @param validatorFunction - A sync or async function `(value: unknown) => boolean | Promise<boolean>`.
	*/
	static registerFormat(name, validatorFunction) {
		return registerFormat(name, validatorFunction);
	}
	/**
	* Remove a globally registered format validator.
	* @param name - The format name to unregister.
	*/
	static unregisterFormat(name) {
		return unregisterFormat(name);
	}
	/** Returns the names of all globally registered format validators. */
	static getRegisteredFormats() {
		return getRegisteredFormats();
	}
	/** Returns a deep clone of the default options. */
	static getDefaultOptions() {
		return deepClone(defaultOptions);
	}
	/**
	* Register a remote schema in the global cache so any instance can resolve `$ref` to it.
	* @param uri - The URI the schema will be known by.
	* @param schema - The schema object or JSON string.
	* @param validationOptions - Optional options used for schema preparation.
	*/
	static setRemoteReference(uri, schema, validationOptions) {
		const _schema = prepareRemoteSchema(schema, uri, validationOptions);
		SchemaCache.cacheSchemaByUri(uri, _schema);
	}
	/** Returns the current global schema reader, or `undefined` if none is set. */
	static getSchemaReader() {
		return getSchemaReader();
	}
	/**
	* Set a global schema reader function used to resolve remote `$ref` URIs.
	* @param schemaReader - A function `(uri: string) => JsonSchema | undefined`, or `undefined` to clear.
	*/
	static setSchemaReader(schemaReader) {
		return setSchemaReader(schemaReader);
	}
	static schemaSymbol = schemaSymbol;
	static jsonSymbol = jsonSymbol;
	static create(options = {}) {
		const isAsync = options.async;
		const isSafe = options.safe;
		delete options.async;
		delete options.safe;
		if (isAsync && isSafe) return new ZSchemaAsyncSafe(options, FACTORY_TOKEN);
		if (isAsync) return new ZSchemaAsync(options, FACTORY_TOKEN);
		if (isSafe) return new ZSchemaSafe(options, FACTORY_TOKEN);
		return new ZSchema(options, FACTORY_TOKEN);
	}
	/**
	* Validate JSON data against a schema.
	* @param json - The data to validate.
	* @param schema - A JSON Schema object or a schema id string (previously registered via `validateSchema`).
	* @param options - Per-call options (`schemaPath`, `includeErrors`, `excludeErrors`).
	* @returns `true` if valid.
	* @throws {@link ValidateError} if validation fails, with a `details` array of structured errors.
	*/
	validate(json, schema, options = {}) {
		return this._validate(json, schema, options);
	}
	/**
	* Validate JSON data against a schema, returning a result object instead of throwing.
	* @param json - The data to validate.
	* @param schema - A JSON Schema object or a schema id string.
	* @param options - Per-call options.
	* @returns `{ valid: true }` on success, or `{ valid: false, err: ValidateError }` on failure.
	*/
	validateSafe(json, schema, options) {
		try {
			this._validate(json, schema, options ?? {});
			return { valid: true };
		} catch (err) {
			return {
				valid: false,
				err
			};
		}
	}
	/**
	* Validate JSON data against a schema asynchronously (supports async format validators).
	* @param json - The data to validate.
	* @param schema - A JSON Schema object or a schema id string.
	* @param options - Per-call options.
	* @returns A promise that resolves to `true` if valid.
	* @throws {@link ValidateError} if validation fails (the promise rejects).
	*/
	validateAsync(json, schema, options) {
		return new Promise((resolve, reject) => {
			try {
				this._validate(json, schema, options || {}, (err, valid) => err || valid !== true ? reject(err) : resolve(valid));
			} catch (err) {
				reject(err);
			}
		});
	}
	/**
	* Validate JSON data against a schema asynchronously, returning a result object.
	* The promise always resolves (never rejects).
	* @param json - The data to validate.
	* @param schema - A JSON Schema object or a schema id string.
	* @param options - Per-call options.
	* @returns A promise resolving to `{ valid: true }` or `{ valid: false, err: ValidateError }`.
	*/
	validateAsyncSafe(json, schema, options) {
		return new Promise((resolve) => {
			try {
				this._validate(json, schema, options || {}, (err, valid) => {
					resolve({
						valid,
						err
					});
				});
			} catch (err) {
				resolve({
					valid: false,
					err
				});
			}
		});
	}
	/**
	* Validate one or more JSON Schemas, compiling and caching them for later use with `validate()`.
	* @param schemaOrArr - A single schema or an array of schemas (for cross-referencing).
	* @returns `true` if all schemas are valid.
	* @throws {@link ValidateError} if any schema is invalid.
	*/
	validateSchema(schemaOrArr) {
		return this._validateSchema(schemaOrArr);
	}
	/**
	* Validate one or more JSON Schemas, returning a result object instead of throwing.
	* @param schemaOrArr - A single schema or an array of schemas.
	* @returns `{ valid: true }` on success, or `{ valid: false, err: ValidateError }` on failure.
	*/
	validateSchemaSafe(schemaOrArr) {
		try {
			this._validateSchema(schemaOrArr);
			return { valid: true };
		} catch (err) {
			return {
				valid: false,
				err
			};
		}
	}
};
/**
* Synchronous safe validator — `validate()` returns `{ valid, err? }` instead of throwing.
* Created via `ZSchema.create({ safe: true })`.
*/
var ZSchemaSafe = class extends ZSchemaBase {
	/** @internal Use ZSchema.create() instead. */
	constructor(options, token) {
		super(options, token);
	}
	/**
	* Validate JSON data against a schema.
	* @param json - The data to validate.
	* @param schema - A JSON Schema object or a schema id string.
	* @param options - Per-call options.
	* @returns `{ valid: true }` on success, or `{ valid: false, err: ValidateError }` on failure.
	*/
	validate(json, schema, options = {}) {
		try {
			this._validate(json, schema, options);
			return { valid: true };
		} catch (err) {
			return {
				valid: false,
				err
			};
		}
	}
	/**
	* Validate one or more JSON Schemas.
	* @param schemaOrArr - A single schema or an array of schemas.
	* @returns `{ valid: true }` on success, or `{ valid: false, err: ValidateError }` on failure.
	*/
	validateSchema(schemaOrArr) {
		try {
			this._validateSchema(schemaOrArr);
			return { valid: true };
		} catch (err) {
			return {
				valid: false,
				err
			};
		}
	}
};
/**
* Asynchronous throw validator — `validate()` returns `Promise<true>` or rejects.
* Created via `ZSchema.create({ async: true })`.
*/
var ZSchemaAsync = class extends ZSchemaBase {
	/** @internal Use ZSchema.create() instead. */
	constructor(options, token) {
		super(options, token);
	}
	/**
	* Validate JSON data against a schema asynchronously.
	* @param json - The data to validate.
	* @param schema - A JSON Schema object or a schema id string.
	* @param options - Per-call options.
	* @returns A promise that resolves to `true` if valid.
	* @throws {@link ValidateError} if validation fails (the promise rejects).
	*/
	validate(json, schema, options = {}) {
		return new Promise((resolve, reject) => {
			try {
				this._validate(json, schema, options, (err, valid) => err || valid !== true ? reject(err) : resolve(valid));
			} catch (err) {
				reject(err);
			}
		});
	}
	/**
	* Validate one or more JSON Schemas (synchronous, throws on error).
	* @param schemaOrArr - A single schema or an array of schemas.
	* @returns `true` if all schemas are valid.
	* @throws {@link ValidateError} if any schema is invalid.
	*/
	validateSchema(schemaOrArr) {
		return this._validateSchema(schemaOrArr);
	}
};
/**
* Asynchronous safe validator — `validate()` returns `Promise<{ valid, err? }>` (never rejects).
* Created via `ZSchema.create({ async: true, safe: true })`.
*/
var ZSchemaAsyncSafe = class extends ZSchemaBase {
	/** @internal Use ZSchema.create() instead. */
	constructor(options, token) {
		super(options, token);
	}
	/**
	* Validate JSON data against a schema asynchronously.
	* The promise always resolves (never rejects).
	* @param json - The data to validate.
	* @param schema - A JSON Schema object or a schema id string.
	* @param options - Per-call options.
	* @returns A promise resolving to `{ valid: true }` or `{ valid: false, err: ValidateError }`.
	*/
	validate(json, schema, options = {}) {
		return new Promise((resolve) => {
			try {
				this._validate(json, schema, options, (err, valid) => {
					resolve({
						valid,
						err
					});
				});
			} catch (err) {
				resolve({
					valid: false,
					err
				});
			}
		});
	}
	/**
	* Validate one or more JSON Schemas.
	* @param schemaOrArr - A single schema or an array of schemas.
	* @returns `{ valid: true }` on success, or `{ valid: false, err: ValidateError }` on failure.
	*/
	validateSchema(schemaOrArr) {
		try {
			this._validateSchema(schemaOrArr);
			return { valid: true };
		} catch (err) {
			return {
				valid: false,
				err
			};
		}
	}
};
//#endregion
//#region src/index.ts
var src_default = ZSchema;
//#endregion
exports.ValidateError = ValidateError;
exports.default = src_default;
exports.getFormatValidators = getFormatValidators;
exports.getRegisteredFormats = getRegisteredFormats;
exports.getSupportedFormats = getSupportedFormats;
exports.isFormatSupported = isFormatSupported;
exports.registerFormat = registerFormat;
exports.unregisterFormat = unregisterFormat;
