// z-schema/dist/format-validators.mjs
import { sortedKeys } from "./utils/json.mjs";
import { isValidRfc3339Date } from "./utils/date.mjs";
import { isValidHostname, isValidIdnHostname } from "./utils/hostname.mjs";
import { parseRfc3339Time } from "./utils/time.mjs";
import isEmailModule from "validator/lib/isEmail.js";
import isIPModule from "validator/lib/isIP.js";
import isURLModule from "validator/lib/isURL.js";
//#region src/format-validators.ts
const dateValidator = (date) => {
	if (typeof date !== "string") return true;
	const matches = /^([0-9]{4})-([0-9]{2})-([0-9]{2})$/.exec(date);
	if (matches === null) return false;
	return isValidRfc3339Date(parseInt(matches[1], 10), parseInt(matches[2], 10), parseInt(matches[3], 10));
};
const dateTimeValidator = (dateTime) => {
	if (typeof dateTime !== "string") return true;
	const s = dateTime.toLowerCase().split("t");
	if (s.length !== 2) return false;
	const datePart = s[0];
	const timePart = s[1];
	const dateMatches = /^([0-9]{4})-([0-9]{2})-([0-9]{2})$/.exec(datePart);
	if (dateMatches === null) return false;
	if (!isValidRfc3339Date(parseInt(dateMatches[1], 10), parseInt(dateMatches[2], 10), parseInt(dateMatches[3], 10))) return false;
	return parseRfc3339Time(timePart) !== null;
};
const emailValidator = (email) => {
	if (typeof email !== "string") return true;
	if (isEmailModule.default(email, {
		require_tld: true,
		allow_ip_domain: true
	})) return true;
	const ipv6Literal = /^(.+)@\[IPv6:([^\]]+)\]$/i.exec(email);
	if (!ipv6Literal) return false;
	const localPart = ipv6Literal[1];
	const addressPart = ipv6Literal[2];
	if (!isIPModule.default(addressPart, 6)) return false;
	return isEmailModule.default(`${localPart}@example.com`, { require_tld: true });
};
const hostnameValidator = (hostname) => {
	if (typeof hostname !== "string") return true;
	return isValidHostname(hostname);
};
const ipv4Validator = (ipv4) => {
	if (typeof ipv4 !== "string") return true;
	return isIPModule.default(ipv4, 4);
};
const ipv6Validator = (ipv6) => {
	if (typeof ipv6 !== "string") return true;
	if (ipv6.includes("%")) return false;
	return isIPModule.default(ipv6, 6);
};
const regexValidator = (input) => {
	if (typeof input !== "string") return true;
	const invalidEscapes = new Set(["a"]);
	for (let idx = 0; idx < input.length; idx++) {
		if (input[idx] !== "\\") continue;
		idx++;
		if (idx >= input.length) return false;
		const escaped = input[idx];
		if (invalidEscapes.has(escaped)) return false;
	}
	try {
		RegExp(input);
		return true;
	} catch (_e) {
		return false;
	}
};
const durationValidator = (input) => {
	if (typeof input !== "string") return true;
	if (!/^P[\x00-\x7F]*$/.test(input)) return false;
	if (!input.startsWith("P")) return false;
	const body = input.slice(1);
	if (body.length === 0) return false;
	if (body.includes("W")) return /^\d+W$/.test(body);
	const parts = body.split("T");
	if (parts.length > 2) return false;
	const datePart = parts[0];
	const timePart = parts.length === 2 ? parts[1] : void 0;
	if (datePart.length > 0 && !/^(?:\d+Y(?:\d+M(?:\d+D)?)?|\d+M(?:\d+D)?|\d+D)$/.test(datePart)) return false;
	const hasDateComponent = /\d+[YMD]/.test(datePart);
	let hasTimeComponent = false;
	if (timePart !== void 0) {
		if (timePart.length === 0) return false;
		if (!/^(?:\d+H(?:\d+M(?:\d+S)?)?|\d+M(?:\d+S)?|\d+S)$/.test(timePart)) return false;
		hasTimeComponent = /\d+[HMS]/.test(timePart);
		if (!hasTimeComponent) return false;
	}
	return hasDateComponent || hasTimeComponent;
};
const uuidValidator = (input) => {
	if (typeof input !== "string") return true;
	return /^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/.test(input);
};
const strictUriValidator = (uri) => typeof uri !== "string" || isURLModule.default(uri);
const hasValidPercentEncoding = (str) => {
	for (let i = 0; i < str.length; i++) if (str[i] === "%") {
		if (i + 2 >= str.length || !/[0-9a-fA-F]/.test(str[i + 1]) || !/[0-9a-fA-F]/.test(str[i + 2])) return false;
	}
	return true;
};
const uriValidator = function(uri) {
	if (typeof uri !== "string") return true;
	if (/[^\x00-\x7F]/.test(uri)) return false;
	if (!hasValidPercentEncoding(uri)) return false;
	const match = uri.match(/^([a-zA-Z][a-zA-Z0-9+.-]*):\/\/([^/?#]*)/);
	if (match) {
		const authority = match[2];
		const atIndex = authority.indexOf("@");
		if (atIndex > 0) {
			const userinfo = authority.substring(0, atIndex);
			if (userinfo.includes("[") || userinfo.includes("]")) return false;
		}
		let hostPort = atIndex >= 0 ? authority.substring(atIndex + 1) : authority;
		if (hostPort.startsWith("[")) {
			const bracketEnd = hostPort.indexOf("]");
			if (bracketEnd >= 0) hostPort = hostPort.substring(bracketEnd + 1);
		}
		const colonIndex = hostPort.lastIndexOf(":");
		if (colonIndex >= 0) {
			const port = hostPort.substring(colonIndex + 1);
			if (port.length > 0 && !/^\d+$/.test(port)) return false;
		}
	}
	return /^[a-zA-Z][a-zA-Z0-9+.-]*:[^"\\<>^{}^`| ]*$/.test(uri);
};
const uriReferenceValidator = (uri) => {
	if (typeof uri !== "string") return true;
	if (/[^\x00-\x7F]/.test(uri)) return false;
	return /^([a-zA-Z][a-zA-Z0-9+.-]*:)?[^"\\<>^{}^`| ]*$/.test(uri);
};
const uriTemplateValidator = (uri) => {
	if (typeof uri !== "string") return true;
	if (!/^([a-zA-Z][a-zA-Z0-9+.-]*:)?[^"\\<>^`| ]*$/.test(uri)) return false;
	let inExpression = false;
	for (let idx = 0; idx < uri.length; idx++) {
		const ch = uri[idx];
		if (ch === "{") {
			if (inExpression) return false;
			inExpression = true;
		} else if (ch === "}") {
			if (!inExpression) return false;
			inExpression = false;
		}
	}
	return !inExpression;
};
const hasValidTildeEscapes = (segment) => {
	for (let i = 0; i < segment.length; i++) if (segment[i] === "~") {
		const next = segment[i + 1];
		if (next !== "0" && next !== "1") return false;
		i++;
	}
	return true;
};
const jsonPointerValidator = (pointer) => {
	if (typeof pointer !== "string") return true;
	if (pointer === "") return true;
	if (!/^(?:\/[^/]*)+$/.test(pointer)) return false;
	const tokens = pointer.split("/").slice(1);
	for (const token of tokens) if (!hasValidTildeEscapes(token)) return false;
	return true;
};
const relativeJsonPointerValidator = (pointer) => {
	if (typeof pointer !== "string") return true;
	const match = pointer.match(/^(0|[1-9]\d*)(.*)$/);
	if (!match) return false;
	const suffix = match[2];
	if (suffix === "" || suffix === "#") return true;
	if (!suffix.startsWith("/")) return false;
	if (!/^(?:\/[^/]*)+$/.test(suffix)) return false;
	const tokens = suffix.split("/").slice(1);
	for (const token of tokens) if (!hasValidTildeEscapes(token)) return false;
	return true;
};
const timeValidator = (time) => {
	if (typeof time !== "string") return true;
	return parseRfc3339Time(time) !== null;
};
const idnEmailValidator = (email) => {
	if (typeof email !== "string") return true;
	return /^[^\s@]+@[^\s@]+$/.test(email);
};
const idnHostnameValidator = (hostname) => {
	if (typeof hostname !== "string") return true;
	return isValidIdnHostname(hostname);
};
const iriValidator = (iri) => {
	if (typeof iri !== "string") return true;
	if (!/^[a-zA-Z][a-zA-Z0-9+.-]*:[^"\\<>^{}^`| ]*$/u.test(iri)) return false;
	try {
		new URL(iri);
		return true;
	} catch (_e) {
		return false;
	}
};
const iriReferenceValidator = (iriReference) => {
	if (typeof iriReference !== "string") return true;
	return /^([a-zA-Z][a-zA-Z0-9+.-]*:)?[^"\\<>^{}^`| ]*$/u.test(iriReference);
};
const inbuiltValidators = {
	date: dateValidator,
	"date-time": dateTimeValidator,
	email: emailValidator,
	hostname: hostnameValidator,
	"host-name": hostnameValidator,
	ipv4: ipv4Validator,
	ipv6: ipv6Validator,
	regex: regexValidator,
	uri: uriValidator,
	"strict-uri": strictUriValidator,
	"uri-reference": uriReferenceValidator,
	"uri-template": uriTemplateValidator,
	"json-pointer": jsonPointerValidator,
	"relative-json-pointer": relativeJsonPointerValidator,
	time: timeValidator,
	"idn-email": idnEmailValidator,
	"idn-hostname": idnHostnameValidator,
	iri: iriValidator,
	"iri-reference": iriReferenceValidator,
	duration: durationValidator,
	uuid: uuidValidator
};
const customValidators = {};
function getFormatValidators(options) {
	return {
		...inbuiltValidators,
		...options?.strictUris ? { uri: strictUriValidator } : {},
		...customValidators,
		...options?.customFormats || {}
	};
}
function registerFormat(name, validatorFunction) {
	customValidators[name] = validatorFunction;
}
function unregisterFormat(name) {
	delete customValidators[name];
}
function getSupportedFormats(customFormats) {
	const merged = {
		...inbuiltValidators,
		...customValidators,
		...customFormats
	};
	return sortedKeys(merged).filter((key) => merged[key] != null);
}
function isFormatSupported(name, customFormats) {
	if (customFormats) {
		const custom = customFormats[name];
		if (custom === null) return false;
		if (custom != null) return true;
	}
	if (name in customValidators) return customValidators[name] != null;
	return name in inbuiltValidators;
}
function getRegisteredFormats() {
	return sortedKeys(customValidators);
}
//#endregion
export { getFormatValidators, getRegisteredFormats, getSupportedFormats, isFormatSupported, registerFormat, unregisterFormat };
