// z-schema/dist/report.mjs
import { isAbsoluteUri } from "./utils/uri.mjs";
import { MAX_ASYNC_TIMEOUT } from "./utils/constants.mjs";
import { isObject } from "./utils/what-is.mjs";
import { Errors, getValidateError } from "./errors.mjs";
import { shallowClone } from "./utils/clone.mjs";
import { get } from "./utils/json.mjs";
import { jsonSymbol, schemaSymbol } from "./utils/symbols.mjs";
//#region src/report.ts
const ASYNC_TIMEOUT_POLL_MS = 10;
var Report = class Report {
	asyncTasks = [];
	commonErrorMessage;
	__$recursiveAnchorStack = [];
	__$dynamicScopeStack = [];
	__validationResultCache = /* @__PURE__ */ new Map();
	errors = [];
	json;
	path = [];
	schemaPath = [];
	rootSchema;
	parentReport;
	options;
	reportOptions;
	validateOptions = {};
	constructor(parentOrOptions, reportOptionsOrValidate, validateOptions) {
		this.parentReport = parentOrOptions instanceof Report ? parentOrOptions : void 0;
		this.options = parentOrOptions instanceof Report ? parentOrOptions.options : parentOrOptions || {};
		if (parentOrOptions instanceof Report) {
			this.reportOptions = reportOptionsOrValidate || {};
			this.validateOptions = validateOptions || parentOrOptions.validateOptions;
			this.__$recursiveAnchorStack = [...parentOrOptions.__$recursiveAnchorStack];
			this.__$dynamicScopeStack = [...parentOrOptions.__$dynamicScopeStack];
			this.__validationResultCache = parentOrOptions.__validationResultCache;
		} else {
			this.reportOptions = {};
			this.validateOptions = reportOptionsOrValidate || {};
			this.__$recursiveAnchorStack = [];
			this.__$dynamicScopeStack = [];
			this.__validationResultCache = /* @__PURE__ */ new Map();
		}
	}
	isValid() {
		if (this.asyncTasks.length > 0) throw new Error("Async tasks pending, can't answer isValid");
		return this.errors.length === 0;
	}
	addAsyncTask(fn, args, asyncTaskResultProcessFn) {
		this.asyncTasks.push([
			fn,
			args,
			asyncTaskResultProcessFn
		]);
	}
	/**
	* Like {@link addAsyncTask}, but automatically saves the current `path` and
	* restores it around `processFn`.  This eliminates the manual
	* path-save/restore boilerplate that every async-aware validator would
	* otherwise need.
	*/
	addAsyncTaskWithPath(fn, args, processFn) {
		const pathBefore = shallowClone(this.path);
		this.asyncTasks.push([
			fn,
			args,
			(result) => {
				const backup = this.path;
				this.path = pathBefore;
				processFn(result);
				this.path = backup;
			}
		]);
	}
	getAncestor(id) {
		if (!this.parentReport) return;
		if (this.parentReport.getSchemaId() === id) return this.parentReport;
		return this.parentReport.getAncestor(id);
	}
	processAsyncTasks(timeout, callback) {
		const validationTimeout = Math.min(Math.max(timeout || 2e3, 0), MAX_ASYNC_TIMEOUT);
		const timeoutAt = Date.now() + validationTimeout;
		let tasksCount = this.asyncTasks.length;
		let timedOut = false;
		const finish = () => {
			setTimeout(() => {
				const valid = this.errors.length === 0;
				callback(valid ? void 0 : getValidateError({ details: this.errors }), valid);
			}, 0);
		};
		const respond = (asyncTaskResultProcessFn) => (asyncTaskResult) => {
			if (timedOut) return;
			asyncTaskResultProcessFn(asyncTaskResult);
			if (--tasksCount === 0) finish();
		};
		if (tasksCount === 0 || this.errors.length > 0 && this.options.breakOnFirstError) {
			finish();
			return;
		}
		for (let i = 0; i < this.asyncTasks.length; i++) {
			const [fn, fnArgs, processFn] = this.asyncTasks[i];
			const respondCallback = respond(processFn);
			fn(...fnArgs, respondCallback);
		}
		const checkTimeout = () => {
			if (timedOut || tasksCount <= 0) return;
			if (Date.now() >= timeoutAt) {
				timedOut = true;
				this.addError("ASYNC_TIMEOUT", [tasksCount, validationTimeout]);
				callback(getValidateError({ details: this.errors }), false);
				return;
			}
			setTimeout(checkTimeout, ASYNC_TIMEOUT_POLL_MS);
		};
		setTimeout(checkTimeout, ASYNC_TIMEOUT_POLL_MS);
	}
	getPath(returnPathAsString) {
		let path = [];
		if (this.parentReport) path = path.concat(this.parentReport.path);
		path = path.concat(this.path);
		if (returnPathAsString !== true) return "#/" + path.map(function(segment) {
			segment = segment.toString();
			if (isAbsoluteUri(segment)) return "uri(" + segment + ")";
			return segment.replace(/~/g, "~0").replace(/\//g, "~1");
		}).join("/");
		return path;
	}
	getSchemaPath() {
		let schemaPath = [];
		if (this.parentReport) schemaPath = schemaPath.concat(this.parentReport.schemaPath);
		schemaPath = schemaPath.concat(this.schemaPath);
		return schemaPath;
	}
	getSchemaId() {
		if (!this.rootSchema) return;
		let path = [];
		if (this.parentReport) path = path.concat(this.parentReport.path);
		path = path.concat(this.path);
		while (path.length > 0) {
			const obj = get(this.rootSchema, path);
			if (obj && obj.id) return obj.id;
			path.pop();
		}
		return this.rootSchema.id;
	}
	hasError(errCode, errParams) {
		for (let idx = 0; idx < this.errors.length; idx++) if (this.errors[idx].code === errCode) {
			let match = true;
			for (let idx2 = 0; idx2 < this.errors[idx].params.length; idx2++) if (this.errors[idx].params[idx2] !== errParams[idx2]) match = false;
			if (match) return match;
		}
		return false;
	}
	addError(errCode, errParams, subReports, schema, keyword) {
		if (!errCode) throw new Error("No errorCode passed into addError()");
		this.addCustomError(errCode, Errors[errCode], errParams, subReports, schema, keyword);
	}
	getJson() {
		if (this.json) return this.json;
		if (this.parentReport) return this.parentReport.getJson();
	}
	addCustomError(errorCode, errorMessage, params, subReports, schema, keyword) {
		if (typeof this.reportOptions.maxErrors === "number" && this.errors.length >= this.reportOptions.maxErrors) return;
		if (!errorMessage) throw new Error("No errorMessage known for code " + errorCode);
		params = params || [];
		for (let idx = 0; idx < params.length; idx++) {
			const param = params[idx] === null || isObject(params[idx]) ? JSON.stringify(params[idx]) : params[idx];
			errorMessage = errorMessage.replace("{" + idx + "}", param.toString());
		}
		const err = {
			code: errorCode,
			params,
			message: errorMessage,
			path: this.getPath(this.options.reportPathAsArray),
			schemaPath: this.getSchemaPath(),
			schemaId: this.getSchemaId(),
			keyword
		};
		err[schemaSymbol] = schema;
		err[jsonSymbol] = this.getJson();
		if (schema && typeof schema === "string") err.description = schema;
		else if (schema && typeof schema === "object") {
			if (schema.title) err.title = schema.title;
			if (schema.description) err.description = schema.description;
		}
		if (subReports != null) {
			if (!Array.isArray(subReports)) subReports = [subReports];
			err.inner = [];
			for (const subReport of subReports) for (const error of subReport.errors) err.inner.push(error);
			if (err.inner.length === 0) err.inner = void 0;
		}
		if (Array.isArray(this.validateOptions.excludeErrors) && this.validateOptions.excludeErrors.includes(errorCode)) return;
		this.errors.push(err);
	}
};
//#endregion
export { Report };
