// z-schema/dist/schema-compiler.mjs
import { getRemotePath, isAbsoluteUri } from "./utils/uri.mjs";
import { NON_SCHEMA_KEYWORDS, getId, isInternalKey } from "./json-schema.mjs";
import { Report } from "./report.mjs";
import { getSchemaReader } from "./z-schema-reader.mjs";
//#region src/schema-compiler.ts
const UNSAFE_TARGETS = [
	Object.prototype,
	Function.prototype,
	Array.prototype
];
/** Returns true if `obj` is a built-in prototype that must not be mutated. */
function isUnsafeTarget(obj) {
	return UNSAFE_TARGETS.includes(obj);
}
/** Safely assign a property on `obj`, refusing prototype-polluting keys. */
function safeSetProperty(obj, key, value) {
	if (isUnsafeTarget(obj)) return;
	/** Reject property names that could pollute Object.prototype (CWE-1321). */
	if (key !== "__proto__" && key !== "constructor" && key !== "prototype") obj[key] = value;
}
/** Safely delete a property from `obj`, refusing to mutate built-in prototypes (CWE-1321). */
function safeDeleteProperty(obj, key) {
	if (isUnsafeTarget(obj)) return;
	if (key !== "__proto__" && key !== "constructor" && key !== "prototype") delete obj[key];
}
const collectIds = (obj, maxDepth = 100) => {
	const ids = [];
	function walk(node, scope, _depth = 0) {
		if (typeof node !== "object" || node == null) return;
		if (_depth >= maxDepth) throw new Error(`Maximum recursion depth (${maxDepth}) exceeded in collectIds. If your schema is deeply nested and valid, increase the maxRecursionDepth option.`);
		let addedScope = false;
		const nodeId = getId(node);
		if (typeof nodeId === "string") {
			let type = isAbsoluteUri(nodeId) ? "absolute" : "relative";
			if (scope.length === 0) type = "root";
			const id = {
				id: nodeId,
				type,
				obj: node
			};
			if (type === "absolute" || type === "root" && isAbsoluteUri(nodeId)) id.absoluteUri = nodeId;
			else if (type === "root" && typeof node.id === "string" && isAbsoluteUri(node.id) && node.id !== nodeId) id.absoluteUri = resolveSchemaScopeId(node.id, node, nodeId);
			else if (type === "relative") {
				id.absoluteParent = scope.filter((x) => x.type === "absolute" || x.type === "root" && x.absoluteUri).slice(-1)[0];
				if (id.absoluteParent) id.absoluteUri = resolveSchemaScopeId(id.absoluteParent.absoluteUri || id.absoluteParent.id, node, id.id);
			}
			ids.push(id);
			scope.push(id);
			addedScope = true;
		}
		if (Array.isArray(node)) for (const item of node) walk(item, scope, _depth + 1);
		else for (const key of Object.keys(node)) {
			if (isInternalKey(key) || NON_SCHEMA_KEYWORDS.includes(key)) continue;
			walk(node[key], scope, _depth + 1);
		}
		if (addedScope) scope.pop();
	}
	walk(obj, []);
	return ids;
};
const collectReferences = (obj, results, scope, path, options, maxDepth = 100, _depth = 0) => {
	results = results || [];
	scope = scope || [];
	path = path || [];
	options = options || {};
	if (typeof obj !== "object" || obj === null) return results;
	if (_depth >= maxDepth) throw new Error(`Maximum recursion depth (${maxDepth}) exceeded in collectReferences. If your schema is deeply nested and valid, increase the maxRecursionDepth option.`);
	const hasRef = typeof obj.$ref === "string" && typeof obj.__$refResolved === "undefined";
	let addedScope = false;
	const isRootScope = scope.length === 0;
	let scopeId = getId(obj);
	if (typeof obj.id === "string" && isAbsoluteUri(obj.id) && (!scopeId || !isAbsoluteUri(scopeId))) scopeId = obj.id;
	if (typeof scopeId === "string" && (isRootScope || !hasRef || options.useRefObjectScope === true)) {
		const base = scope.length > 0 ? scope[scope.length - 1] : void 0;
		scope.push(resolveSchemaScopeId(base, obj, scopeId));
		addedScope = true;
	}
	if (hasRef) results.push({
		ref: resolveReference(scope[scope.length - 1], obj.$ref),
		key: "$ref",
		obj,
		path: path.slice(0)
	});
	if (typeof obj.$recursiveRef === "string" && typeof obj.__$recursiveRefResolved === "undefined") results.push({
		ref: resolveReference(scope[scope.length - 1], obj.$recursiveRef),
		key: "$recursiveRef",
		obj,
		path: path.slice(0)
	});
	if (typeof obj.$dynamicRef === "string" && typeof obj.__$dynamicRefResolved === "undefined") results.push({
		ref: resolveReference(scope[scope.length - 1], obj.$dynamicRef),
		key: "$dynamicRef",
		obj,
		path: path.slice(0)
	});
	if (typeof obj.$schema === "string" && typeof obj.__$schemaResolved === "undefined") results.push({
		ref: resolveReference(scope[scope.length - 1], obj.$schema),
		key: "$schema",
		obj,
		path: path.slice(0)
	});
	if (Array.isArray(obj)) for (let i = 0; i < obj.length; i++) {
		path.push(i);
		collectReferences(obj[i], results, scope, path, options, maxDepth, _depth + 1);
		path.pop();
	}
	else {
		const keys = Object.keys(obj);
		for (const key of keys) {
			if (isInternalKey(key) || NON_SCHEMA_KEYWORDS.includes(key)) continue;
			path.push(key);
			collectReferences(obj[key], results, scope, path, options, maxDepth, _depth + 1);
			path.pop();
		}
	}
	if (addedScope) scope.pop();
	return results;
};
const resolveReference = (base, ref) => {
	if (isAbsoluteUri(ref)) return ref;
	const baseStr = base ?? "";
	if (ref[0] === "#") {
		const hashIndex = baseStr.indexOf("#");
		return (hashIndex === -1 ? baseStr : baseStr.slice(0, hashIndex)) + ref;
	}
	if (!baseStr) return ref;
	const hashIndex = baseStr.indexOf("#");
	const baseNoFrag = hashIndex === -1 ? baseStr : baseStr.slice(0, hashIndex);
	if (isAbsoluteUri(baseNoFrag)) try {
		return new URL(ref, baseNoFrag).toString();
	} catch {}
	let baseDir = baseNoFrag;
	if (!baseDir.endsWith("/")) {
		const lastSlash = baseDir.lastIndexOf("/");
		baseDir = lastSlash === -1 ? "" : baseDir.slice(0, lastSlash + 1);
	}
	return baseDir + ref;
};
const isSimpleIdentifier = (id) => id[0] !== "#" && !id.includes("/") && !id.includes(".") && !id.includes("#");
const resolveIdScope = (base, id) => {
	if (isAbsoluteUri(id)) return id;
	const baseStr = base ?? "";
	if (isSimpleIdentifier(id)) {
		const hashIndex = baseStr.indexOf("#");
		return (hashIndex === -1 ? baseStr : baseStr.slice(0, hashIndex)) + "#" + id;
	}
	return resolveReference(base, id);
};
const resolveSchemaScopeId = (base, schema, id) => {
	if (typeof schema.$id === "string") return resolveReference(base, id);
	return resolveIdScope(base, id);
};
var SchemaCompiler = class {
	constructor(validator) {
		this.validator = validator;
	}
	collectAndCacheIds(schema) {
		const ids = collectIds(schema, this.validator.options.maxRecursionDepth);
		for (const item of ids) if (item.absoluteUri) {
			this.validator.scache.cacheSchemaByUri(item.absoluteUri, item.obj);
			if (item.type === "relative" && item.absoluteParent && isSimpleIdentifier(item.id)) {
				const altAbsoluteUri = resolveReference(item.absoluteParent.absoluteUri || item.absoluteParent.id, item.id);
				this.validator.scache.cacheSchemaByUri(altAbsoluteUri, item.obj);
			}
		} else if (item.type === "root") this.validator.scache.cacheSchemaByUri(item.id, item.obj);
	}
	compileSchema(report, schema, options) {
		report.commonErrorMessage = "SCHEMA_COMPILATION_FAILED";
		if (typeof schema === "string") {
			const loadedSchema = this.validator.scache.getSchemaByUri(report, schema);
			if (typeof loadedSchema === "undefined") {
				report.addError("SCHEMA_NOT_REACHABLE", [schema]);
				return false;
			}
			schema = loadedSchema;
		}
		if (Array.isArray(schema)) {
			if (!options?.noCache) schema.forEach((s) => this.collectAndCacheIds(s));
			return this.compileArrayOfSchemas(report, schema);
		} else if (typeof schema === "boolean") return true;
		else if (!options?.noCache) this.collectAndCacheIds(schema);
		const canMutateSchemaObject = schema !== Object.prototype && schema !== Function.prototype && schema !== Array.prototype;
		if (canMutateSchemaObject && schema.__$compiled && schema.id && this.validator.scache.checkCacheForUri(schema.id) === false) schema.__$compiled = void 0;
		if (schema.__$compiled) return true;
		if (canMutateSchemaObject && !schema.$schema && this.validator.options.version !== "none") schema.$schema = this.validator.getDefaultSchemaId();
		if (schema.id && typeof schema.id === "string" && !options?.noCache) this.validator.scache.cacheSchemaByUri(schema.id, schema);
		let isRoot = false;
		if (!report.rootSchema) {
			report.rootSchema = schema;
			isRoot = true;
		}
		const isValidExceptReferences = report.isValid();
		if (canMutateSchemaObject) safeDeleteProperty(schema, "__$missingReferences");
		const useRefObjectScope = this.validator.options.version === "draft2019-09" || this.validator.options.version === "draft2020-12";
		const refs = collectReferences(schema, void 0, void 0, void 0, { useRefObjectScope }, this.validator.options.maxRecursionDepth);
		for (const refObj of refs) {
			let response = this.validator.scache.getSchemaByUri(report, refObj.ref, schema);
			if (typeof response === "undefined") {
				const schemaReader = getSchemaReader();
				if (schemaReader) {
					const remotePath = getRemotePath(refObj.ref);
					const s = schemaReader(remotePath);
					if (s) {
						s.id = remotePath;
						const subreport = new Report(report);
						if (!this.compileSchema(subreport, s)) report.errors = report.errors.concat(subreport.errors);
						else response = this.validator.scache.getSchemaByUri(report, refObj.ref, schema);
					}
				}
			}
			if (typeof response === "undefined") {
				const hasNotValid = report.hasError("REMOTE_NOT_VALID", [refObj.ref]);
				const isAbsolute = isAbsoluteUri(refObj.ref);
				let isDownloaded = false;
				const ignoreUnresolvableRemotes = this.validator.options.ignoreUnresolvableReferences === true;
				if (isAbsolute) isDownloaded = this.validator.scache.checkCacheForUri(refObj.ref);
				if (hasNotValid) {} else if (ignoreUnresolvableRemotes && isAbsolute) {} else if (isDownloaded) {} else {
					report.path.push(...refObj.path);
					report.addError("UNRESOLVABLE_REFERENCE", [refObj.ref]);
					report.path = report.path.slice(0, -refObj.path.length);
					if (isValidExceptReferences && canMutateSchemaObject && !isUnsafeTarget(schema)) {
						schema.__$missingReferences = schema.__$missingReferences || [];
						schema.__$missingReferences.push(refObj);
					}
				}
			}
			safeSetProperty(refObj.obj, `__${refObj.key}Resolved`, response);
		}
		const isValid = report.isValid();
		if (isValid && canMutateSchemaObject) schema.__$compiled = true;
		if (isRoot) report.rootSchema = void 0;
		return isValid;
	}
	compileArrayOfSchemas(report, arr) {
		let compiled = 0, lastLoopCompiled;
		do {
			report.errors = report.errors.filter((e) => e.code !== "UNRESOLVABLE_REFERENCE");
			lastLoopCompiled = compiled;
			compiled = this.compileArrayOfSchemasLoop(report, arr);
			for (const sch of arr) if (sch.__$missingReferences) {
				for (let idx2 = sch.__$missingReferences.length - 1; idx2 >= 0; idx2--) {
					const refObj = sch.__$missingReferences[idx2];
					const response = arr.find((x) => x.id === refObj.ref);
					if (response) {
						safeSetProperty(refObj.obj, `__${refObj.key}Resolved`, response);
						sch.__$missingReferences.splice(idx2, 1);
					}
				}
				if (sch.__$missingReferences.length === 0) safeDeleteProperty(sch, "__$missingReferences");
			}
		} while (compiled !== arr.length && compiled !== lastLoopCompiled);
		return report.isValid();
	}
	compileArrayOfSchemasLoop(mainReport, arr) {
		let compiledCount = 0;
		for (const schema of arr) {
			const report = new Report(mainReport);
			if (this.compileSchema(report, schema)) compiledCount++;
			mainReport.errors = mainReport.errors.concat(report.errors);
		}
		return compiledCount;
	}
};
//#endregion
export { SchemaCompiler };
