// z-schema/dist/z-schema.mjs
import { deepClone } from "./utils/clone.mjs";
import { jsonSymbol, schemaSymbol } from "./utils/symbols.mjs";
import { defaultOptions } from "./z-schema-options.mjs";
import { SchemaCache, prepareRemoteSchema } from "./schema-cache.mjs";
import "./z-schema-versions.mjs";
import { getRegisteredFormats, registerFormat, unregisterFormat } from "./format-validators.mjs";
import { getSchemaReader, setSchemaReader } from "./z-schema-reader.mjs";
import { FACTORY_TOKEN, ZSchemaBase } from "./z-schema-base.mjs";
//#region src/z-schema.ts
var ZSchema = class ZSchema extends ZSchemaBase {
	/** @internal Use ZSchema.create() instead. */
	constructor(options, token) {
		super(options, token);
	}
	/**
	* Register a global format validator available to all instances.
	* @param name - The format name (e.g. `'email'`, `'date'`).
	* @param validatorFunction - A sync or async function `(value: unknown) => boolean | Promise<boolean>`.
	*/
	static registerFormat(name, validatorFunction) {
		return registerFormat(name, validatorFunction);
	}
	/**
	* Remove a globally registered format validator.
	* @param name - The format name to unregister.
	*/
	static unregisterFormat(name) {
		return unregisterFormat(name);
	}
	/** Returns the names of all globally registered format validators. */
	static getRegisteredFormats() {
		return getRegisteredFormats();
	}
	/** Returns a deep clone of the default options. */
	static getDefaultOptions() {
		return deepClone(defaultOptions);
	}
	/**
	* Register a remote schema in the global cache so any instance can resolve `$ref` to it.
	* @param uri - The URI the schema will be known by.
	* @param schema - The schema object or JSON string.
	* @param validationOptions - Optional options used for schema preparation.
	*/
	static setRemoteReference(uri, schema, validationOptions) {
		const _schema = prepareRemoteSchema(schema, uri, validationOptions);
		SchemaCache.cacheSchemaByUri(uri, _schema);
	}
	/** Returns the current global schema reader, or `undefined` if none is set. */
	static getSchemaReader() {
		return getSchemaReader();
	}
	/**
	* Set a global schema reader function used to resolve remote `$ref` URIs.
	* @param schemaReader - A function `(uri: string) => JsonSchema | undefined`, or `undefined` to clear.
	*/
	static setSchemaReader(schemaReader) {
		return setSchemaReader(schemaReader);
	}
	static schemaSymbol = schemaSymbol;
	static jsonSymbol = jsonSymbol;
	static create(options = {}) {
		const isAsync = options.async;
		const isSafe = options.safe;
		delete options.async;
		delete options.safe;
		if (isAsync && isSafe) return new ZSchemaAsyncSafe(options, FACTORY_TOKEN);
		if (isAsync) return new ZSchemaAsync(options, FACTORY_TOKEN);
		if (isSafe) return new ZSchemaSafe(options, FACTORY_TOKEN);
		return new ZSchema(options, FACTORY_TOKEN);
	}
	/**
	* Validate JSON data against a schema.
	* @param json - The data to validate.
	* @param schema - A JSON Schema object or a schema id string (previously registered via `validateSchema`).
	* @param options - Per-call options (`schemaPath`, `includeErrors`, `excludeErrors`).
	* @returns `true` if valid.
	* @throws {@link ValidateError} if validation fails, with a `details` array of structured errors.
	*/
	validate(json, schema, options = {}) {
		return this._validate(json, schema, options);
	}
	/**
	* Validate JSON data against a schema, returning a result object instead of throwing.
	* @param json - The data to validate.
	* @param schema - A JSON Schema object or a schema id string.
	* @param options - Per-call options.
	* @returns `{ valid: true }` on success, or `{ valid: false, err: ValidateError }` on failure.
	*/
	validateSafe(json, schema, options) {
		try {
			this._validate(json, schema, options ?? {});
			return { valid: true };
		} catch (err) {
			return {
				valid: false,
				err
			};
		}
	}
	/**
	* Validate JSON data against a schema asynchronously (supports async format validators).
	* @param json - The data to validate.
	* @param schema - A JSON Schema object or a schema id string.
	* @param options - Per-call options.
	* @returns A promise that resolves to `true` if valid.
	* @throws {@link ValidateError} if validation fails (the promise rejects).
	*/
	validateAsync(json, schema, options) {
		return new Promise((resolve, reject) => {
			try {
				this._validate(json, schema, options || {}, (err, valid) => err || valid !== true ? reject(err) : resolve(valid));
			} catch (err) {
				reject(err);
			}
		});
	}
	/**
	* Validate JSON data against a schema asynchronously, returning a result object.
	* The promise always resolves (never rejects).
	* @param json - The data to validate.
	* @param schema - A JSON Schema object or a schema id string.
	* @param options - Per-call options.
	* @returns A promise resolving to `{ valid: true }` or `{ valid: false, err: ValidateError }`.
	*/
	validateAsyncSafe(json, schema, options) {
		return new Promise((resolve) => {
			try {
				this._validate(json, schema, options || {}, (err, valid) => {
					resolve({
						valid,
						err
					});
				});
			} catch (err) {
				resolve({
					valid: false,
					err
				});
			}
		});
	}
	/**
	* Validate one or more JSON Schemas, compiling and caching them for later use with `validate()`.
	* @param schemaOrArr - A single schema or an array of schemas (for cross-referencing).
	* @returns `true` if all schemas are valid.
	* @throws {@link ValidateError} if any schema is invalid.
	*/
	validateSchema(schemaOrArr) {
		return this._validateSchema(schemaOrArr);
	}
	/**
	* Validate one or more JSON Schemas, returning a result object instead of throwing.
	* @param schemaOrArr - A single schema or an array of schemas.
	* @returns `{ valid: true }` on success, or `{ valid: false, err: ValidateError }` on failure.
	*/
	validateSchemaSafe(schemaOrArr) {
		try {
			this._validateSchema(schemaOrArr);
			return { valid: true };
		} catch (err) {
			return {
				valid: false,
				err
			};
		}
	}
};
/**
* Synchronous safe validator — `validate()` returns `{ valid, err? }` instead of throwing.
* Created via `ZSchema.create({ safe: true })`.
*/
var ZSchemaSafe = class extends ZSchemaBase {
	/** @internal Use ZSchema.create() instead. */
	constructor(options, token) {
		super(options, token);
	}
	/**
	* Validate JSON data against a schema.
	* @param json - The data to validate.
	* @param schema - A JSON Schema object or a schema id string.
	* @param options - Per-call options.
	* @returns `{ valid: true }` on success, or `{ valid: false, err: ValidateError }` on failure.
	*/
	validate(json, schema, options = {}) {
		try {
			this._validate(json, schema, options);
			return { valid: true };
		} catch (err) {
			return {
				valid: false,
				err
			};
		}
	}
	/**
	* Validate one or more JSON Schemas.
	* @param schemaOrArr - A single schema or an array of schemas.
	* @returns `{ valid: true }` on success, or `{ valid: false, err: ValidateError }` on failure.
	*/
	validateSchema(schemaOrArr) {
		try {
			this._validateSchema(schemaOrArr);
			return { valid: true };
		} catch (err) {
			return {
				valid: false,
				err
			};
		}
	}
};
/**
* Asynchronous throw validator — `validate()` returns `Promise<true>` or rejects.
* Created via `ZSchema.create({ async: true })`.
*/
var ZSchemaAsync = class extends ZSchemaBase {
	/** @internal Use ZSchema.create() instead. */
	constructor(options, token) {
		super(options, token);
	}
	/**
	* Validate JSON data against a schema asynchronously.
	* @param json - The data to validate.
	* @param schema - A JSON Schema object or a schema id string.
	* @param options - Per-call options.
	* @returns A promise that resolves to `true` if valid.
	* @throws {@link ValidateError} if validation fails (the promise rejects).
	*/
	validate(json, schema, options = {}) {
		return new Promise((resolve, reject) => {
			try {
				this._validate(json, schema, options, (err, valid) => err || valid !== true ? reject(err) : resolve(valid));
			} catch (err) {
				reject(err);
			}
		});
	}
	/**
	* Validate one or more JSON Schemas (synchronous, throws on error).
	* @param schemaOrArr - A single schema or an array of schemas.
	* @returns `true` if all schemas are valid.
	* @throws {@link ValidateError} if any schema is invalid.
	*/
	validateSchema(schemaOrArr) {
		return this._validateSchema(schemaOrArr);
	}
};
/**
* Asynchronous safe validator — `validate()` returns `Promise<{ valid, err? }>` (never rejects).
* Created via `ZSchema.create({ async: true, safe: true })`.
*/
var ZSchemaAsyncSafe = class extends ZSchemaBase {
	/** @internal Use ZSchema.create() instead. */
	constructor(options, token) {
		super(options, token);
	}
	/**
	* Validate JSON data against a schema asynchronously.
	* The promise always resolves (never rejects).
	* @param json - The data to validate.
	* @param schema - A JSON Schema object or a schema id string.
	* @param options - Per-call options.
	* @returns A promise resolving to `{ valid: true }` or `{ valid: false, err: ValidateError }`.
	*/
	validate(json, schema, options = {}) {
		return new Promise((resolve) => {
			try {
				this._validate(json, schema, options, (err, valid) => {
					resolve({
						valid,
						err
					});
				});
			} catch (err) {
				resolve({
					valid: false,
					err
				});
			}
		});
	}
	/**
	* Validate one or more JSON Schemas.
	* @param schemaOrArr - A single schema or an array of schemas.
	* @returns `{ valid: true }` on success, or `{ valid: false, err: ValidateError }` on failure.
	*/
	validateSchema(schemaOrArr) {
		try {
			this._validateSchema(schemaOrArr);
			return { valid: true };
		} catch (err) {
			return {
				valid: false,
				err
			};
		}
	}
};
//#endregion
export { ZSchema };
