// z-schema/dist/z-schema-options.mjs
import { MAX_ASYNC_TIMEOUT } from "./utils/constants.mjs";
import { shallowClone } from "./utils/clone.mjs";
import { CURRENT_DEFAULT_SCHEMA_VERSION } from "./json-schema-versions.mjs";
//#region src/z-schema-options.ts
const defaultOptions = {
	version: CURRENT_DEFAULT_SCHEMA_VERSION,
	asyncTimeout: 2e3,
	forceAdditional: false,
	assumeAdditional: false,
	enumCaseInsensitiveComparison: false,
	forceItems: false,
	forceMinItems: false,
	forceMaxItems: false,
	forceMinLength: false,
	forceMaxLength: false,
	forceProperties: false,
	ignoreUnresolvableReferences: false,
	noExtraKeywords: false,
	noTypeless: false,
	noEmptyStrings: false,
	noEmptyArrays: false,
	strictUris: false,
	strictMode: false,
	reportPathAsArray: false,
	breakOnFirstError: false,
	pedanticCheck: false,
	ignoreUnknownFormats: false,
	formatAssertions: null,
	customValidator: null,
	maxRecursionDepth: 100
};
const normalizeOptions = (options) => {
	let normalized;
	if (typeof options === "object") {
		let keys = Object.keys(options);
		for (const key of keys) if (defaultOptions[key] === void 0) throw new Error("Unexpected option passed to constructor: " + key);
		keys = Object.keys(defaultOptions);
		for (const key of keys) if (options[key] === void 0) options[key] = shallowClone(defaultOptions[key]);
		normalized = options;
	} else normalized = shallowClone(defaultOptions);
	if (normalized.asyncTimeout != null) normalized.asyncTimeout = Math.min(Math.max(normalized.asyncTimeout, 0), MAX_ASYNC_TIMEOUT);
	if (normalized.strictMode === true) {
		normalized.forceAdditional = true;
		normalized.forceItems = true;
		normalized.forceMaxLength = true;
		normalized.forceProperties = true;
		normalized.noExtraKeywords = true;
		normalized.noTypeless = true;
		normalized.noEmptyStrings = true;
		normalized.noEmptyArrays = true;
	}
	return normalized;
};
//#endregion
export { defaultOptions, normalizeOptions };
