// z-schema/src/index.ts
// Export types and interfaces from relevant files

import { ZSchema } from './z-schema.js';

export type { ErrorCode, ErrorParam, Errors } from './errors.js';
export { ValidateError } from './errors.js';
export type { FormatValidatorFn, FormatValidatorsOptions } from './format-validators.js';
export {
  getFormatValidators,
  getRegisteredFormats,
  getSupportedFormats,
  isFormatSupported,
  registerFormat,
  unregisterFormat,
} from './format-validators.js';
export type { JsonSchemaType } from './json-schema.js';
export type { JsonSchemaCommon } from './json-schema.js';
export type {
  JsonSchema,
  JsonSchemaDraft4,
  JsonSchemaDraft6,
  JsonSchemaDraft7,
  JsonSchemaDraft201909,
  JsonSchemaDraft202012,
  JsonSchemaVersion,
} from './json-schema-versions.js';
export type { Report, SchemaErrorDetail } from './report.js';
export type { ZSchema, ZSchemaAsync, ZSchemaAsyncSafe, ZSchemaSafe } from './z-schema.js';
export type { ValidateOptions, ValidateResponse } from './z-schema-base.js';
export type { ZSchemaOptions } from './z-schema-options.js';
export type { SchemaReader } from './z-schema-reader.js';

export default ZSchema;
